%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% Installation guide for regulatory Sequence Analysis Tools
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\documentclass{book}
%\documentstyle[makeidx]{book}
\makeindex
\include{rsat_latex_commands}

\begin{document}

\title{Regulatory Sequence Analysis Tools \\
Installation guide}

\author{
	Jacques van Helden \\
	\email{jvhelden@ulb.ac.be} \\
	\url{http://www.bigre.ulb.ac.be/Users/jvanheld/} \\
	\bigre
}


\maketitle

\newpage
\tableofcontents
\newpage

\chapter{Description and requirements}

\section{Description}

This documents describes the installation procedure for the software
package \textbf{Regulatory Sequence Analysis Tools} (\RSAT).

\section{Requirements}

\subsection{Operating system}

\RSAT is a unix-based package. It has been installed successfully on
the following operating systems.

\begin{enumerate}
\item Linux

\item Mac OSX

\item Sun Solaris

\item Dec Alpha

\item cygwin (under MS Windows 98) (except for the graphical
librairies, because I did not find a cygwin version of GD.pm)

\end{enumerate}

\RSAT is not compatible with any version of Microsoft Windows and I
have no intention to make it compatible in a foreseeable future. Since
most programs are written in perl, part of them might run under
windows, but some others will certainly not, because they include
calls to unix system commands.

\subsection{Perl language}

Most of the programs in \RSAT are written in perl. Version 5.1 or
later is recommended.

\subsection{Perl modules}

Some perl modules are required for the graphical tools of \RSAT, and
for some other specific programs. The perl modules can be found in the
Comprehensive Perl Archive Network (\url{http://www.cpan.org/}), or
can be installed with the command \program{cpan}.

\begin{description}
\item[GD.pm] Interface to Gd Graphics Library. Used by
  \program{XYgraph} and \program{feature-map}.

\item[PostScript::Simple] Produce PostScript files from Perl. Used by
  \program{feature-map}.

\item[Math::CDF] Is used to calculate some probability distribution
  functions. In particular, it is used by the program
  \program{position-analysis} to calculate the P-value of the
  chi2. Note that this librairy is currently restricted to a precision
  of 1e-16. For the discrete functions (binomial , Poisson,
  hypergeometric) \RSAT relies on a custom library (
  \$RSAT/perl-scripts/lib/RSAT/stats.pm) which reaches a precision of
  1e-300.

\item[Util::Properties] This module is required to load property
  files, which are used to specify the site-specific configuration of
  your \RSAT server. Property filesa are also useful to write your own
  perl clients for the Web service interface to \RSAT (RSATWS).

\item[Storable] This module is required to bring persistence to data
  structures ike organisms. The slow procedure of organisms loading
  can then be done only once. This library can be easyly installed via
  CPAN.

\item[XML::LibXML] This module is required for parsing and writing XML and uses the 
XML::Parser::Expat library. It is necessary for some RSAT applications.


\end{description}

\subsection{Additional libraries to access EnsEMBL genomes}

Two programs, \program{retrieve-ensembl-seq.pl} and
\program{get-ensembl-genome.pl}, require two EnsEMBL Perl APIs and
bioperl to work on your machine.

To obtain bioperl:

\begin{footnotesize}
\begin{verbatim}
## For this example, we install Bioperl and EnsEMBL libraries 
## in $RSAT/lib, but you can install it in some other place
mkdir -p $RSAT/lib
cd $RSAT/lib

## Login into the BioPerl CVS server
cvs -d :pserver:cvs@code.open-bio.org:/home/repository/bioperl login
### (password is 'cvs')

## download the current version
cvs -d :pserver:cvs@code.open-bio.org:/home/repository/bioperl \
  checkout bioperl-live

\end{verbatim}
\end{footnotesize}

%% (or cvs -d :pserver:cvs@code.open-bio.org:/home/repository/bioperl checkout -d bioperl-branch-1.4 -r branch-1-4 bioperl-live for older branch)

To obtain EnsEMBL:

\begin{footnotesize}
\begin{verbatim}
## Login on the EnsEMBL CVS server
cvs -d :pserver:cvsuser@cvs.sanger.ac.uk:/cvsroot/ensembl login
## (password is 'CVSUSER')

## Download the current version of EnsEMBL
cvs -d :pserver:cvsuser@cvs.sanger.ac.uk:/cvsroot/ensembl \
  checkout -r branch-ensembl-47 ensembl
## (you need to adapt the branch number, e.g. 47, as it updates).

## Download the current version of the EnsEMBL module for comparative genomics
cvs -d :pserver:cvsuser@cvs.sanger.ac.uk:/cvsroot/ensembl \
  checkout -r branch-ensembl-47 ensembl-compara
## (you need to adapt the branch number, e.g. 47, as it updates).
\end{verbatim}
\end{footnotesize}

Adapt the 3 following lines in the RSAT configuration file
\file{RSAT\_config.props} to specify the actual path of the bioperl
and ensembl libraries on your computer.

\begin{footnotesize}
\begin{verbatim}
ensembl=/home/rsat/rsa-tools/lib/ensembl/modules
compara=/home/rsat/rsa-tools/lib/ensembl-compara/modules
bioperl=/home/rsat/rsa-tools/lib/bioperl-live
\end{verbatim}
\end{footnotesize}

You also need to define the url of the ensembl database in that configuration file:

\begin{footnotesize}
\begin{verbatim}
## EnsEMBL host
## Used by the EnsEMBL-accessing tools (retrieve-ensembl-seq,
## get-ensembl-genome).
## URL of the server for the EnsEMBL DB. By default, the
## main ensembl server is called, but a local server can be specified.
ensembl_host=ensembldb.ensembl.org
\end{verbatim}
\end{footnotesize}

The EnsEMBL libraries also require the SQL client Perl module
\file{DBD::mysql}, as well as \file{DBI}, which can be installed with cpan (for this you need
root privileges), and MySQL. To access EnsEMBL versions above 47, you need
port 5306 to be opened.

Detailed information about the EnsEMBL libraries can be obtained on
the EnsEMBL web site
(\footnote{\url{http://www.ensembl.org/info/using/api/api\_installation.html}}).

\subsection{Python language}

Some of the programs in \RSAT are written in python. Version 2.4 or
later is recommended.

\subsection{Java language}

Some of the NeAT tools are written in Java. Java Runtime Environment 5 or higher is recommended.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Installation
\chapter{Obtaining \RSAT distribution}

For the time being, \RSAT is distributed as a compressed archive. In a
near future, we will also distribute it via an anonymous CVS server,
which will greatly facilitate the updates.

\textbf{Note} The CVS distribution will soon be available for external
users, but we still need to configure the CVS server to accept a guest
login. For the time being, the CVS distribution is still restricted to
the people from the lab. Inbetween, the only distribution mode for
external users is the compressed archive. If you are not member of the
BiGRe laboratory, please skip the section \textit{Installation from
  the CVS repository}.

\section{Donwloading \RSAT from the CVS server}

\textbf{Warning:} we currently cannot manage user profiles in our CVS
server, so the \RSAT code is still distributed as a compressed
archive. Please skip this section if you downloaded a compessed
archive from the Web server (a file named
\file{rsa-tools\_2009XXXX.tar.gz}, where XXXXXX indicates the date).

\subsection{CVS configuration}

You need to indicate your \program{cvs} client to use \program{ssh}
as remote shell application. For this, you can specify an environment
variable.

if your shell is tcsh, add the following line to the \file{.cshrc}
file in your home directory.

\begin{footnotesize}
\begin{verbatim}
setenv CVS_RSH ssh
\end{verbatim}
\end{footnotesize}


If your shell is bash, add the following line to the \file{.bashrc}
file in your home directory.

\begin{footnotesize}
\begin{verbatim}
export CVS_RSH=ssh
\end{verbatim}
\end{footnotesize}



\subsection{Obtaining a first version of \RSAT programs}

The following command should be used the first time you retrieve the
tools from the server (you need to replace \texttt{[mylogin]} by the
login name you received when signing the \RSAT license).

\begin{footnotesize}
\begin{verbatim}
cvs -d [mylogin]@cvs.bigre.ulb.ac.be:/cvs/rsat co rsa-tools
\end{verbatim}
\end{footnotesize}


This will create a directory \file{rsa-tools} on your computer, and
store the programs in it. Note that at this stage the programs are not
yet functional, because you still need to install genomes, which are
not included in the CVS distribution.

\subsection{Updating \RSAT programs}

Once the tools have been retrieved, you can obtain updates very
easily. For this, you need to change your directory to the rsa-tools
directory, and use the \texttt{cvs} command in the following way.

\begin{footnotesize}
\begin{verbatim}
cd rsa-tools
cvs update -d
\end{verbatim}
\end{footnotesize}



\section{Installation from a compressed archive}

Uncompress the archive containing the programs. The archive is
distributed \texttt{tar} format.

The \texttt{.tar.gz} file can be uncompressed with the command
\program{tar}, which are part of the default unix installation.

\begin{footnotesize}
\begin{verbatim}
tar -xpzf rsa-tools_yyyymmdd.tar.gz
\end{verbatim}
\end{footnotesize}

\chapter{Initializing \RSAT}

\section{Adding \RSAT to your path}

\begin{enumerate}

\item Create an environment variable named \textit{RSAT} and
  containing the path of rsa-tools.

  The way to create an environment variable depends on your shell. To
  know you shell, you can type

\begin{footnotesize}
\begin{verbatim}
echo $SHELL
\end{verbatim}
\end{footnotesize}


Now, if we assume that \RSAT have been installed in the directory

\begin{footnotesize}
\begin{verbatim}
/home/rsat/rsa-tools
\end{verbatim}
\end{footnotesize}


you should type the following command.

If your shell is bash:
\begin{footnotesize}
\begin{verbatim}
export RSAT=/home/rsat/rsa-tools
\end{verbatim}
\end{footnotesize}


If your shell is csh or tcsh, you need to type a slightly different
command:

\begin{footnotesize}
\begin{verbatim}
setenv RSAT /home/rsat/rsa-tools
\end{verbatim}
\end{footnotesize}


\item Add the path of the \RSAT \ perl scripts, python scripts and
  binaries to your path. In addition, add java jar files to your
  classpath.

If your shell is bash:
\begin{footnotesize}
\begin{verbatim}
export PATH=${PATH}:${RSAT}/bin
export PATH=${PATH}:${RSAT}/perl-scripts
export PATH=${PATH}:${RSAT}/python-scripts
export CLASSPATH=${CLASSPATH}:${RSAT}/java/lib/NeAT_javatools.jar
\end{verbatim}
\end{footnotesize}


If your shell is csh or tcsh:
\begin{footnotesize}
\begin{verbatim}
set path=($path $RSAT/bin)
set path=($path $RSAT/perl-scripts)
set path=($path $RSAT/python-scripts)
set classpath=($classpath $RSAT/java/lib/NeAT_javatools.jar)
rehash
\end{verbatim}
\end{footnotesize}


(the \texttt{rehash} command updates the list of executable programs)

\end{enumerate}

If you are using a different shell than bash, csh or tcsh, the
specification of environment variables might differ from the syntax
above.  In case of doubt, ask your system administrator how to
configure your environment variables and your path.

The specification of the environment variables and paths are required
each time you want to use \RSAT. You can add these specification to
your personal profile.  This file is normally found at the root of
your personal account, in the file \file{.bashrc} if your shell is
bash, or \file{.cshrc} if your shell is csh or tcsh. If you don't know
how to proceed, ask your system administrator.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Directories

\section{Initializing the directories}

In addition to the programs, the installation of rsa-tools requires
the creation of a few directories for storing data, access logs (for
the web server), and temporary files.

The distribution includes a series of make scripts which will
facilitate this step. You just need go to the rsa-tools directory, and
start the appropriate make file.

\begin{footnotesize}
\begin{verbatim}
cd rsa-tools
make -f makefiles/init_RSAT.mk init
\end{verbatim}
\end{footnotesize}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Config file
\section{Adapting \RSAT \  local configuration}

The \RSAT distribution comes with a template configuration file named
\file{RSAT\_config\_default.props} and located in the \file{rsa-tools}
directory.

Copy this file to create your own config file \file{RSAT\_config.props}.

\begin{footnotesize}
\begin{verbatim}
cp RSAT_config_default.props RSAT_config.props
\end{verbatim}
\end{footnotesize}


You need to edit this file and specify the parameters of your local
configuration. In particular, it is essential to specify the variable
RSAT, which specifies the RSAT main directory.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Compile C/C++ programs
\section{Compiling C programs in \RSAT}

Some of the tools available in \RSAT (info-gibbs, matrix-scan-quick,
count-words) are written in the C. The distribution only contains
the sources of these tools, because the binaries are operating
system-dependent. The programs can be compiled in a very easy way.

\begin{footnotesize}
\begin{verbatim}
cd ${RSAT}
make -f makefiles/init_RSAT.mk compile_all
\end{verbatim}
\end{footnotesize}


This will compile and install the following programs in the directory
\file{\$RSAT/bin}. 

The installation directory can be changed by redefining the BIN
variable. For instance, you can type the following command to install
the compiled programs in \file{/usr/local/bin}.

\begin{footnotesize}
\begin{verbatim}
cd ${RSAT}
make -f makefiles/init_RSAT.mk compile_all BIN=/usr/local/bin
\end{verbatim}
\end{footnotesize}


\begin{itemize}
\item \program{info-gibbs}: a gibbs sampling algorithm based on
  optimization of the the information content of the motif
  \cite{Defrance:2009}.

\item \program{count-words}: an efficient algorithm for counting word
  occurrences in DNA sequences. This program is much faster than
  \program{oligo-analysis}, but it only returns the occurrences and
  frequencies, whereas oligo-analysis returns over-representation
  statistics and supports many additional
  options. \program{count-words} is routinely used to compute word
  frequencies in large genome sequences, for calibrating the Markov
  models used by \program{oligo-analysis}.

\item \program{matrix-scan-quick}: an efficient algorithm for scanning
  sequences with a position-specific scoring matrix. As its name
  indicates, \program{matrix-scan-quick} is \emph{much} faster than
  the Perl script \program{matrix-scan}, but presents reduced
  functionalities (only computes the weight, returns either a list of
  sites or the weight score distribution).
\end{itemize}


% To compile info-gibbs.
% \verb|cd| to the \verb|contrib/info-gibbs|
% directory and run

% \begin{verbatim}
% cd ${RSAT}/contrib/info-gibbs
% make all
% ln -s info-gibbs ${RSAT}/bin/info-gibbs
% \end{verbatim}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Downloading genomes from RSAT main server

\input{ig_downloading_genomes.tex}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Tests for the command line tools
\chapter{Testing the command-line tools}

\section{Testing the access to the programs}

\subsection{Perl scripts}

From now on, you should be able to use the perl scripts from the
command line. To test this, run:

\begin{footnotesize}
\begin{verbatim}
random-seq -help
\end{verbatim}
\end{footnotesize}


This should display the on-line help for the random sequence
generator.

\begin{footnotesize}
\begin{verbatim}
random-seq -l 200 -r 4 -a a:t 0.3 c:g 0.2
\end{verbatim}
\end{footnotesize}

Should generate a random sequence of 200 nucleotides, with approximate
proportions of 60\% A/T and 40\% G/C.


\subsection{Testing Perl graphical librairies}

\RSAT includes some graphical tools (\program{feature-map} and
\program{XYgraph}), which require a proper installation of Perl
modules.

\begin{description}
\item[GD.pm] Interface to Gd Graphics Library.
\item[PostScript::Simple]  Produce PostScript files from Perl.
\end{description}

To test if these modules are available on your machine, type.

\begin{footnotesize}
\begin{verbatim}
feature-map -help
\end{verbatim}
\end{footnotesize}

If the modules are available, you should see the help message of the
program feature-map. If not, you will see an error message complaining
about the missing librairies. In such a case, ask your system
administrator to install the missing modules.

\subsection{Python scripts}

From 2009 releases, the \RSAT distribution includes some Python
scripts. To test that they are running correctly, you can try.

\begin{footnotesize}
\begin{verbatim}
random-motif  -l 10 -c 0.85 -N 20
\end{verbatim}
\end{footnotesize}

This command will generate a 20-columns position-specific scoring
matrix (PSSM) with 85\% conservation of one residue in each column.

\subsection{C programs}

You can test the correct installation of the C programs with the
following command.

\begin{footnotesize}
\begin{verbatim}
random-seq -l 1000 -a a:t 0.4 c:g 0.1 | count-words -l 2 -v 1 -2str
\end{verbatim}
\end{footnotesize}

The first program (\program{random-seq}) is a Perl script, which
generates a random sequence. The output is directly piped to the C
program \program{count-words}, which computes the frequencies and
occurrences of each dinucleotide.

\section{Testing genome installation}

We will now test if the genomes are correctly installed. You can
obtain the list of supported organisms with the command:

\begin{footnotesize}
\begin{verbatim}
supported-organisms
\end{verbatim}
\end{footnotesize}


If this command returns no result, it means that genomes were either
not installed, or not correctly configured. In such a case, check the
directories in the \file{data/genomes} directory, and check that the
file \file{data/supported\_organisms.pl}.

Once you can obtain the list of installed organisms, try to retrieve
some upstream sequences. You can first read the list of options for the
\program{retrieve-seq} program.

\begin{footnotesize}
\begin{verbatim}
retrieve-seq -help
\end{verbatim}
\end{footnotesize}


Select an organism (say \org{Saccharomyces cerevisiae}), and
retrieve all the start codons with the following options :

\begin{footnotesize}
\begin{verbatim}
retrieve-seq -org Saccharomyces_cerevisiae \
        -type upstream -from 0 -to +2 -all \
        -format wc -nocomment
\end{verbatim}
\end{footnotesize}


This should return a set of 3 bp sequences, mostly ATG (in the case of
\org{Saccharomyces cerevisiae} at least)

\section{Further steps}

The installation is now finished, you can start the user's guide.

In case you would like to install additional genomes that are not
supported on \RSAT main server, the next chapter indicates yiou how to
proceed.


\chapter{Installing third-party programs}

\section{complementary programs for the analysis of regulatory
  sequences}

The \RSAT distribution only contains the programs developed by the
\RSAT \ team. 

A few additional programs, developed by third parties, can be
integrated in the package. All third-party programs may be loacated in
the directory \emph{bin} directory of the \RSAT distribution. In order
to obtain these programs, please download them from their original
site.

In particular, we recommend to install the following programs.

\begin{description}
\item[\program{vmatch}]: developed by Stefan Kurtz, is used used by the
program \program{purge-sequences}, for the detection of sequence repeats.

\item[\program{seqlogo}]:  developed by Thomas D. Schneider, is used
used by the program \program{convert-matrix} to generate logos. It can
be downloaded from $<$http://weblogo.berkeley.edu/$>$. \program{seqlogo} is
the command-line version of \program{WebLogo}.

Download the source code archive and uncompress it. Copy the following
 files to the directory \emph{bin} of your \RSAT distribution:
\program{seqlogo, logo.pm, template.pm} and \program{template.eps}.

\program{seqlogo} requires a recent version of \program{gs} (ghostscript)
$<$http://www.cs.wisc.edu/\~{}ghost$>$ to create PNG and PDF output, and
\program{ImageMagic's convert} $<$http://www.imagemagick.org$>$ to create GIFs.

\item[\program{patser}]: developed by Jerry Hertz, is used for
  matrix-based pattern matching.

\item[matrix-based pattern discovery]: several other pattern discovery
  programs have been embedded in the \RSAT program
  \program{multiple-family-analysis}:
\program{consensus} (Jerry Hertz),
\program{meme} (Tim Bailey),
\program{MotifSampler} (Gert Thijs),
\program{gibbs} (Andrew Neuwald).


\end{description}


\begin{table}
\begin{center}
\begin{tabular}{lll}
\hline
Program & author  & URL \\
\hline
vmatch & Stefan Kurtz & http://www.vmatch.de/ \\
seqlogo & Thomas Sneider & http://weblogo.berkeley.edu/ \\
patser & Jerry Hertz & ftp://ftp.genetics.wustl.edu/pub/stormo/Consensus/ \\
consensus & Jerry Hertz &  ftp://ftp.genetics.wustl.edu/pub/stormo/Consensus/ \\
meme & Tim Bailey & http://meme.sdsc.edu/meme/website/meme-download.html \\
MotifSampler & Gert Thijs & http://www.esat.kuleuven.ac.be/$\tilde{\hspace{0.4em}}$thijs/download.html \\
gibbs & Andrew Neuwald & ftp://ftp.ncbi.nih.gov/pub/neuwald/gibbs9\_95/ \\
\hline
\end{tabular}
\end{center}
\caption{\label{table:other_programs} Programs from other developers
  which are complementary to the \RSAT package.}
\end{table}

We particularly recommend the installation of \program{mkvtree} and
\program{vmatch} (Stefan Kurtz), because these programs are used by
the program purge-seq to discard redundant sequence fragments.

In order to add functionalities to \RSAT, install some or all of these
programs and include their binaries path rsa-tools/bin. If you are not
familiar with the installation of unix programs, ask assistance to
your system administrator.


\section{Recommended programs for the Network Analysis Tools (NeAT)}


For the Network Analysis Tools (NeAT), we recommend to install the
following programs, which offer complementary functionalities for the
analysis of networks/graphs.

Some programs come in the contrib directory of the \RSAT
distribution. Some others have to be downloaded from their original
distribution site.

To compile the program \program{floydwarshall} located in the
\file{contrib/floydwarshall} directory of \RSAT, use this command :

\begin{footnotesize}
\begin{verbatim}
gcc $RSAT/contrib/floydwarshall/floydwarshall.c -o $RSAT/bin/floydwarshall
\end{verbatim}
\end{footnotesize}


In addition, the contributed programs REA and kWalks need to be installed.

To install kWalks, type:

\begin{footnotesize}
\begin{verbatim}
cd $RSAT/contrib/kwalks
tar xzvf kwalks.tgz
cd src
make
\end{verbatim}
\end{footnotesize}


The KWALKS\_ROOT variable in the RSAT config file should point to
\$RSAT/contrib/kwalks/bin

To install REA, type:

\begin{footnotesize}
\begin{verbatim}
cd $RSAT/contrib/REA
tar xzvf REA.tgz
make
\end{verbatim}
\end{footnotesize}

The REA\_ROOT variable in the RSAT config file should point to
\$RSAT/contrib/REA

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Parsing and installing additional genomes

\input{ug_installing_organisms.tex}

% Bibliography is not working yet. I have to debug
\bibliographystyle{plain}
\bibliography{rsat_bibliography}

\end{document}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

