%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% Installation guide for regulatory Sequence Analysis Tools
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%\documentclass{book}
\documentclass[12pt,a4paper, oneside]{scrreprt} % KOMA-class Neukam and Kohm, scrbook alternatively
%\documentstyle[makeidx]{book}
\makeindex
\include{rsat_latex_commands}

\begin{document}

\RSATtitlePage{Installation guide}

% \title{Regulatory Sequence Analysis Tools \\
% Installation guide}
% \author{
% 	Jacques van Helden \\
% 	\email{jvhelden@ulb.ac.be} \\
% 	\bigre
% }
% \maketitle

\newpage
\tableofcontents
\newpage

\chapter{Description and requirements}

\section{Description}

This documents describes the installation procedure for the software
suite \textbf{Regulatory Sequence Analysis Tools} (\RSAT), which
offers several dozens of tools to detect cis-regulatory elements in
DNA sequences
\cite{Thomas-Chollier:2008:W119-27,vanHelden:2003:3593-6,vanHelden:2000:177-87}.

\section{Requirements}

\subsection{Operating system}

\RSAT is a unix-based package. It has been installed successfully on
the following operating systems.

\begin{enumerate}
\item Linux

\item Mac OSX

\item Sun Solaris

\item Dec Alpha

\end{enumerate}

\RSAT is not compatible with any version of Microsoft Windows and I
have no intention to make it compatible in a foreseeable future. Since
most programs are written in perl, part of them might run under
windows, but some others will certainly not, because they include
calls to unix system commands.

\subsection{Perl language}

Most of the programs in \RSAT are written in perl. Version 5.1 or
later is recommended. A set of Perl modules is required, the \RSAT
package includes a script to install them automatically (see
Chapter~\ref{chap:perl_modules}).

\subsection{Python language}

Some of the programs in \RSAT are written in python. Version 2.4 or
later is recommended.

\subsection{Java language}

Some of the NeAT tools are written in Java. Java Runtime Environment 5
or higher is recommended.

\subsection{Helper applications}

\subsubsection{wget}

The program \program{wget}, is used  to download
\begin{enumerate}

\item some helper programs developed by third-parties, which can be
  installed in \RSAT;

\item genomes from the \RSAT server to your local \RSAT installation.

\end{enumerate}

\program{wget} is part of linux distribution. If it is not installed
on your computer, you can download it from
\url{http://www.gnu.org/software/wget/}.  An installation package for
Mac OSX can be found at
\url{http://download.cnet.com/Wget/3000-18506_4-128268.html}.


\subsubsection{gnuplot}

The standard version of the \RSAT program \program{XYgraph} export
figures in bitmapformat (png, jpeg). If you want to support vectorial
drawings (pdf), which give a much better resolution for printing, you
need to install the freeware software \program{gnuplot} (4.2 or
later), which can be downloaded from \url{http://www.gnuplot.info/}.



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Installation
\chapter{Obtaining \RSAT distribution}

For the time being, \RSAT is distributed as a compressed archive. In a
near future, we will also distribute it via an anonymous CVS server,
which will greatly facilitate the updates.

\textbf{Note} The CVS distribution will soon be available for external
users, but we still need to configure the CVS server to accept a guest
login. For the time being, the CVS distribution is still restricted to
the people from the lab. Inbetween, the only distribution mode for
external users is the compressed archive. If you are not member of the
BiGRe laboratory, please skip the section \textit{Installation from
  the CVS repository}.

\section{Downloading \RSAT from the CVS server}

\textbf{Warning:} we currently cannot manage user profiles in our CVS
server, so the \RSAT code is still distributed as a compressed
archive. Please skip this section if you downloaded a compessed
archive from the Web server (a file named
\file{rsa-tools\_2009XXXX.tar.gz}, where XXXX indicates the date).

\subsection{CVS configuration}

You need to indicate your \program{cvs} client to use \program{ssh}
as remote shell application. For this, you can specify an environment
variable.

\emph{If your shell is tcsh}, add the following line to the
\file{.cshrc} file in your home directory.

\begin{lstlisting}
setenv CVS_RSH ssh
\end{lstlisting}


\emph{If your shell is bash}, add the following line to the
\file{.bashrc} file in your home directory.

\begin{lstlisting}
export CVS_RSH=ssh
\end{lstlisting}



\subsection{Obtaining a first version of \RSAT programs}

The following command should be used the first time you retrieve the
tools from the server (you need to replace \texttt{[mylogin]} by the
login name you received when signing the \RSAT license).

\begin{lstlisting}
cvs -d [mylogin]@cvs.bigre.ulb.ac.be:/cvs/rsat co rsa-tools
\end{lstlisting}


This will create a directory \file{rsa-tools} on your computer, and
store the programs in it. Note that at this stage the programs are not
yet functional, because you still need to install genomes, which are
not included in the CVS distribution.

\subsection{Updating \RSAT programs}

Once the tools have been retrieved, you can obtain updates very
easily. For this, you need to change your directory to the rsa-tools
directory, and use the \texttt{cvs} command in the following way.

\begin{lstlisting}
cd rsa-tools
cvs update -d
\end{lstlisting}

\section{Installation from a compressed archive}

Uncompress the archive containing the programs. The archive is
distributed \texttt{tar} format.

The \texttt{.tar.gz} file can be uncompressed with the command
\program{tar}, which are part of the default unix installation.

\begin{lstlisting}
tar -xpzf rsa-tools_yyyymmdd.tar.gz
\end{lstlisting}

\chapter{Initializing \RSAT}

\input{ug_configuration.tex}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Directories

\section{Initializing the directories}

In addition to the programs, the installation of rsa-tools requires
the creation of a few directories for storing data, access logs (for
the web server), and temporary files.

The distribution includes a series of make scripts which will
facilitate this step. You just need go to the rsa-tools directory, and
start the appropriate make file.

\begin{lstlisting}
cd $RSAT
make -f makefiles/init_RSAT.mk init
\end{lstlisting}
% $

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Config file
\section{Adapting \RSAT \  local configuration}

The \RSAT distribution comes with a template configuration file named
\file{RSAT\_config\_default.props} and located in the \file{rsa-tools}
directory.

Copy this file to create your own config file \file{RSAT\_config.props}.

\begin{lstlisting}
cp RSAT_config_default.props RSAT_config.props
\end{lstlisting}


\subsection{Editing the \RSAT\ local configuration file}

You need to edit the file \file{RSAT\_config.props} and specify the
parameters of your local configuration. In particular, \emph{it is
  crucial to specify the full path of the variable RSAT}, which
specifies the RSAT main directory.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Check the RSAT paths
\subsection{Checking the RSAT path}

The \RSAT programs should now be included in your path.  To check if
this is done properly, just type:

\begin{lstlisting}
random-seq -l 350
\end{lstlisting}

If your configuration is correct, this command should return a random
sequence of 350 nucleotides.

Don't worry if you see a warning looking like this:

\begin{footnotesize}
\begin{verbatim}
; WARNING	The tabular file with the list of supported organism cannot be read
; WARNING	Missing file	/no_backup/rsa-tools/public_html/data/supported_organisms.tab
\end{verbatim}
\end{footnotesize}

This warning will disappear after we download the first organism in
\RSAT.

You are now able to use any program from the \RSAT package, untill you
quit your session. It is however not very convenient to set the path
manually each time you open a new connection. You can modify your
default configuration by including the above commands in the file
\file{.cshrc} (in tcsh) or \file{.bashrc} (in bash) which should be
found at the root of your home directory. If you don't know how to
modify this file, see the system administrator.

\subsection{Checking the java path (required for pathway tools only)}

The \program{java} language is necessary for the path finding tools
included in the \program{NeAT} suite. If you want to use those tools,
you can check the correct setting for the java-based pathway tools, by
typing the following command.

\begin{lstlisting}
java graphtools.util.ListTools
\end{lstlisting}

This should display the list of available pathway analysis tools,
which are part of the NeAT package.



\chapter{Installing Perl modules}
\label{chap:perl_modules}

Some Perl modules are required for the graphical tools of \RSAT, and
for some other specific programs. The perl modules can be found in the
Comprehensive Perl Archive Network (\url{http://www.cpan.org/}), or
can be installed with the command \program{cpan}.

\section{Before installing Perl modules: install the GD library}

The Perl module GD.pm requires the prior installation of the GD
library. 

\begin{itemize}

\item On \emph{Linux} systems, this library can be installed with the
  package manager of your distribution (e.g. \command{apt-get}).


\item On \emph{Mac OSX} systems, the installation of the GD library is
  quite tricky. The bet is to follow the installation protocol of the
  libgd Web site: \url{http://www.libgd.org/DOC_INSTALL_OSX}.

\end{itemize}

\section{Automatic installation of Perl modules}

The simplest way to install all the required Perl modules is to tye
the command below. \emph{Beware:} this command sudo requires
administrator rights on the computer. If you don't have the root
password, please consult your system administrator.

\begin{lstlisting}
cd ${RSAT}

## Display the list of Perl modules that will be installed
make -f makefiles/install_RSAT.mk list_perl_modules

## Install the Perl modules
make -f makefiles/install_RSAT.mk install_perl_modules
\end{lstlisting}
%% closing $

\section{Utilization of the Perl modules in \RSAT}

For information, we describe hereafter the list of modules that will
be installed with this command, and the reason why it is useful to
install them before running \RSAT programs.

If you are not interested by technical details, you can skip this
section.


\begin{enumerate}
\item \program{GD.pm} Interface to Gd Graphics Library. Used by
  \program{XYgraph} and \program{feature-map}. 

\item \program{PostScript::Simple} Produce PostScript files from Perl. Used by
  \program{feature-map}.

\item \program{Statistics::Distributions} is used to calculate some probability
  distribution functions. In particular, it is used by the program
  \program{position-analysis} to calculate the P-value of the
  chi2. 

  \textbf{Notes} 
  \begin{itemize}
  \item In previous releases, the chi2 P-value was computed using
    \program{Math::CDF}, but the precision was limited to
    1e-15. \program{Statistics::Distributions} can compute P-values
    down to 1e-65.
  \item For the discrete functions (binomial , Poisson,
    hypergeometric) \RSAT relies on a custom library (
    \$RSAT/perl-scripts/lib/RSAT/stats.pm) which reaches a precision
    of 1e-300.
  \end{itemize}

\item \program{File::Spec}, \program{POSIX} and \program{Data::Dumper}
  are required for some functions of \program{matrix-scan}.

\item \program{Util::Properties} is required to load property files, which are
  used to specify the site-specific configuration of your \RSAT
  server. Property filesa are also useful to write your own perl
  clients for the Web service interface to \RSAT (RSATWS).

\item \program{Class::Std::Fast} and \program{Storable} are required
  to bring persistence to data structures like organisms. This library
  can be easyly installed via CPAN.

\item \program{XML::LibXML} is required for parsing and writing XML and uses
  the XML::Parser::Expat library. It is necessary for some RSAT
  applications.

\item \program{DBD::mysql} and \program{DBI} those two libraries are
  required by the program \program{retrieve-ensembl-seq} in order to
  access the ENSEMBL database.

\item \program{SOAP::WSDL}, \program{Module::Build::Compat} and
  \program{Util::Properties} (already mentioned above) are required
  for the Web services. \RSAT Web services is a convenient interface
  that permits to write Perl scripts to run \RSAT queries on a remote
  server.

\item \program{Bio::Perl} is required for the Ensembl API, which in
  turn is required for handling genomes installed on the ENSEMBL
  database (\url{http://www.ensembl.org/}).

\end{enumerate}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Compile C/C++ programs
\section{Compiling C programs in \RSAT}

Some of the tools available in \RSAT (\program{info-gibbs},
\program{matrix-scan-quick}, \program{count-words}) are written in the
\program{C} language. The distribution only contains the sources of
these tools, because the binaries are operating system-dependent. The
programs can be compiled in a very easy way.

\begin{lstlisting}
cd ${RSAT}
make -f makefiles/init_RSAT.mk compile_all
\end{lstlisting}

This will compile and install the following programs in the directory
\file{\$RSAT/bin}. 

The installation directory can be changed by redefining the BIN
variable. For instance, if you have the system adminstrator
privileges, you could install the compiled programs in the standard
directory for compiled packages (\file{/usr/local/bin}).

\begin{lstlisting}
cd ${RSAT}
make -f makefiles/init_RSAT.mk compile_all BIN=/usr/local/bin
\end{lstlisting}


\begin{itemize}
\item \program{info-gibbs}: a gibbs sampling algorithm based on
  optimization of the the information content of the motif
  \cite{Defrance:2009}.

\item \program{count-words}: an efficient algorithm for counting word
  occurrences in DNA sequences. This program is much faster than
  \program{oligo-analysis}, but it only returns the occurrences and
  frequencies, whereas oligo-analysis returns over-representation
  statistics and supports many additional
  options. \program{count-words} is routinely used to compute word
  frequencies in large genome sequences, for calibrating the Markov
  models used by \program{oligo-analysis}.

\item \program{matrix-scan-quick}: an efficient algorithm for scanning
  sequences with a position-specific scoring matrix. As its name
  indicates, \program{matrix-scan-quick} is \emph{much} faster than
  the Perl script \program{matrix-scan}, but presents reduced
  functionalities (only computes the weight, returns either a list of
  sites or the weight score distribution).
\end{itemize}


\section{Compiling the Stubb for Web services}

If you want to use the Web services, you need to compile the Stubb in
order to synchronize the Perl modules of your client with the service
specification of the server.


\begin{lstlisting}
## Change directory to the sample Perl Web services clients
cd $RSAT/ws_clients/perl_clients

## Compile the stubb.
## Your computer will open a connection to the Web services server 
## and collect the WSDL specifications of the supported services.
make stubb

## Test the connection to the Web services
make test
\end{lstlisting}

%$

\section{Importing and installing programs from third parties}

In addition to the binaries incorporated in the suite itself, \RSAT
uses some programs developed by third parties for specific
tasks. Those programs can be downloaded and installed automatically
using the makefile \file{install\_rsat.mk}.

Before doing this, you must make sure that the program \program{wget}
(For Linux: \url{http://www.gnu.org/software/wget/}; for Mac OSX
\url{http://download.cnet.com/Wget/3000-18506_4-128268.html}).

\begin{lstlisting}
cd $RSAT
make -f makefiles/install_rsat.mk install_ext_apps
\end{lstlisting}
%$

\section{Libraries required to support EnsEMBL genomes}
\label{sect:ensembl_libraries}

Since 2009, a series of \RSAT programs support a direct access to the
EnsEMBL database in order to ensure a convenient access to genomes
from higher organisms \cite{Sand:2009}.

\begin{itemize}
\item \program{supported-organisms-ensembl}
\item \program{ensembl-org-info}
\item \program{retrieve-ensembl-seq.pl} 
\item \program{get-ensembl-genome.pl}
\end{itemize}

Those programs require to install a few Perl libraries as well as a
MySQL client on your machine. 

The first requirement is the \program{BioPerl} module, which has in
principle been installed in Chapter~\ref{chap:perl_modules}).

% To obtain BioPerl, you first need to install the program \program{git}
% (\url{http://git-scm.com/}). You can then execute the following
% instructions.

% \begin{lstlisting}
% ## For this example, we install Bioperl and EnsEMBL libraries 
% ## in $RSAT/lib, but you can install it in some other place
% mkdir -p $RSAT/lib
% cd $RSAT/lib


% ## Easy installation using GIT, protocol from http://www.bioperl.org/wiki/Using_Git
% git clone git://github.com/bioperl/bioperl-live.git

% ## To test the version of BioPerl that you are running, use:
% perl -MBio::Perl -le 'print Bio::Perl->VERSION;'
% \end{lstlisting}
% %%$

To obtain EnsEMBL \footnote{Full instructions at
  \url{http://useast.ensembl.org/info/docs/api/api_cvs.html}}.

\begin{lstlisting}

## Make sure you start from the right directory
cd $RSAT/lib

## Login on the EnsEMBL CVS server
cvs -d :pserver:cvsuser@cvs.sanger.ac.uk:/cvsroot/ensembl login
## (password is CVSUSER)

## Download the current version of EnsEMBL
cvs -d :pserver:cvsuser@cvs.sanger.ac.uk:/cvsroot/ensembl \
  checkout -r branch-ensembl-61 ensembl
## (you need to adapt the branch number, e.g. 61, as it updates).

## Download the current version of the EnsEMBL module for comparative genomics
cvs -d :pserver:cvsuser@cvs.sanger.ac.uk:/cvsroot/ensembl \
  checkout -r branch-ensembl-61 ensembl-compara
## (you need to adapt the branch number, e.g. 61, as it updates).
\end{lstlisting}
%%$ 

Adapt the 3 following lines in the RSAT configuration file
\file{RSAT\_config.props} to specify the actual path of the bioperl
and ensembl libraries on your computer. \emph{The path must be adapted
  to fit your local configuration}.

\begin{lstlisting}
ensembl=[RSAT_PARENT_PATH]/rsa-tools/lib/ensembl/modules
compara=[RSAT_PARENT_PATH]/rsa-tools/lib/ensembl-compara/modules
bioperl=[RSAT_PARENT_PATH]/rsa-tools/lib/bioperl-live
\end{lstlisting}

You also need to define the URL of the Ensembl database in that
configuration file:

\begin{lstlisting}
## EnsEMBL host
## Used by the EnsEMBL-accessing tools (retrieve-ensembl-seq,
## get-ensembl-genome).
## URL of the server for the EnsEMBL DB. By default, the
## main ensembl server is called, but a local server can be specified.
ensembl_host=ensembldb.ensembl.org
\end{lstlisting}

Finally, you need to include the BioPerl and Ensembl librairies in the
Perl module path (specified by the environment variable named
\$PERL5LIB).

If your shell is bash, add the following lines in the file
\file{.bashrc} at the root of your account.

\begin{lstlisting}
export PERL5LIB=${PERL5LIB}:${RSAT}/lib/ensembl/modules
export PERL5LIB=${PERL5LIB}:${RSAT}/lib/ensembl-compara/modules
export PERL5LIB=${PERL5LIB}:${RSAT}/lib/ensembl-variation/modules
\end{lstlisting}

If your shell is tcsh or chs, add the following lines in the file
\file{.cshrc} at the root of your account.

\begin{lstlisting}
setenv PERL5LIB ${PERL5LIB}:${RSAT}/lib/ensembl/modules
setenv PERL5LIB ${PERL5LIB}:${RSAT}/lib/ensembl-compara/modules
setenv PERL5LIB ${PERL5LIB}:${RSAT}/lib/ensembl-variation/modules
\end{lstlisting}



The EnsEMBL libraries also require the SQL client Perl module
\file{DBD::mysql}, as well as \file{DBI}, which can be installed with
\program{cpan} (for this you need root privileges). Note that the
installation of the CPAN module DBD:mysql requires a prior
installation of a MySQL client on your machine
(\url{http://dev.mysql.com/downloads/}).

To access EnsEMBL versions above 47, you need port 5306 to be opened
on your machine. This might require to consult the system
administrator of your network in order to ensure that the Firewall
accepts this port.

Detailed information about the EnsEMBL libraries can be obtained on
the EnsEMBL web site
(\footnote{\url{http://www.ensembl.org/info/using/api/api\_installation.html}}).



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Downloading genomes from RSAT main server

\input{ig_downloading_genomes.tex}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Tests for the command line tools
\input{ig_testing_config}


%% \section{Further steps}

%% The installation is now finished, you can start the user's guide.

%% In case you would like to install additional genomes that are not
%% supported on \RSAT main server, the next chapter indicates yiou how to
%% proceed.


\chapter{Installing third-party programs}

\section{complementary programs for the analysis of regulatory
  sequences}

The \RSAT distribution only contains the programs developed by the
\RSAT \ team. 

A few additional programs, developed by third parties, can be
integrated in the package. All third-party programs may be loacated in
the directory \emph{bin} directory of the \RSAT distribution. In order
to obtain these programs, please download them from their original
site.

In particular, we recommend to install the following programs.

\begin{description}
\item[\program{vmatch}]: developed by Stefan Kurtz, is used used by the
program \program{purge-sequences}, for the detection of sequence repeats.

\item[\program{seqlogo}]:  developed by Thomas D. Schneider, is used
used by the program \program{convert-matrix} to generate logos. It can
be downloaded from $<$http://weblogo.berkeley.edu/$>$. \program{seqlogo} is
the command-line version of \program{WebLogo}.

Download the source code archive and uncompress it. Copy the following
 files to the directory \emph{bin} of your \RSAT distribution:
\program{seqlogo, logo.pm, template.pm} and \program{template.eps}.

\program{seqlogo} requires a recent version of \program{gs}
(ghostscript\urlref{http://www.ghostscript.com/}) to create PNG and
PDF output, and \program{ImageMagic's
  convert}\urlref{http://www.imagemagick.org/} to create GIFs.

\item[\program{patser}]: developed by Jerry Hertz, is used for
  matrix-based pattern matching.

\item[matrix-based pattern discovery]: several other pattern discovery
  programs have been embedded in the \RSAT program
  \program{multiple-family-analysis}:
\program{consensus} (Jerry Hertz),
\program{meme} (Tim Bailey),
\program{MotifSampler} (Gert Thijs),
\program{gibbs} (Andrew Neuwald).


\end{description}


\begin{table}
\begin{center}
\begin{tabular}{lll}
  \hline
  Program & author  & URL \\
  \hline
  vmatch & Stefan Kurtz & \footnotesize{\url{http://www.vmatch.de/}} \\
  seqlogo & Thomas Sneider & \footnotesize{\url{http://weblogo.berkeley.edu/}} \\
  patser & Jerry Hertz & \footnotesize{\url{ftp://ftp.genetics.wustl.edu/pub/stormo/Consensus/}} \\
  consensus & Jerry Hertz &  \footnotesize{\url{ftp://ftp.genetics.wustl.edu/pub/stormo/Consensus/}} \\
  meme & Tim Bailey & \footnotesize{\url{http://meme.sdsc.edu/}} \\
  MotifSampler & Gert Thijs & \footnotesize{\url{http://www.esat.kuleuven.ac.be/~thijs/download.html}} \\
%%  gibbs & Andrew Neuwald & \footnotesize{\url{ftp://ftp.ncbi.nih.gov/pub/neuwald/gibbs9\_95/}} \\
  \hline
\end{tabular}
\end{center}
\caption{\label{table:other_programs} Programs from other developers
  which are complementary to the \RSAT package.}
\end{table}

We particularly recommend the installation of \program{mkvtree} and
\program{vmatch} (Stefan Kurtz), because these programs are used by
the program purge-seq to discard redundant sequence fragments.

In order to add functionalities to \RSAT, install some or all of these
programs and include their binaries path rsa-tools/bin. If you are not
familiar with the installation of unix programs, ask assistance to
your system administrator.


\section{Recommended programs for the Network Analysis Tools (NeAT)}


For the Network Analysis Tools (NeAT), we recommend to install the
following programs, which offer complementary functionalities for the
analysis of networks/graphs.

Some programs come in the contrib directory of the \RSAT
distribution. Some others have to be downloaded from their original
distribution site.

To compile the program \program{floydwarshall} located in the
\file{contrib/floydwarshall} directory of \RSAT, use this command :

\begin{lstlisting}
gcc $RSAT/contrib/floydwarshall/floydwarshall.c -o $RSAT/bin/floydwarshall
\end{lstlisting}


In addition, the contributed programs REA and kWalks need to be installed.

To install kWalks, type:


\begin{lstlisting}
cd $RSAT/contrib/kwalks; 
tar xzvf kwalks.tgz; 
cd src; 
make clean; make
\end{lstlisting}

You can test taht the compilation worked by running the following
command.

\begin{lstlisting}
$RSAT/contrib/kwalks/bin/lkwalk
\end{lstlisting}
%% $

This should display the help message of \program{lkwalk}.

Check that the KWALKS\_ROOT variable in the RSAT config file
(\file{\$RSAT/RSAT\_config.props}) points to the correct path (it
should be the absolute path of \file{\$RSAT/contrib/kwalks/bin/}).

To install REA, type:

\begin{lstlisting}
cd $RSAT/contrib/REA; 
tar xzvf REA.tgz; 
rm -f *.o; 
make
\end{lstlisting}

You can test taht the compilation worked by running the following
command.

\begin{lstlisting}
$RSAT/contrib/REA/REA
\end{lstlisting}

This should display the help message of \program{REA}.

Check that the REA\_ROOT variable in the RSAT config file
(\file{\$RSAT/RSAT\_config.props}) points to the correct path (it
should be the absolute path of \file{\$RSAT/contrib/REA}).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Parsing and installing additional genomes

\input{ug_installing_organisms.tex}

% Bibliography is not working yet. I have to debug
\bibliographystyle{plain}
\bibliography{rsat_bibliography}

\end{document}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

