%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% Installation guide for regulatory Sequence Analysis Tools
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\documentclass{article}
%\documentstyle[makeidx]{book}
\makeindex
\include{rsat_latex_commands}

\begin{document}

\title{Regulatory Sequence Analysis Tools \\
Installation guide}

\author{
	Jacques van Helden \\
	\email{jvanheld@scmbb.ulb.ac.be} \\
	\url{http://www.scmbb.ulb.ac.be/$\tilde{\hspace{0.4em}}$jvanheld/} \\
	\scmbb 
}


\maketitle

\newpage
\tableofcontents
\newpage

\section{Description}

This documents describes the installation procedure for the software
package \textbf{Regulatory Sequence Analysis Tools} (\RSAT).

\section{Requirements}

\subsection{Operating system}

\RSAT is a unix-based package. It has been installed successfully on
the following operating systems.

\begin{enumerate}
\item Linux

\item Mac OSX

\item Sun Solaris

\item Dec Alpha

\item cygwin (under MS Windows 98) (except for the graphical
librairies, because I did not find a cygwin version of GD.pm)

\end{enumerate}

\RSAT is not compatible with any version of Microsoft Windows and I
have no intention to make it compatible in a foreseeable future. Since
most programs are written in perl, part of them might run under
windows, but some others will certainly not, because they include
calls to unix system commands.

\subsection{Perl language}

The programs in \RSAT are written in perl. Version 5.1 or
later is recommended.

\subsection{Perl modules}

Some perl modules are required for the graphical tools of \RSAT, and
for some other specific programs.

\begin{description}
\item[GD.pm] Interface to Gd Graphics Library. Used by
  \program{XYgraph} and \program{feature-map}.
\item[PostScript::Simple] Produce PostScript files from Perl. Used by
  \program{feature-map}.
\item[bioperl] \RSAT was developed independently of the bioperl
  project, but for some recent programs, I used bioperl classes
  (e.g. for reading and exporting taxonomic trees). It is thus useful
  to install a recent version of bioperl to take benefit of the full
  functionalities of \RSAT.
\end{description}

These modules can be found in the Comprehensive Perl Archive Network
(\url{http://www.cpan.org/}). For \program{bioperl}, we recommend to
use the CVS distribution (\url{http://cvs.open-bio.org/}), which
includes more recent updates.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Installation
\section{Installation}

For the time being, \RSAT is distributed as a compressed archive. In a
near future, we will also distribute it via a CVS server, which will
greatly facilitate the updates.


\RSAT can be distributed either as a compressed archive, or via the CVS
server. The CVS distribution greatly facilitates updates.

\textbf{Note} The CVS distribution will soon be available for external
users, but we still need to configure the CVS server to accept a guest
login. For the time being, the CVS distribution is still restricted to
the people from the lab. Inbetween, the only distribution mode for
external users is the compressed archive.

\subsection{Installation from the CVS repository}


Before being able to retrieve \RSAT from the CVS repository, you need
an account on our server. For this, please contact Jacques van Helden
(\email{jvanheld@scmbb.ulb.ac.be}).

\subsubsection{First installation}

The following command should be used the first time you retrieve the
tools from the server:
\begin{verbatim}
cvs -d mylogin@cvs.scmbb.ulb.ac.be:/cvs co rsa-tools
\end{verbatim}

This will create a directory \file{rsa-tools} on your computer, and
store the programs in it. Note that at this stage the programs are not
yet functional, because you still need to install genomes, which are
not included in the CVS distribution.

\subsubsection{Updates}

Once the tools have been retrieved, you can obtain updates very
easily. For this, you need to change your directory to the rsa-tools
directory, and use the \texttt{cvs} command in the following way.

\begin{verbatim}
cd rsa-tools
cvs update .
\end{verbatim}


\subsection{Installation from a compressed archive}

Uncompress the archive containing the programs. The archive is
distributed in \texttt{zip} or \texttt{tar} format.

The \texttt{.zip} files can be uncompressed with the command
\program{unzip}.

\begin{verbatim}
unzip rsa-tools_yyyymmdd.zip
\end{verbatim}
where yyyymmdd stands for the version number (delivery date).

If the \program{unzip} command is not supported on your system, you
can uncompress the \texttt{.tar.gz} archive with the commands
\program{gunzip} and \program{tar}, which are part of the default unix
installation.

\begin{verbatim}
gunzip rsa-tools_yyyymmdd.tar.gz
tar -xpf rsa-tools_yyyymmdd.tar
\end{verbatim}

\section{Adding \RSAT to your path}

\begin{enumerate}

\item Create an environment variable named RSAT and containing the
  path of rsa-tools. 

The precise command depends on your shell. To know you shell, you can
type
\begin{verbatim}
echo $SHELL
\end{verbatim}

Now, if we assume that \RSAT have been installed in the directory

\begin{verbatim}
/home/myaccount/rsa-tools
\end{verbatim}

you should type the following command.

If your shell is bash:
\begin{verbatim}
export RSAT=/home/myaccount/rsa-tools
\end{verbatim}

If your shell is csh or tcsh:
\begin{verbatim}
setenv RSAT /home/myaccount/rsa-tools
\end{verbatim}



\item Add the path of the \RSAT perl scripts and binaries to your
  path.

If your shell is bash:
\begin{verbatim}
export PATH=${PATH}:${RSAT}/bin
export PATH=${PATH}:${RSAT}/perl-scripts
\end{verbatim}

If your shell is csh or tcsh:
\begin{verbatim}
set path=($path $RSAT/bin)
set path=($path $RSAT/perl-scripts)
rehash
\end{verbatim}

(the \texttt{rehash} command updates the list of executable programs)

\end{enumerate}

If you are using a different shell than bash, csh or tcsh, the
specification of environment variables is slightly different.  In case
of doubt, ask your system administrator how to configure your
environment variables and your path.

The specification of the environment variables and paths are required
each time you want to use \RSAT. You can add these specification to
your personal profile.  This file is normally found at the root of
your personal account, in the file \file{.bashrc} if your shell is
bash, or \file{.cshrc} if your shell is csh or tcsh. If you don't know
how to proceed, ask your system administrator.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Directories

\section{Initializing the directories}

In addition to the programs, the installation of rsa-tools requires
the creation of a few directories for storing data, access logs (for
the web server), and temporary files.  

The distribution includes a series of make scripts which will
facilitate this step. You just need go to the rsa-tools directory, and
start the appropriate make file.

\begin{verbatim}
cd rsa-tools
make -f makefiles/init_RSAT.mk init
\end{verbatim}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Config file
\section{Configuring \RSAT for utilization on the command line}

The \RSAT distribution comes with a template configuration file named
\file{RSA.config.default} and located in the \file{rsa-tools}
directory.

Copy this file to create your own config file \file{RSA.config}.

\begin{verbatim}
cp RSA.config.default RSA.config
\end{verbatim}

In principle, this default configuration file is sufficient to run the
tools on the command-line. 

You only need to edit it if you want to install a web server of the
tools, or if you want to specify custom settings (for example the
installation of additional genomes on a separate hard drive).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Genome Download

\input{ig_installing_genomes.tex}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Tests for the command line tools
\section{Testing the command-line tools}

\subsection{Testing the access to perl scripts}

From now on, you should be able to use the perl scripts from the
command line. To test this, run: 

\begin{verbatim}
random-seq -help
\end{verbatim}

This should display the on-line help for the random sequence
generator. 

\begin{verbatim}
random-seq -l 200 -r 4 -a a:t 0.3 c:g 0.2
\end{verbatim}

Should generate a random sequence.

\subsection{Testing genome installation}

We will now test if the genomes are correctly installed. You can
obtain the list of supported organisms with the command:

\begin{verbatim}
supported-organisms
\end{verbatim}

If this command returns no result, it means that genomes were either
not installed, or not correctly configured. In such a case, check the
directories in the \file{data/genomes} directory, and check that the
file \file{data/supported\_organisms.pl}.

Once you can obtain the list of installed organisms, try to retrieve
some upstream sequences. You can first read the list of options for the
\program{retrieve-seq} program.

\begin{verbatim}
retrieve-seq -help
\end{verbatim}

Select an organism (say \organism{Saccharomyces cerevisiae}), and
retrieve all the start codons with the following options :

\begin{verbatim}
retrieve-seq -org Saccharomyces_cerevisiae \
        -type upstream -from 0 -to +2 -all \
        -format wc -nocomment 
\end{verbatim}

This should return a set of 3 bp sequences, mostly ATG (in the case of
\organism{Saccharomyces cerevisiae} at least)

\subsection{Testing the graphical scripts}

\RSAT includes two graphical tools, \program{feature-map} and
\program{XYgraph}. These tools require the following  perl modules: 

\begin{description}
\item[GD.pm] Interface to Gd Graphics Library.
\item[PostScript::Simple]  Produce PostScript files from Perl.
\end{description}

To test if these modules are available on your machine, type.

\begin{verbatim}
feature-map -help
\end{verbatim}

If the modules are available, you should see the help message of the
program feature-map. If not, you will see an error message complaining
about the missing librairies. In such a case, ask your system
administrator to install the missing modules.

\subsection{Installing third-party programs}

The \RSAT distribution only contains the programs developed by Jacques
van Helden. A few additional programs, developed by third parties, can
be integrated in the package. In order to obtain these programs,
please download them from their original site.

In particular, we recommend to install the following programs.

\begin{description}
\item[\program{vmatch}]: developed by Stefan Kurtz, is used used by the
program \program{purge-sequences}, for the detection of sequence repeats.

\item[\program{patser}]: developed by Jerry Hertz, is used for
  matrix-based pattern matching.

\item[matrix-based pattern discovery]: several other pattern discovery
  programs have been embedded in the \RSAT program
  \program{multiple-family-analysis}: 
\program{consensus} (Jerry Hertz),
\program{meme} (Tim Bailey),
\program{MotifSampler} (Gert Thijs),
\program{gibbs} (Andrew Neuwald).

\end{description}


\begin{table}
\begin{center}
\begin{tabular}{lll}
\hline
Program & author  & URL \\
\hline
vmatch & Stefan Kurtz & http://www.vmatch.de/ \\
patser & Jerry Hertz & ftp://ftp.genetics.wustl.edu/pub/stormo/Consensus/ \\
consensus & Jerry Hertz &  ftp://ftp.genetics.wustl.edu/pub/stormo/Consensus/ \\
meme & Tim Bailey & http://meme.sdsc.edu/meme/website/meme-download.html \\
MotifSampler & Gert Thijs & http://www.esat.kuleuven.ac.be/$\tilde{\hspace{0.4em}}$thijs/download.html \\
gibbs & Andrew Neuwald & ftp://ftp.ncbi.nih.gov/pub/neuwald/gibbs9\_95/ \\
\hline
\end{tabular}
\end{center}
\caption{\label{table:other_programs} Programs from other developers
  which are complementary to the \RSAT package.}
\end{table}

I particularly recommend the installation of \program{mkvtree} and
\program{vmatch} (Stefan Kurtz), because these programs are used by
the program purge-seq to discard redundant sequence fragments.

In order to add functionalities to \RSAT, install some or all of these
programs and include their binaries path rsa-tools/bin. If you are not
familiar with the installation of unix programs, ask assistance to
your system administrator.


\section{Further steps}

The installation is now finished, you can start the user's guide. 

\end{document}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

