\documentclass{article}
%\documentstyle[makeidx]{book}
\makeindex
\include{rsat_latex_commands}

\begin{document}

\title{Regulatory Sequence Analysis Tools \\
Web server installation}

\author{
	Jacques van Helden \\
	\email{jvanheld@scmbb.ulb.ac.be} \\
	\url{http://www.scmbb.ulb.ac.be/$\tilde{\hspace{0.4em}}$jvanheld/} \\
	\scmb 
}


\maketitle

\newpage
\tableofcontents
\newpage

\section{Description}

This documents describes the installation procedure for the web server
of the \textbf{Regulatory Sequence Analysis Tools} (\RSAT).

It assumes that you already installed the perl scripts and the
genomes, as described in the \RSAT installation guide.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Web server installation

\section{Installing a local web server}

The Regulatory Sequence Analysis Tools include a web server, which
offers a user-friendly interface for biologists. The main server is
available for academic users at \url{http://rsat.ulb.ac.be/rsat/}. A
few additional mirrors have been installed in different countries.

\subsection{Web server pages}

The web pages are located in the directory
\file{rsa-tools/public\_html}. This directory contains both the HTML
pages, and the CGI scripts.

\subsection{Configuration of the Apache server}

In order to provide web access to the rsa-tools, you need to adapt the
configuration of your web server. This requires root privileges
(system administrator). 


The detailed settings depend on your web server program. We provide
here an example of typical settings for the Apache server (the most
widely used web server).

In summary, the configuration includes the following steps.

\begin{enumerate}
\item Open the apache configuration file \file{/etc/httpd/httpd.conf}

\item Specify an alias for rsa-tools in the appropriate section.

\begin{verbatim}
Alias /rsa-tools "/home/myaccount/rsa-tools/public_html/"
\end{verbatim}


\item Associate .cgi extension to CGI scripts

Make sure the following line is present in the config file. If the
server has not yet been configured, the line is commented, and you
need to remove the \texttt{\#} character before it.

\begin{verbatim}
AddHandler cgi-script .cgi
\end{verbatim}

\item Give authorization to execute CGI scripts in the rsa-tools directory

\begin{verbatim}
ScriptAlias /rsa-tools/ "/home/myaccount/rsa-tools/public_html/
\end{verbatim}

\item Specify the access options for the rsa-tools directory.

\begin{verbatim}
<Directory "/home/myaccount/rsa-tools/public_html">
   AllowOverride None
   Options ExecCGI Indexes
   Order allow,deny
   Allow from all
</Directory>
\end{verbatim}

\end{enumerate}

These are the basic steps to configure the web access to
\RSAT. Depending on your operating system, you probably need to
specify some additional settings. For example, on the Max OSX version
of Apache server allows to define a user-specific configuration in the
directory \file{/etc/httpd/users}.

Note that you need to restart the werb server for these changes to
take effect.

\subsection{Editing the configuration file for the \RSAT server}

If you want to install a web server, you need to edit two variables on
the file \file{RSA.config}. Open this file with a text editor, and
specify the variables \texttt{\$config\_site} and \texttt{\$WWW\_RSA}
according to your local configuration.


\subsection{testing the web server}

To test the werb server, open a web browser and connect your \RSAT server

\url{http://www.myserver/rsa-tools/} 

(of course you  need to adapt the URL according to your IP address).

If the connection works, try to execute the demonstration of the
following pages.

\begin{description}
\item[\program{retrieve-seq}] to test the correct installation of genomes. 

\item[\program{oligo-analysis}] to test the correct installation of
background oligonucleotide frequencies.

\item[\program{feature-map}] to test the correct installation of the
graphical librairies.

\end{description}

\section{Managing a local web server}

\subsection{Access logs}

Each time a script is executed via the \RSAT server, some basic
information is stored in a log file. This information is minimal: it
is restricted to the time, name of the script executed, and the IP
address of the client machine. We do not want to store any additional
information (e.g. selected organism, lists of genes), for obvious
confidentiality reasons.

The log files are saved in the directory \file{rsa-tools/logs}. There
is one file per month.


\subsection{Cleaning the temporary directory}

The web server stores result files in a temporary directory
\file{\$RSAT/public\_html/tmp/}. These files should remain 3 days on
the server, in order to allow users to consult their results.

\subsubsection{Manual cleaning}

The \RSAT package includes a make script to clean old files in the
temporary directory.

\begin{verbatim}
cd rsa-tools
make -f makefiles/server.mk clean_tmp
\end{verbatim}

This command cleans all the files older than 3 days. You can clean
more recent files by modifying the variable CLEAN\_DATE.

\begin{verbatim}
make -f makefiles/server.mk clean_tmp  CLEAN\_DATE=1
\end{verbatim}

This will clean all files older than 1 day.

\subsubsection{Automatic cleaning}

The automatic management of the temporary directory can be greatly
facilitated the \program{crontab} command. For this, you need to add a
command to your personal \texttt{crontab} configuration file.

\begin{enumerate}
\item Start to edit the crontab command file

\begin{verbatim}
crontab -e
\end{verbatim}

This will open your \file{crontab} file with your default text editor
(this default editor can be specified with the environment variable
EDITOR or VISUAL).

\item Add the following line to the \file{crontab} file. 

\begin{verbatim}
02 04 * * * make -f /home/myaccount/rsa-tools/makefiles/server.mk clean_tmp
\end{verbatim}

This will execute the make script \file{server.mk}, with the target
\texttt{clean\_tmp}, every day, at 04:02 AM. 

\item Save the modified crontab file and close your text editor.

\end{enumerate}

In principle, you will receive an email from \program{crontab} each
time the command is executed.


\end{document}
