%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%% COMPARATIVE GENOMICS
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Comparative genomics}

\subsection{Genome-wise comparison of protein sequences}

In this section, I will explain how to use the program
\program{genome-blast}.

\subsection{Getting putative homologs, orthologs and paralogs}

In this section, I will explain how to use the program
\program{get-orthologs}.

\subsection{Phylogenetic profiles}

The notion of \concept{phylogenetic profile} was introduced by
Pellegrini et al. (1999). They identified putative orthologs for all
the genes of \organism{Escherichia coli K12} in all the complete genomes
available at that time, and built a table with one row per gene, one
column per genome. Each cell of this table indicates if an ortholog of
the considered gene (row) has been identified in the considered genome
(column). Pellegrini et al. (1999) showed that genes having similar
phylogenetic profiles are generally involved in common biological
processes. The analysis of phylogenetic footprints is thus a powerful
way to identify functional grouping in completely sequenced genomes.

The program \program{get-orthologs} can be used to obtain the
phylogenetic profiles. The principle is to submit the complete list of
protein-coding genes of the query organism. We process in two steps : 

\begin{enumerate}

\item With \program{get-orthologs}, we can identify the putative
  ortholgos for all the genes of the query organism, using the
  criterion of \concept{bidirectional best hit (BBH)}. This generate a
  large table with one row per pair of putative orthologs.

\item We then use \program{convert-classes} to convert the ortholog
  table into profiles (one row per gene, one column per genome).

\end{enumerate}

We will illustrate this by calculating the phylogenetic profiles of
all the genes from \organism{Saccharomyces cerevisiae} across all the
Fungi. We use a level of verbosity of 2, in order to get information
about the progress of the calculations.

It is well known that the sole criterion of BBH presents problem for
the identification of putative orthologs. In particular, there is a
riks to obtain irrelevant matches, due to partial matches between a
protein and some spurious domains. To avoid this, we will add a
constraint on the percentage of identity (min 30\%), and on the
alignment length (min 50 aa). These limits are somewhat arbitrary, we
use them to illustrate the principe, and leave to each user the
responsibility to choose the criteria that she/he considers as
relevant. Finally, we will use a more stringent threshold on E-value
than the default one.

\begin{small}
\begin{verbatim}

## Identify all the putative orthologs (BBH)
get-orthologs -v 2 \
  -i $RSAT/data/genomes/Saccharomyces_cerevisiae/genome/cds.tab  \
  -org Saccharomyces_cerevisiae \
  -taxon Fungi \
  -uth rank 1 -lth ali_len 50 -lth ident 30 -uth e_value 1e-10 \
  -return e_value,bit_sc,ident,ali_len \
  -o Saccharomyces_cerevisiae_Fungi_bbh.tab

## Convert ortholog table into a profile table
## with the IDs of the putative orthologs
convert-classes -v 2 \
  -i Saccharomyces_cerevisiae_Fungi_bbh.tab  \
  -from tab -to profiles \
  -ccol 2 -mcol 3 -scol 1 -null "<NA>" \
  -o Saccharomyces_cerevisiae_Fungi_phyloprofiles_ids.tab

\end{verbatim}
\end{small}


The resulting table indicates the identifier of the ortholog
genes. The option \option{-null} was used to specify that the string
\texttt{<NA>} should be used to indicate the absence of putative
orhtolog.

Another option would be to obtain a ``quantitative'' profile, where
each cell indicates the E-value of the match between the two
orthologs. This can be done by specifying a different score column
with the option \option{-scol} of \program{convert-classes}.

\begin{small}
\begin{verbatim}

## Convert ortholog table into a profile table
## with the E-value of the putative orthologs
convert-classes -v 2 \
  -i Saccharomyces_cerevisiae_Fungi_bbh.tab  \
  -from tab -to profiles \
  -ccol 2 -mcol 3 -scol 4 -null "<NA>" \
  -o Saccharomyces_cerevisiae_Fungi_phyloprofiles_evalue.tab

\end{verbatim}
\end{small}

