%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%% Installing organisms
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\chapter{Installing organisms}


\RSAT includes a series of tools to install and maintain the latest
version of genomes.

\section{Original data sources}

Genomes supported on \RSAT were obtained from various sources.

Genomes can be installed either from the \RSAT web site, or from their
original sources.  

\begin{itemize}
\item NCBI/Genbank (\url{ftp://ftp.ncbi.nih.gov/genomes/})

\item ENSEMBL (\url{http://www.ensembl.org/})

\item The EBI genome directory (\url{ftp://ftp.ebi.ac.uk/pub/databases/genomes/Eukaryota/})

\end{itemize}

Other genomes can also be found on the web site of a diversity of
genome-sequencing centers.

\section{Requirement : wget}

The download of genomes relies on the application \program{wget},
which is part of linux distribution. \program{wget} is a ``web
aspirator'', which allows to downlaod whole directories from ftp and
http sites. You can check if the program is installed on your machine.

\begin{verbatim}
wget -help
\end{verbatim}

This command should return the help pages for \program{wget}.  If you
obtain an error message (``command not found''), you need to ask your
system administrator to install it.

\section{Importing organisms from the \RSAT main server}

The simplest way to install organisms on our \RSAT site is to download
the RSAT-formatted files from the web server. For this, you can use a
web aspirator (for example the program \program{wget}). 

Beware, the full installation (including Mammals) requires a large
disk space (several dozens of Gb). You should better start installting
a small genome and test it before processing to the full
installation. We illustrate the approach with one of the smallest
sequenced genome: \textit{Mycoplasma genitalium}.

To download the genome in your \RSAT folder, tye the following
command.

\begin{small}
\begin{verbatim}
cd $RSAT
wget -rNL http://rsat.scmbb.ulb.ac.be/rsat/data/genomes/Mycoplasma_genitalium/
\end{verbatim}
\end{small}

This will create a local mirror of the \RSAT data repository. You can
check the result by typing.

\begin{small}
\begin{verbatim}
ls -l $RSAT/rsat.scmbb.ulb.ac.be/rsat/data/genomes/Mycoplasma_genitalium/
\end{verbatim}
\end{small}

When the download is complete, move the newly transferred genome to
the data directory of your \RSAT installation.

\begin{small}
\begin{verbatim}
mv $RSAT/rsat.scmbb.ulb.ac.be/rsat/data/genomes/Mycoplasma_genitalium \
    $RSAT/data/genomes/
\end{verbatim}
\end{small}

You need now to declare the newly installed organism. 

\begin{small}
\begin{verbatim}
install-organism -v 1 -task config \
    -org Mycoplasma_genitalium -up_from -400 -up_to -1
\end{verbatim}
\end{small}

You can now check the configuration file.

\begin{small}
\begin{verbatim}
tail -20 $RSAT/data/supported_organisms.pl
\end{verbatim}
\end{small}

If the installation was successfull, you should see something like this : 

\begin{tiny}
\begin{verbatim}
#### Mycoplasma_genitalium      Mycoplasma genitalium   2006/01/04 22:08:42
$supported_organism{'Mycoplasma_genitalium'}->{'name'} = "Mycoplasma genitalium";
$supported_organism{'Mycoplasma_genitalium'}->{'data'} = "$RSA/data/genomes/Mycoplasma_genitalium";
$supported_organism{'Mycoplasma_genitalium'}->{'last_update'} = "2006/01/04 22:08:42";
$supported_organism{'Mycoplasma_genitalium'}->{'features'} = "$RSA/data/genomes/Mycoplasma_genitalium/genome/feature.tab";
$supported_organism{'Mycoplasma_genitalium'}->{'genome'} = "$RSA/data/genomes/Mycoplasma_genitalium/genome/contigs.txt";
$supported_organism{'Mycoplasma_genitalium'}->{'seq_format'} = "filelist";
$supported_organism{'Mycoplasma_genitalium'}->{'taxonomy'} = "Bacteria; Firmicutes; Mollicutes; Mycoplasmataceae; Mycoplasma";
$supported_organism{'Mycoplasma_genitalium'}->{'synonyms'} = "$RSA/data/genomes/Mycoplasma_genitalium/genome/feature_names.tab";
$supported_organism{'Mycoplasma_genitalium'}->{'up_from'} = -400;
$supported_organism{'Mycoplasma_genitalium'}->{'up_to'} = -1;

return 1;
\end{verbatim}
\end{tiny}


\section{Installing genomes from  their original source}

The parsing of genomes from their original data
sources is more tricky than the synchronization from the \RSAT server.

In principle, if you succeeded, with the protocol above, to obtain the
genomes from the \RSAT server, you don't need to proceed to new
installations yourself and you can skip the rest of this chapter.

\subsection{The \RSAT genome files}


\begin{enumerate}
\item genome sequence
\item feature table
\item list of names/synonyms
\end{enumerate}

\subsubsection{Genome sequence} 

The genome must be in raw format (text files containing the sequence
without any space or carriage return). If the organism contains
several chromosomes, there should be one separate file per contig
(chromosome). 

In addition, the genome directory must contain one file listing the
contig (chromosome) files. You can find an example in the directory
\file{\$RSAT/data/genomes/Saccharomyces\_cerevisiae/genome/}.


\subsubsection{Feature table}

A feature-table giving the basic information about genes. This is
a tab-delimited text file. Each row contains information about one
gene. The columns contain the following information: 
\begin{enumerate}

\item Identifier

\item Feature type (e.g. ORF, tRNA, ...)

\item Name

\item Chromosome. This must correspond to one of the sequence
identifiers from the fasta file.

\item Left limit

\item Right limit

\item Strand (D for direct, R for reverse complemet)

\item Description. A one-sentence description of the gene function.

\end{enumerate}

\subsubsection{Gene names (synonyms)}

Optionally, you can provide a synonym file, which contains two
columns:

\begin{enumerate}
\item ID. This must be one identifier found in the feature table
\item Synonym
\end{enumerate}

Multiple synonyms can be given for a gene, by adding several lines with
the same ID in the first column.

\subsubsection{Example}

\begin{verbatim}
cd $RSAT/data/genomes/Saccharomyces_cerevisiae/genome/

## The list of sequence files
cat contigs.txt

## The sequence files
ls -l *.raw

## The feature table
head -30 feature.tab

## The gene names/synonyms
head -30 feature_names.tab

\end{verbatim}


\section{Parsing genomes from NCBI/Genbank}

The easiest way to install an organism in \RSAT is to download the
complete genome files from the NCBI
\urlref{ftp://ftp.ncbi.nih.gov/genomes/}, and to parse it with the
program \program{parse-genbank.pl}.

\subsection{Downloading genomes from NCBI/Genbank}

\RSAT includes a makefile to download genomes from different sources.
We provide hereafter a protocol to create a download directory in your
account, and download genomes in this directory. Beware, genomes
require a lot of disk space, especially for those of higher
organisms. To avoid filling up your hard drive, we illustrate the protocol
with the smallest procaryote genome to date: \textit{Mycoplasma
  genitamlium}.


\begin{verbatim}
## Creating a directory for downloading genomes in your home account
cd $HOME
mkdir -p downloads
cd downloads

## Creating a link to the makefile which allows you to dowload genomes
ln -s $RSAT/makefiles/downloads.mk ./makefile
\end{verbatim}

We will now download a small genome from NCBI/Genbank. 

\begin{verbatim}
## Downloading one directory from NCBI Genbank
cd $HOME/downloads/
make one_genbank_dir GB_DIR=genomes/Bacteria/Mycoplasma_genitalium
\end{verbatim}


\subsection{Parsing genomes from NCBI/Genbank}

The program \program{parse-genbank.pl} extract genome information
(sequence, gene location, ...) from Genbank flat files, and exports
the result in a set of tab-delimited files.

\begin{verbatim}



\end{verbatim}


\section{Parsing genomes from EMBL files}

The program \program{parse-embl.pl} reads flat files in EMBL format,
and exports genome sequences and features (CDS, tRNA, ...) in
different files.

As an example, we can parse a yeast genome sequenced by the
``Genolevures'' project
\urlref{http://natchaug.labri.u-bordeaux.fr/Genolevures/download.php}.

Let us assume that you want to parse the genome of the species
\textit{Debaryomyces hansenii}.

Before parsing, you need to download the files in your account, 

\begin{itemize}
\item Create a directory for storing the EMBL files. The last level of
  the directory should be the name of the organism, where spaces are
  replaced by underscores. Let us assume that you store them in
  the directory \file{\$HOME/downloads/Debaryomyces\_hansenii}.

\item Download all the EMBL file for the selected organism. Save each
  name under its original name (the contig ID), followed by the
  extension \texttt{.embl})

\end{itemize}

We will check the content of this directory.

\begin{verbatim}
ls -1 $HOME/downloads/Debaryomyces_hansenii
\end{verbatim}

On my computer, it gives the following result

\begin{verbatim}
CR382133.embl
CR382134.embl
CR382135.embl
CR382136.embl
CR382137.embl
CR382138.embl
CR382139.embl
\end{verbatim}

The following instruction will parse this genome.

\begin{verbatim}
parse-embl.pl -v 1 -i  $HOME/downloads/Debaryomyces_hansenii
\end{verbatim}

If you do not specify the output directory, a directory is
automatically created by combining the current date and the organism
name.  The verbose messages will indicate you the path of this
directory, something like
\file{\$HOME/parsed\_data/embl/20050309/Debaryomyces\_hanseni}.


\subsection{Installing a genome in the main \RSAT directory}

Once the genome has been parsed, the simplest way to make it available
 for all the users is to install it in the \RSAT genome directory. You
 can already check the genomes installed in this directory.

\begin{verbatim}
ls -1 $RSAT/data/genomes/
\end{verbatim}

There is one subdirectory per organism. For example, the yeast data is
 in \file{\$RSAT/data/genomes/Saccharomyces\_cerevisiae/}. This
 directory is further subdivided in folders: \file{genome} and
 \file{oligo-frequencies}.

We will now create a directory to store data about
 Debaryomyces\_hansenii, and transfer the newly parsed genome in this
 directory.

\begin{verbatim}
## Create the directory
mkdir -p $RSAT/data/genomes/Debaryomyces_hansenii/genome

## Transfer the data in this directory
mv $HOME/parsed_data/embl/20050309/Debaryomyces_hanseni/* \
  $RSAT/data/genomes/Debaryomyces_hansenii/genome

## Check the transfer
ls -ltr $RSAT/data/genomes/Debaryomyces_hansenii/genome
\end{verbatim}

\subsection{Updating the configuration file}

The fact to add a directory is not sufficient for \RSAT to be aware of
the new organism. For this, we must update the configuration file. We
will also specify the default upstream sequence length. For a yeast
(\textit{Debaryomyces hansenii}), a good guess is 800bp (this is at
least the value I chose for \textit{Saccharomyces cerevisiae}).

\begin{verbatim}
install-organism -v 1 -org Debaryomyces_hansenii -task config -up_from -800

## Check the last lines of the configuration file
tail -15 $RSAT/data/supported_organisms.pl
\end{verbatim}

\section{Checking that the organism is installed properly}

To check the installation, start by checking whether your newly
installed now appears in the list of supported organisms.

\begin{verbatim}
supported-organisms
\end{verbatim}

Will give you a list of installed organisms.


\subsection{Retrieving sequences}

As soon as the configuration file has been updated, we should be in
state to retrieve sequences for the newly installed genome. We will
check this by retrieving a the start codons.

\begin{verbatim}
retrieve-seq -org Debaryomyces_hansenii -all -from 0 -to 2
\end{verbatim}

\subsection{Checking the composition of start codons}

Once the organism is found in your configuration, you need to check
whether sequences are retrieved properly. A good test for this is to
retrieve all the start codons, and check whether they are made of the
expected codons (mainly ATG, plus some alternative start codons like
GTG or TTG for bacteria).

We will now analyze the trinucleotide composition of the start
codons. In principle, al of them should be ATG for an eucaryote
organism.

\begin{verbatim}
retrieve-seq -org Debaryomyces_hansenii -all -from 0 -to 2 \
    | oligo-analysis -l 3 -1str -v 1 -return occ,freq -sort
\end{verbatim}

\subsection{Checking the start and stop codons with \program{install-organisms}}

The program \program{install-organisms} includes an option to
automatically check all the start and stop codons from a parsed
organism.

\begin{verbatim}
install-organism -v 1 -org Debaryomyces_hansenii -task start_stop
\end{verbatim}

You can then check the composition of the start and stop codons.

\begin{verbatim}
cd $RSAT/data/genomes/Debaryomyces_hansenii/genome/
more Debaryomyces_hansenii_start_codon_frequencies
more Debaryomyces_hansenii_stop_codon_frequencies
\end{verbatim}

The stop codons should be TAA, TAG or TGA, for any organism. For
eucaryotes, all start codons should be ATG. For some procaryotes,
alternative start codons (GTG, TGG) are frequent.

\subsection{Calibrating oligonucleotide and dyad frequencies with \program{install-organisms}}

The programs \program{oligo-analysis} and \program{dyad-analysis}
  require calibrated frequencies for the background models. These
  frequencies are calculated automatically with
  \program{install-organism}.

\textbf{Warning: } this task requires several hours of computation (a
few hours for small bacterial genomes, and several days for the human
genome).

\begin{verbatim}
install-organism -v 1 -org Debaryomyces_hansenii \
    -task allup,oligos,dyads,upstream_freq,clean
\end{verbatim}

\subsection{Installing a genome in your own account}

We describe below how this information should be formatted to be used
in rsa-tools.

In this chapter, we explain how to add support for an organism on your
local configuration of \RSAT. This assumes that you have the complete
sequence of a genome, and a table describing the predicted location of
genes.

First, prepare a directory where you will store the data for your
organism. For example:

\begin{verbatim}
mkdir -p $HOME/rsat-add/data/Mygenus_myspecies/
\end{verbatim}


One you have this information, start the program
\program{install-organism}. You will be asked to enter the information
needed for genome installation.

\section{Updating your local configuration}


\begin{itemize}
\item Modify the local config file

\item You need to define an environment variable called
  RSA\_LOCAL\_CONFIG, containing the full path of the local config
  file.

\end{itemize}
