<HTML>
<HEAD>
   <TITLE>RSA-tools - Change summary</TITLE>
</HEAD>
<BODY TEXT="#000000" BGCOLOR="#FFEEDD" LINK="#0000EE" VLINK="#55008B" ALINK="#FF0000">

<CENTER><H2><A HREF="RSAT_home.cgi">RSA-tools</A> - 
Change summary
</H2></CENTER>

<P>
The Regulatory Sequence Analysis tools are evolving regularly. We
summarize here the important steps in this evolution.  This summary
only started in September 2002 (although the tools exist since
1997). In addition, it only mention important changes, and I forget
some ones.  An exhaustive list of program modifications (including
those anterior to September 2002) can be obtained on request to
Jacques van Helden
(<a
href=mailto:jvanheld@scmbb.ulb.ac.be>jvanheld@scmbb.ulb.ac.be</a>).

<p>

<table border=1  width=650>

<tr valign=top>
<th>Date</th>
<th>Program</th>
<th>Description</th>
</tr>

<!--
<tr valign=top>
<td>2006</td>
<td><a href=></a></td>
<td></td>
</tr>
-->

<tr valign=top>
<td>2007/05/12</td>
<td>All the programs</td>
<td>Output tables can now be re-sorted on any criterion by cliking on the column headers</td>
</tr>

<tr valign=top>
<td>2006/12/07</td>
<td>
<a href=oligo-analysis_form.cgi>oligo-analysis</a>
<br><a href=dyad-analysis_form.cgi>dyad-analysis</a>
</td>
<td>Added the possibility to send the result (PSSM) to matrix-scan.</td>
</tr>


<tr valign=top>
<td>2006/12/07</td>
<td>
<a href=oligo-analysis_form.cgi>oligo-analysis</a>
<br><a href=dyad-analysis_form.cgi>dyad-analysis</a>
</td>
<td>Included a conversion from pattern assembly fo PSSM via convert-matrix.</td>
</tr>


<tr valign=top>
<td>2006/11</td>
<td><a href=matrix-scan_form.cgi>matrix-scan</a></td>
<td>Added a web interface to the program <i>matrix-scan</i></td>
</tr>

<tr valign=top>
<td>2006/04/12</td>
<td><a href=dyad-analysis_form.cgi>dyad-analysis</a></td>
<td>Corrected imprecisions with the calculation of expected frequency (with the monad background)
<br>Cross-cecked the results by comparing dyads with spacing 0 to oligo-analysis results.
<br>Improved the output (more consistent with oligo-analysis)
</td>
</tr>

<tr valign=top>
<td>2006/02</td>
<td>All the scripts reading sequences</td>
<td>Added support for tab-delimited sequence files. Each row corresponds to one sequence. 
<ol>
<li>The first column contains the sequence ID
<li>the second column the sequence. 
<li>All subsequent columns are ignored.
</ol></td>
</tr>

<tr valign=top>
<td>2006/02/08</td>
<td>Pattern matching and pattern discovery algorithms<a href=></a></td>
<td>Added an option "mask", which allows to mask either lowercases or uppercases (transform soft masking into hard masking)</td>
</tr>

<tr valign=top>
<td>2006/01/23</td>
<td><a href=retrieve-seq_form.cgi>retrieve-seq</a></td>
<td>Added an option to retrieve sequences for multiple organisms (coupled with get-orthologs)</td>
</tr>

<tr valign=top>
<td>2006/01/19</td>
<td><a href=get-orthologs_form.cgi>get-orthologs</a></td>
<td>Created a form for get-orthologs</td>
</tr>

<tr valign=top>
<td>2006/01/01</td>
<td><a href=retrieve-seq_form.cgi>retrieve-seq</a></td>
<td>Fixed a bug with gene names : genes without names had "NULL" as
name, instead of the identifier. From now on, the locus_tag is
systematically used as identifier, and, when a gene has no name, its
ID is used as primary name.</td>
</tr>


<tr valign=top>
<td>2005/12/14</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a></td>
<td>Added the option to select over-represented, under-represented or both types of oligonucleotides</td>
</tr>


<tr valign=top>
<td>2005/11/18</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a> and <a href=dyad-analysis_form.cgi>dyad-analysis</a></td>
<td>Changed the correction for multi-testing. Bfore, the E-value was calculated as 

<i><pre>
  E-value = P-value * nb_possible_patterns
</pre></i>

The E-value is now calculated according to the number of patterns
really tested for significance (i.e. after filtering on occurrences,
frequencies, ...).

<i><pre>
  E-value = P-value * nb_tested_patterns
</pre></i>


<p>In most cases it should not make any difference with previous
results, but for small sequence sets (< 10000bp) or large patterns (>
6nt), the previous correction was too severe, resulting in a loss of
sensitivity.

</td>
</tr>


<tr valign=top>
<td>2005/08/31</td>
<td><a href=dyad-analysis_form.cgi>dyad-analysis</a></td>
<td>Added an option to select fields to be returned, and specify an upper and a lower thresholds on any of these fields</td>
</tr>


<tr valign=top>
<td>2005/07/21</td>
<td><a href=genome-scale-dna-pattern_form.cgi>genome-scale-dna-pattern</a></td>
<td>Added the option Feature type (which was already present in retrieve sequences)</td>
</tr>

<tr valign=top>
<td>2005/07/21</td>
<td><a href=genome-scale-patser_form.cgi>genome-scale-patser</a></td>
<td>Added the option Feature type (which was already present in retrieve sequences)</td>
</tr>

<tr valign=top>
<td>2005/06/12</td>
<td><a href=dyad-analysis_form.cgi>dyad-analysis</a></td>
<td>Fixed a bug with the calculation of relative frequencies,expected
frequencies, and probabilities.</td>
</tr>

<tr valign=top>
<td>2005/06/12</td>
<td><a href=dyad-analysis_form.cgi>dyad-analysis</a></td>
<td><li>Changed the counting mode when counting on boths
strands. Instead of considering the 2 strands as two independent
sequences with T=2*(L-k+1) possible positions for D dyads, I consider
pairs of reverse-complementary dyads as the single event, and the
number of possible positions for any of the pairs (direct or RC dyad)
is T=L-k+1. 
<p>Occurrence counts remain unchanged, except for the
revere-palindromic dyads, which appear twice less frequent (they are
noww "seen" on a single strand.
<p>Probabilities were adapted accordingly. For non-reverse-plindromic
dyads, this can slightly change the calucated P-values. for reverse
complementary palindromic dyads, it does not change anything , since
the P-value were already calculated this way.
</td>
</tr>

<tr valign=top>
<td>2005/06/12</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a></td>
<td>Fixed a bug with the Markov chains (expected frequencies summed to 4 instead of 1)</td>
</tr>

<tr valign=top>
<td>2005/05/11</td>
<td><a href=random-seq_form.cgi>random-seq</a></td>
<td>Markov models are calibrated on non-coding upstream sequences (upstream-noorf) rather than intergenic.</td>
</tr>


<tr valign=top>
<td>2005/04/09</td>
<td><a href=retrieve-seq_form.cgi>retrieve-seq</a></td>
<td>Fixed a bug with the determination of neighbour ORFS. This bug
resulted in incorrect clipping with the option "no overlapping
ORF". In particular, it waas incompatible with downstream sequecnes
(all sizes=0).</td>
</tr>


<tr valign=top>
<td>2004/09/18</td>
<td><a href=retrieve-seq_form.cgi>retrieve-seq</a></td>
<td>
<ul>
<li>Homonyms are now supported (i.e. several genes with the same name)
<li>fixed a bug by treating the cases of overlapping upstream genes
(e.g. in bacterial operons) : in this case, the sequence has a length
of 0
</ul>
</td>
</tr>

<tr valign=top>
<td>2004/06/09</td>
<td><a href=convert-matrix_form.cgi>convert-matrix</a></td>
<td>Implemented a web interface for a new script, convert-matrix (in construction)</td>
</tr>

<tr valign=top>
<td>2004/06/03</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a></td>
<td>Added the possibility to specify an upper and a lower threshold on each output parameter</td>
</tr>

<tr valign=top>
<td>2004/05/12</td>
<td>
<a href=retrieve-seq_form.cgi>retrieve-seq</a>
<br><a href=genome-scale-dna-pattern_form.cgi>genome-scale dna-pattern</a>
<br><a href=genome-scale-patser_form.cgi>genome-scale patser</a>
</td>
<td>Fixed a bug (introduced May 7) which caused each sequence to be
duplicated with the option "all genes"</td>
</tr>

<tr valign=top>
<td>2004/05/08</td>
<td><a href=tut_random-genes.html>random-genes tutorial</a></td>
<td>Wrote a tutorial for random-genes</td>
</tr>

<tr valign=top>
<td>2004/05/07</td>
<td><a href=retrieve-seq_form.cgi>retrieve-seq</a>
<br><a href=random-genes_form.cgi>random-genes</a>
<br><a href=gene-info_form.cgi>gene-info</a></td>
<td>Additional feature types are now supported: tRNA, rRNA, scRNA</td>
</tr>

<tr valign=top>
<td>2004/05/07</td>
<td><a href=patser_form.cgi>patser</a></td>
<td>The default parameter for lower threshold estimation has been set
to weight score, because the adjusted information is often too
stringent (no answer)</td>
</tr>

<tr valign=top>
<td>2004/04/02</td>
<td><a href=feature-map_form.cgi>feature-map</a></td>
<td>The backbone (black line) of each sequence now reflects the
sequence length. This is useful to show the different sequence lengths
when the upstream ORFs have been clipped with retrieve-seq.</td>
</tr>

<tr valign=top>
<td>2004/04/02</td>
<td><a href=dna-pattern_form.cgi>dna-pattern</a></td>
<td>dna pattern now exports the start and end position of each
sequence (this information is used to draw the the feature-map)</td>
</tr>

<tr valign=top>
<td>2004/03/27</td>
<td><a href=feature-map_form.cgi>feature-map</a></td>
<td>Feature-maps can now be exported in postscript (ps) to obtain a
high quality printing. The form contains an option to select image
format. </td>
</tr>


<tr valign=top>
<td>2004/01/13</td>
<td><a href=random-genes_form.cgi>random-genes</a></td>
<td>Added an option to return multiple group of random genes</td>
</tr>


<tr valign=top>
<td>2003/12/30</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a></td>
<td>Added an option to return pattern count distributions. This
returnss one row per patttern and each row contains a frequency
distribution for the pattern occurrences.</td>
</tr>


<tr valign=top>
<td>2003/10/20</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a></td>
<td>Fixed a bug in the calculation of expected frequencies when oligos
were counted on both strands with alphabet, Markov and lexicon : the
expected frequency did not sum the pairs of reverse complements, which
resulted in an under-estimation of expected frequencies, and thus in
an over-estimation of the significance.
</td>
</tr>



<tr valign=top>
<td>2003/08/29</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a></td>
<td>Added the option 'One row per sequence', which returns a table
with one row per gene, one column per pattern (occurrence counts
only)</td>
</tr>

<tr valign=top>
<td>2003/07/10</td>
<td><a href=retrieve-seq_form.cgi>retrieve-seq</a></td>
<td>Added the option "server" for the output type. This avoids to
display the sequence, whilst allowing to send it to other tools</td>
</tr>

<tr valign=top>
<td>2003/07/07</td>
<td>Genomes</td>
<td>First installation of Rattus norvegicus</td>
</tr>

<tr valign=top>
<td>2003/07/07</td>
<td><a href=dyad-analysis_form.cgi>dyad-analysis</a></td>
<td>For the pattern assembly, changed the allowed flanking bases from
2 to 1, to avoid assembling too many unrelated patterns</td>
</tr>


<tr valign=top>
<td>2003/07/06</td>
<td><a href=gene-info_form.cgi>gene-info</a></td>

<td>The program orf-info has been renamed gene-info. The reason is
that mRNAs are now supported. An option allows to chose the feature
type between CDSand mRNA.</td>

</tr>


<tr valign=top>
<td>2003/06/06</td>
<td><a href=retrieve-seq_form.cgi>retrieve-seq</a></td>

<td>Added an option "feature-type" allowing to choose between CDSs and
mRNAs.</td>

</tr>


<tr valign=top>
<td>2003/06/02</td>
<td>
<a href=patser_form.cgi>patser</a><br>
<a href=genome-scale-patser_form.cgi>genome-scale-patser</a><br>
</td>
<td>Added the option to return a score table (one row per sequence, one column per match)</td>
</tr>

<tr valign=top>
<td>2003/05/14</td>
<td><a href=tutorials/tutorials.html>tutorials</a></td>
<td>
Added several pages to the tutorials.
<ul>
<li>String-based representations
<li>Matrix-based representations
<li>patser
<li>TRANSFAC
<li>Generating random sequences
<li>Selecting random genes
</ul>
</td>
</tr>

<tr valign=top>
<td>2003/05/06</td>
<td><a href=position-analysis_form.cgi>position-analysis</a></td>
<td>Implemented the web interface for position-analysis.</td>
</tr>

<tr valign=top>
<td>2003/04/28</td>
<td>parsers</td>
<td>The genbank parser has been improved to take into account more
synonyms (by parsing the field 'note=', and return more details about
the organism (taxonomy) and the contigs (date of the genbank
deposition).</td>
</tr>



<tr valign=top>
<td>2003/04/28</td>
<td><a href=supported-organisms.cgi>supported organisms</a></td>
<td>New script which returns information (taxonomy, parsing date,
default upstream boundaries, ...) about supported organisms</td>
</tr>

<tr valign=top>
<td>2002/11/16</td>
<td><a href=dyad-analysis_form.cgi>dyad-analysis</a></td>
<td>Support different bacground models : intergenic (as previously), upstream, upstream-noorf</td>
</tr>

<tr valign=top>
<td>2002/10/22</td>
<td><a href=pattern-assembly_form.cgi>pattern-assembly</a></td>
<td>Added a form for pattern-assembly (previously, it was only available through the result of dyad-assembly or oligo-assembly)</td>
</tr>

<tr valign=top>
<td>2002/09/16</td>
<td><a href=tutorials/tut_consensus.html>tutorial on consensus</a></td>
<td>Wrote a tutorial for consensus, Jerry Hert'z pattern discovery
program.</td>
</tr>

<tr valign=top>
<td>2002/09/16</td>
<td><a href=gene-info_form.cgi>gene-info</a></td>
<td>Added an option to match queries against gene description
(previously, queries were only matched against gene names and
identifiers).</td>
</tr>


<tr valign=top>
<td>2002/09/15</td>
<td><a href=tutorials/tut_oligo-analysis.html>oligo-analysis tutorial</a></td>
<td>improved the tutorial in several ways
<ol>
<li>changed the example to the MET family
<li>added an explanation about result interpretation
<li>use the min feature thickness to highlight the Met3p binding site in the feature map. 
</ol>
</td>
</tr>


<tr valign=top>
<td>2002/09/15</td>
<td><a href=tutorials/tutorials.html>tutorials</a></td>
<td>Wrote a <a href=tutorials/tut_genome-scale-patser.html>tutorial for genome-scale patser</a></td>
</tr>


<tr valign=top>
<td>2002/09/13</td>
<td><a href=retrieve-seq_form.cgi>retrieve-seq</a></td>
<td>The default is now to prevent overlap with upstream ORFs. This is
essential for bacteria, to avoid taking large fragments of coding
sequences, when the gene is in the middle of an operon. Actually, for
bacteria, the distance to the upstream ORF is < 50bp for more than
half of the genes.</td>
</tr>

<tr valign=top>
<td>2002/09/14</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a></td>
<td>different background models are supported: intergenic, all
upstream sequences (with or without overlap with upstream ORFs).
</td>
</tr>


<tr valign=top>
<td>2002/09/12</td>
<td><a href=random-genes_form.cgi>random-genes</a></td>
<td>A new program, allowing to select random genes in a given
genome. The selected genes can be piped to retrieve-seq. This program
is important for testing the rate of false positive returned by the
different pattern discovery programs.</td>
</tr>

<tr valign=top>
<td>2002/09/12</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a> and <a
href=dyad-analysis_form.cgi>dyad-analysis</a></td>
<td>The programs now return an E-value, in addition to the P-value and
sig index.</td>
</tr>


<tr valign=top>
<td>2002/09/12</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a></td>
<td>For double strand counts of occurrences, I changed the way to
calculate the binomial probability. See the <a
href=help.oligo-analysis.html#proba>manual</a> for the
detail.</td>
</tr>

</table>


</BODY>
</HTML>





