<HTML>
<HEAD>
   <TITLE>RSA-tools - Change summary</TITLE>
</HEAD>
<BODY TEXT="#000000" BGCOLOR="#FFEEDD" LINK="#0000EE" VLINK="#55008B" ALINK="#FF0000">

<CENTER><H2><A HREF="RSA_home.cgi">RSA-tools</A> - 
Change summary
</H2></CENTER>

<P>
The Regulatory Sequence Analysis tools are evolving regularly. We
summarize here the important steps in this evolution.  This summary
only started in September 2002 (although the tools exist since 1997).

<P>
A detailed list of program modifications (including those anterior to
September 2002) can be obtained on request to Jacques van Helden (<a
href=mailto:jvanheld@ucmb.ulb.ac.be>jvanheld@ucmb.ulb.ac.be</a>).

<p>

<table border=1>

<tr valign=top>
<th>Date</th>
<th>Program</th>
<th>Description</th>
</tr>


<tr valign=top>
<td></td>
<td></td>
<td></td>
</tr>

<tr valign=top>
<td>2004/04/02</td>
<td><a href=feature-map_form.cgi>feature-map</a></td>
<td>The backbone (black line) of each sequence now reflects the
sequence length. This is useful to show the different sequence lengths
when the upstream ORFs have been clipped with retrieve-seq.</td>
</tr>

<tr valign=top>
<td>2004/04/02</td>
<td><a href=dna-pattern_form.cgi>dna-pattern</a></td>
<td>dna pattern now exports the start and end position of each
sequence (this information is used to draw the the feature-map)</td>
</tr>

<tr valign=top>
<td>2004/03/27</td>
<td><a href=feature-map_form.cgi>feature-map</a></td>
<td>Feature-maps can now be exported in postscript (ps) to obtain a
high quality printing. The form contains an option to select image
format. </td>
</tr>


<tr valign=top>
<td>2004/01/13</td>
<td><a href=random-genes_form.cgi>random-genes</a></td>
<td>Added an option to return multiple group of random genes</td>
</tr>


<tr valign=top>
<td>2003/12/30</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a></td>
<td>Added an option to return pattern count distributions. This
returnss one row per patttern and each row contains a frequency
distribution for the pattern occurrences.</td>
</tr>


<tr valign=top>
<td>2003/10/20</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a></td>
<td>Fixed a bug in the calculation of expected frequencies when oligos
were counted on both strands with alphabet, Markov and lexicon : the
expected frequency did not sum the pairs of reverse complements, which
resulted in an under-estimation of expected frequencies, and thus in
an over-estimation of the significance.
</td>
</tr>



<tr valign=top>
<td>2003/08/29</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a></td>
<td>Added the option 'One row per sequence', which returns a table
with one row per gene, one column per pattern (occurrence counts
only)</td>
</tr>

<tr valign=top>
<td>2003/07/10</td>
<td><a href=retrieve-seq_form.cgi>retrieve-seq</a></td>
<td>Added the option "server" for the output type. This avoids to
display the sequence, whilst allowing to send it to other tools</td>
</tr>

<tr valign=top>
<td>2003/07/07</td>
<td>Genomes</td>
<td>First installation of Rattus norvegicus</td>
</tr>

<tr valign=top>
<td>2003/07/07</td>
<td><a href=dyad-analysis_form.cgi>dyad-analysis</a></td>
<td>For the pattern assembly, changed the allowed flanking bases from
2 to 1, to avoid assembling too many unrelated patterns</td>
</tr>


<tr valign=top>
<td>2003/07/06</td>
<td><a href=gene-info_form.cgi>gene-info</a></td>

<td>The program orf-info has been renamed gene-info. The reason is
that mRNAs are now supported. An option allows to chose the feature
type between CDSand mRNA.</td>

</tr>


<tr valign=top>
<td>2003/06/06</td>
<td><a href=retrieve-seq_form.cgi>retrieve-seq</a></td>

<td>Added an option "feature-type" allowing to choose between CDSs and
mRNAs.</td>

</tr>


<tr valign=top>
<td>2003/06/02</td>
<td>
<a href=patser_form.cgi>patser</a><br>
<a href=genome-scale-patser_form.cgi>genome-scale-patser</a><br>
</td>
<td>Added the option to return a score table (one row per sequence, one column per match)</td>
</tr>

<tr valign=top>
<td>2003/05/14</td>
<td><a href=tutorials/tutorials.html>tutorials</a></td>
<td>
Added several pages to the tutorials.
<ul>
<li>String-based representations
<li>Matrix-based representations
<li>patser
<li>TRANSFAC
<li>Generating random sequences
<li>Selecting random genes
</ul>
</td>
</tr>

<tr valign=top>
<td>2003/05/06</td>
<td><a href=position-analysis_form.cgi>position-analysis</a></td>
<td>Implemented the web interface for position-analysis.</td>
</tr>

<tr valign=top>
<td>2003/04/28</td>
<td>parsers</td>
<td>The genbank parser has been improved to take into account more
synonyms (by parsing the field 'note=', and return more details about
the organism (taxonomy) and the contigs (date of the genbank
deposition).</td>
</tr>



<tr valign=top>
<td>2003/04/28</td>
<td><a href=supported-organisms.cgi>supported organisms</a></td>
<td>New script which returns information (taxonomy, parsing date,
default upstream boundaries, ...) about supported organisms</td>
</tr>

<tr valign=top>
<td>2002/11/16</td>
<td><a href=dyad-analysis_form.cgi>dyad-analysis</a></td>
<td>Support different bacground models : intergenic (as previously), upstream, upstream-noorf</td>
</tr>

<tr valign=top>
<td>2002/10/22</td>
<td><a href=pattern-assembly_form.cgi>pattern-assembly</a></td>
<td>Added a form for pattern-assembly (previously, it was only available through the result of dyad-assembly or oligo-assembly)</td>
</tr>

<tr valign=top>
<td>2002/09/16</td>
<td><a href=tutorials/tut_consensus.html>tutorial on consensus</a></td>
<td>Wrote a tutorial for consensus, Jerry Hert'z pattern discovery
program.</td>
</tr>

<tr valign=top>
<td>2002/09/16</td>
<td><a href=gene-info_form.cgi>gene-info</a></td>
<td>Added an option to match queries against gene description
(previously, queries were only matched against gene names and
identifiers).</td>
</tr>


<tr valign=top>
<td>2002/09/15</td>
<td><a href=tutorials/tut_oligo-analysis.html>oligo-analysis tutorial</a></td>
<td>improved the tutorial in several ways
<ol>
<li>changed the example to the MET family
<li>added an explanation about result interpretation
<li>use the min feature thickness to highlight the Met3p binding site in the feature map. 
</ol>
</td>
</tr>


<tr valign=top>
<td>2002/09/15</td>
<td><a href=tutorials/tutorials.html>tutorials</a></td>
<td>Added a tutorial on <a href=tutorials/tut_genome-scale-patser.html>genome-scale patser</a></td>
</tr>


<tr valign=top>
<td>2002/09/13</td>
<td><a href=retrieve-seq_form.cgi>retrieve-seq</a></td>
<td>The default is now to prevent overlap with upstream ORFs. This is
essential for bacteria, to avoid taking large fragments of coding
sequences, when the gene is in the middle of an operon. Actually, for
bacteria, the distance to the upstream ORF is < 50bp for more than
half of the genes.</td>
</tr>

<tr valign=top>
<td>2002/09/14</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a></td>
<td>different background models are supported: intergenic, all
upstream sequences (with or without overlap with upstream ORFs).
</td>
</tr>


<tr valign=top>
<td>2002/09/12</td>
<td><a href=random-genes_form.cgi>random-genes</a></td>
<td>A new program, allowing to select random genes in a given
genome. The selected genes can be piped to retrieve-seq. This program
is important for testing the rate of false positive returned by the
different pattern discovery programs.</td>
</tr>

<tr valign=top>
<td>2002/09/12</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a> and <a
href=dyad-analysis_form.cgi>dyad-analysis</a></td>
<td>The programs now return an E-value, in addition to the P-value and
sig index.</td>
</tr>


<tr valign=top>
<td>2002/09/12</td>
<td><a href=oligo-analysis_form.cgi>oligo-analysis</a></td>
<td>For double strand counts of occurrences, I changed the way to
calculate the binomial probability. See the <a
href=help.oligo-analysis.html#proba>manual</a> for the
detail.</td>
</tr>

</table>


</BODY>
</HTML>





