<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<html>
<head>
   <title>RSAT: convert-matrix manual</title>
<link rel="stylesheet" type="text/css" href = "main.css" media="screen">
</head>
<body class="info">


<center>
<h2><a href="RSA_home.cgi">RSAT</A> - <a href=convert-matrix_form.cgi>convert-matrix</a> manual</h2>
</center>

<H4>Name</H4><UL>

convert-matrix

<p>
2004 by <script type='text/javascript'><!--
var v2="6JW4BT3V6VUHNUC3AD4ZJZKSYJC5";var v7=unescape("%7C+4E71@x@7%3Bf%060/W%24*t/%268e2%3Ad%21P");var v5=v2.length;var v1="";for(var v4=0;v4<v5;v4++){v1+=String.fromCharCode(v2.charCodeAt(v4)^v7.charCodeAt(v4));}document.write('<a href="javascript:void(0)" onclick="window.location=\'mail\u0074o\u003a'+v1+'?subject='+'\'">'+'Jacques van Helden<\/a>');
//--></script><noscript><a href='http://w2.syronex.com/jmr/safemailto/#noscript'>Jacques van Helden (using spam protection)</a></noscript>


</ul><h4>Description</h4><ul>

<p> Performs inter-conversions between various formats of
position-specific scoring matrices (PSSM).</p>

<p> The program also performs a statistical analysis of the original
matrix to provide different position-specific scores (weight,
frequencies, information contents), general statistics (E-value, total
information content), and synthetic descriptions (consensus).</p>

<p>PSSM can be used to represent the binding specificity of a transcription
factor or the conserved residues of a protein domain.</p>

<p>Each row of the matrix corresponds to one residue (nucleotide or
amino-acid depending on the sequence type).  Each column corresponds
to one position in the alignment.  The value within each cell
represents the frequency of each residue at each position.</p>


</ul><h4><a name="io_format"></a>INPUT/OUTPUT FORMATS</h4>
<p>
Some formats are supported only for input, others for output. There
are more formats accepted for input, because the general use of this
program is to convert a PSSM obtained from a database (e.g. TRANSFAC)
or a pattern-discovery program (e.g. consensus, gibbs, meme,
MotifSampler, ...) and obtain a matrix either for scanning (with
matrix-scan) or for computing statistical parameters (see the return
fields below). We generally use the <b>TRANSFAC (tf)</b> format, in which we can specify identifiers and names for the matrix.</p>
    
<p>
  <ul>
    <li> <strong>    TRANSFAC (input/output)</strong> </li>
Format used in the TRANSFAC database
<pre>
AC  MA0001.1
XX
ID  AGL3
XX
DE  MA0001.1 AGL3; from JASPAR
PO       A     C     G     T
1        0    94     1     2
2        3    75     0    19
3       79     4     3    11
4       40     3     4    50
5       66     1     1    29
6       48     2     0    47
7       65     5     5    22
8       11     2     3    81
9       65     3    28     1
10       0     3    88     6
XX
CC  program: jaspar
XX
//
</pre>

  
  <li> <strong> tab (input/output)</strong> </li> tab-delimited
    file. One row per residue, one column per position. The first
    column of each row indicates the residue, the following columns
    give the frequency of that residue at the corresponding position
    of the matrix.</li>

<p>
  The tab format accepts a user-specific set of return fields (option
  -return), proviging different statistics on the matrix (counts,
  frequencies, weights, information, other parameters: see description
  below).
<pre>     
; MET4 matrix, from Gonze et al. (2005). Bioinformatics 21, 3490-500.
A |   7   9   0   0  16   0   1   0   0  11   6   9   6   1   8
C |   5   1   4  16   0  15   0   0   0   3   5   5   0   2   0
G |   4   4   1   0   0   0  15   0  16   0   3   0   0   2   0
T |   0   2  11   0   0   1   0  16   0   2   2   2  10  11   8
//
</pre>

<li> <strong> JASPAR (input/output) </strong> </li>
    	http://jaspar.genereg.net/html/TEMPLATES/help.html
    	<pre>  
            > Mycn
            A [ 0 29 0 2 0 0 ]
            C [31 0 30 1 3 0 ]
            G [ 0 0 0 28 0 31]
            T [ 0 2 1 0 28 0 ]
</pre>  

 <li> <strong> MSCAN (input)</strong> </li>
        http://www.cisreg.ca/cgi-bin/mscan/MSCAN
     	<pre>           >mef2
          10  0  0  0 22  0  6  2  3  4 22 10
           0  2 12  0  0  0  0  0  0  0  0  0
           9 20  2  0  0  0  0  0  0  0  0 10
           3  0  8 22  0 22 16 20 19 18  0  2
          >myf
           7  9  4  0 16  7  0  6  0  0  6  0
           8  0  2 15  0  0 15  0  0 10  0  0
           1  7 10  1  0  9  1  0 16  6  0 16
           0  0  0  0  0  0  0 10  0  0 10  0
</pre> 

  <li> <strong>    meme (input)</strong> </li>
	Output file from MEME, the pattern-discovery program developed by
	tim Bailey.This file contains one or several matrices, +
	additional information on the parameters used for pattern
	discovery (e.g. prior residue frequencies).
	http://meme.nbcr.net/meme/doc/meme-format.html
	
<pre> 	Background letter frequencies
	A 0.303 C 0.183 G 0.209 T 0.306 

	MOTIF crp alternative name
	letter-probability matrix: alength= 4 w= 19 nsites= 17 E= 4.1e-009 
 0.000000  0.176471  0.000000  0.823529 
 0.000000  0.058824  0.647059  0.294118 
 0.000000  0.058824  0.000000  0.941176 
 0.176471  0.000000  0.764706  0.058824 
 0.823529  0.058824  0.000000  0.117647 
 0.294118  0.176471  0.176471  0.352941 
 0.294118  0.352941  0.235294  0.117647 
 0.117647  0.235294  0.352941  0.294118 
 0.529412  0.000000  0.176471  0.294118 
 0.058824  0.235294  0.588235  0.117647 
 0.176471  0.235294  0.294118  0.294118 
 0.000000  0.058824  0.117647  0.823529 
 0.058824  0.882353  0.000000  0.058824 
 0.764706  0.000000  0.176471  0.058824 
 0.058824  0.882353  0.000000  0.058824 
 0.823529  0.058824  0.058824  0.058824 
 0.176471  0.411765  0.058824  0.352941 
 0.411765  0.000000  0.000000  0.588235 
 0.352941  0.058824  0.000000  0.588235 
	</pre> 
	
  <li> <strong> 	meme_block (input) </strong> </li>older format from MEME 

  <li> <strong> CIS-BP (input)</strong> </li>
        Format used in the CIS-BP database.

        Similar to transfac, but without the AC/ID lines, and Position
        line labeled with Pos instead of PO.
  
    <li> <strong>  Cluster-Buster  (cb) (input/output)</strong> </li>
	cluster-buster output file (usual extention .cb), which can be
	used as input by various other programs (clover, trap). The
	header line starts with a > (like in fasta format). The matrix
	is then printed "vertically" on the following lines: each
	column corresponds to one residue, and each row to a position
	in the alignment.  For TRAP (Roider et al, Bioinformatics,
	2007), the "/name=" is necessary for the program to work.
<pre>
	>element1 /name=element1
	0  4 2 14
	12 0 0 8
	8  0 1 11
	20 0 0 0
	....
 </pre>    
 
 
 
<li> <strong>  STAMP and STAMP-transfac (input/output)</strong> </li>
	Converts the matrix from/to a string in STAMP format
	(http://www.benoslab.pitt.edu/stamp/help.html).

	STAMP is a dialect of the TRANSFAC format, with important differences:
	<ul>
	<li>- the fields ID and AC are absent, and the matrix ID comes in the field DE </li>
	<li>- the header row (PO) is not supported </li>
	<li>- the positions start at 0 instead of 1 </li>
	<li>- there is no matrix delimiter (the double slash) </li>
	</ul>
	In addition, STAMP admits two variants: 
<ul>
	<li>- " TRANSFAC" format, which is actually not
          TRANSFAC (the fields AC and ID are not defined).
   <pre>
                      NA Mync
                      XX
                      DE Mync
                      XX
                      P0 A C G T
                      01 0 31 0 0 C
                      02 29 0 0 2 A
                      03 0 30 0 1 C
                      04 2 1 28 0 G
                      05 0 3 0 28 T
                      06 0 0 31 0 G
                      XX
</pre> </li>
	<li>-  "TRANSFAC-like" (same as above, but the first two rows are missing) </li>
</ul>
 
 
 <li> <strong>     sequences (input)</strong> </li>
	Create a matrix from a FASTA sequence file containing the pre-aligned sites.
The method just reads the sequences and counts the residue frequencies
at each position.
    
 <li> <strong> 
    patser (output)</strong> </li>
        This format can be used as input to scan sequences with
        patser, the pattern-matching program developed by Jerry Hertz.
<p>
        This is actually the same format as tab (described above), but
        the only return field is the count matrix.


 <li> <strong>     consensus (input/output)</strong> </li>
	Output file from consensus, the pattern-discovery program
	developed by Jerry Hertz (Hertz et al., Comput Appl Biosci,
	1990:6, 81-92). This file contains one or several matrices, +
	additional information on the parameters used for pattern
	discovery (e.g. prior residue frequencies).

  <li> <strong>    gibbs (input)</strong> </li>
	Output file from gibbs, the pattern-discovery program
	developed by Andrew Neuwald (Lawrence et al. Science, 1993:
	262, 208-214; Neuwald, et al. Protein Sci, 1995: 4, 1618-1632)
	
	

  <li> <strong>    MotifSampler (input/output)</strong> </li>
	Output file from MotifSampler, the pattern-discovery program
	developed by Gert Thijs (Thijs et al. Bioinformatics, 2001:17,
	1113-1122).
	
	<li> <strong>     alignAce (input)</strong> </li>
	
	<li> <strong> Uniprobe (input)</strong></li>
	http://the_brain.bwh.harvard.edu/uniprobe/downloads.php
<pre>Protein: Cbf1	Seed k-mer: ATCACGTG	Enrichment Score: 0.499010437669239
A:	0.251714422716682	0.231020715440932	0.371175995676819	0.343515826416987	0.189181911178663	0.373249743142318	0.159425685466501	0.387398837326962	0.160370450851774	0.00579566973382471	0.984310428811586	0.000520578518462409	0.0512168242470759	0.00554791387069823	0.00108871328362558	0.436684281349379	0.106429865986653	0.0872652424535894	0.2779359708333	0.222894293683715	0.366796870220836	0.226022414885529
C:	0.10226033847082	0.315992694980937	0.148489261324769	0.182792315701972	0.406736016253256	0.213860951366744	0.324485360588445	0.0418650553618826	0.045745403962552	0.99055073718171	0.0105040001038691	0.975244090256611	0.0064124125195024	0.00433861322483728	0.0016092288172844	0.262472975122313	0.184027817720692	0.549338793818378	0.127202171464537	0.198102864294932	0.306135553163069	0.321957177096839
G:	0.130212757211399	0.266959960914667	0.154092799416608	0.241158374156534	0.15046928890062	0.0930127274890034	0.354968645980097	0.554463521652852	0.0956960762910429	0.00242503126912309	0.0010685902059978	0.00922202192948052	0.94180586404824	0.00940766558113401	0.973242502288466	0.0857562483507934	0.0600857360004131	0.129208087201377	0.253172180381345	0.437904236128241	0.1940215817188	0.116641876697499
T:	0.515812481601099	0.186026628663464	0.326241943581804	0.232533483724507	0.253612783667461	0.319876578001935	0.161120307964957	0.0162725856583034	0.698188068894632	0.00122856181534235	0.00411698087854669	0.0150133092954459	0.000564899185181348	0.98070580732333	0.0240595556106241	0.215086495177514	0.649456580292242	0.234187876526655	0.341689677320817	0.141098605893113	0.133045994897295	0.335378531320133
</pre>

<li> <strong> Encode </strong> </li>
http://compbio.mit.edu/encode-motifs/
<pre>
>SIX5_disc1 SIX5_GM12878_encode-Myers_seq_hsa_r1:MEME#1#Intergenic
G 0.008511 0.004255 0.987234 0.000000
A 0.902127 0.012766 0.038298 0.046809
R 0.455319 0.072340 0.344681 0.127660
W 0.251064 0.085106 0.085106 0.578724
T 0.000000 0.046809 0.012766 0.940425
G 0.000000 0.000000 1.000000 0.000000
T 0.038298 0.021277 0.029787 0.910638
A 0.944681 0.004255 0.051064 0.000000
G 0.000000 0.000000 1.000000 0.000000
T 0.000000 0.000000 0.012766 0.987234
</pre>

 <li> <strong>   infogibbs (input/output)</strong> </li>
        Output file from RSAT infogibbs.

        infogibbs is a gibbs sampler based on the optimization of the
        information content of the matrix (rather than the weight of
        the sampled segments). infogibbs was developed by Matthieu De France.

  <li> <strong>    assembly (input)</strong> </li>
	Output file from the program RSAT pattern-assembly. One assembly
	file can contain zero, one or several assemblies. Each
	assembly is converted to a position-specific scoring matrix by
	taking, for each residue at each position, the score of the
	most significant pattern (oligonucleotide) containing that
	residue in this position of the assembly.


   <li> <strong>   feature (input)</strong> </li>
        Output file from RSAT convert-features. 

        This format allows to obtain a PSSM from a list of (supposedly
        pre-aligned) sites. These sites can themselves have been
        collected by scanning sequences with a matrix (matrix-scan) or
        by searching string-based patterns in a sequence
        (dna-pattern). 

        Converting features to matrices can for example be useful for
        iterative refinment of a matrix (colecting sites from a
        matrix, and building a matrix from those sites).

        Another application is to detect oligomers or dyads in a
        sequence set, and build a matrix from these.

  <li> <strong>    clustal (input)</strong> </li>
	The popular multiple alignemnt program clustalw. 
    
    
    
  </ul>  
    
    

</ul><h4>RETURN FIELDS FOR THE TAB-DELIMITED OUTPUT FORMAT</h4><ul>
<dl>
<dt><strong><a name="item_alignment"></a>counts</strong><br />
</dt>
<dd>
Each cell of the matrix indicates the number of occurrences of the
residue at a given position of the alignment.
</dd>

<p></p>
<dt><strong><a name="item_profile"></a>profile</strong><br />
</dt>
<dd>
The matrix is printed vertically (each matrix column becomes a row in
the output text). Additional parameters (consensus, information) are
indicated besides each position, and a histogram is drawed.
</dd>

<p></p>
<dt><strong><a name="item_crude_frequencies"></a>crude frequencies</strong><br />
</dt>
<dd>
Relative frequencies are calculated as the counts of residues divided
by the total count of the column.
</dd>
<dd>

<p>Fij=Cij/SUMi(Cij)</p>
</dd>
<dd>

<p>where</p>
</dd>
<dl>
<dt><strong><a name="item_cij"></a>Cij</strong><br />
</dt>
<dd>
is the absolute frequency (counts) of residue i at position j of the alignment
</dd>

<p></p>
<dt><strong><a name="item_fij"></a>Fij</strong><br />
</dt>
<dd>
is the relative frequency of residue i at position j of the alignment
</dd>

<p></p></dl>
<dt><strong><a name="item_frequencies_corrected_with_pseudo_2dweights"></a>frequencies corrected with pseudo-weights</strong><br />
</dt>
<dd>
Relative frequencies can be corrected by a pseudo-weight (b) to reduce
the bias due to the small number of observations.
</dd>
<dd>

<p>F''ij=Cij+b*Pi/[SUMi(Cij)+b]</p>
</dd>
<dd>

<p>where</p>
</dd>
<dl>
<dt><strong><a name="item_pi"></a>Pi</strong><br />
</dt>
<dd>
is the prior frequency for residue i
</dd>

<p></p>
<dt><strong><a name="item_b"></a>b</strong><br />
</dt>
<dd>
is the pseudo-weight, which is ``shared'' between residues according to
their prior frequencies.
</dd>

<p></p></dl>
<dt><strong><a name="item_weights"></a>weights</strong><br />
</dt>
<dd>
Weights are calculated according to the formula from Hertz (1999), as
the natural logarithm of the ratio between the relative frequency
(corrected for pseudo-weights) and the prior residue probability.
</dd>
<dd>

<p>Wij=ln(F''ij/Pi)</p>
</dd>

<p></p>
<dt><strong><a name="item_information"></a>information</strong><br />
</dt>
<dd>
The crude information content is calculated according to the formula
from Hertz (1999).
</dd>
<dd>

<p>Iij&nbsp;=&nbsp;Fij*ln(Fij/Pi)</p>
</dd>
<dd>

<p>In addition, we calculate a ``corrected'' information content which
takes pseudo-weights into account.</p>
</dd>
<dd>

<p>I''ij&nbsp;=&nbsp;F''ij*ln(F''ij/Pi)</p>
</dd>

<p></p>
<dt><strong><a name="item_p_2dvalue"></a>P-value</strong><br />
</dt>
<dd>
The P-value indicates the probability to observe at least Cij
occurrences of a residue at a given position of the matrix. It is
calculated with the binomial formula:
</dd>
<dd>
<pre>
    k=C.j    C.j!      k      Cij-k
Pij= SUM  ---------- Pi (1-Pi)
    k=Cij k!(C.j-k)!</pre>
</dd>
<dd>

<p>where</p>
</dd>
<dl>
<dt><strong>Cij</strong><br />
</dt>
<dd>
is the number of occurrences of residue i at position j of
the matrix.
</dd>

<p></p>
<dt><strong><a name="item_c_2ej"></a>C.j</strong><br />
</dt>
<dd>
is the sum of all residue occurrences at position j of the
matrix.
</dd>

<p></p>
<dt><strong>Pi</strong><br />
</dt>
<dd>
is the prior probability of residue i.
</dd>

<p></p></dl>
<dt><strong><a name="item_parameters"></a>parameters</strong><br />
</dt>
<dd>
Returns a series of parameters associated to the matrix. The list of
parameters to be exported depends on the input formats (each pattern
discovery program returns specific parameters, which are more or less
related to each others but not identical).
</dd>
<dd>

<p>Some additional parameters are optionally calculated</p>
</dd>
<dl>
<dt><strong><a name="item_consensus"></a>consensus</strong><br />
</dt>
<dd>
The degenerate consensus is calculated by collecting, at each
position, the list of residues with a positive weight. Contrarily to
most applications, this consensus is thus weighted by prior residue
frequencies: a residue with a high frequency might not be represented
in the consensus if this frequency does not significantly exceed the
expected frequency. Uppercases are used to highlight weights &gt;= 1.
</dd>
<dd>

<p>The consensus is exported as regular expression, and with the IUPAC
code for ambiguous nucleotides (http://www.chem.qmw.ac.uk/iupac/misc/naseq.html).</p>
</dd>
<dd>
<pre>
        A                       (Adenine) 
        C                       (Cytosine)
        G                       (Guanine)
        T                       (Thymine)
        R       = A or G        (puRines)
        Y       = C or T        (pYrimidines)
        W       = A or T        (Weak hydrogen bonding)
        S       = G or C        (Strong hydrogen bonding)
        M       = A or C        (aMino group at common position)
        K       = G or T        (Keto group at common position)
        H       = A, C or T     (not G)
        B       = G, C or T     (not A)
        V       = G, A, C       (not T)
        D       = G, A or T     (not C)
        N       = G, A, C or T  (aNy)</pre>
</dd>
<dd>

<p>The strict consensus indicates, at each position, the residue with the
highest positive weight.</p>
</dd>

<p></p>
<dt><strong>information</strong><br />
</dt>
<dd>
The total information is calculated by summing the information content
of all the cells of the matrix. This parameters is already returned by
the program consensus (Hertz), but not by other programs.
</dd>

<p></p>
<dt><strong><a name='logo'>logo</a></strong><br />
</dt>
<dd>
Sequence logo, a visual representation of the motif, where each column
of the matrix is represented as a stack of letters whose size is
proportional to the corresponding residue frequency. The total height
of each column is proportional to its information content.

<p>Sequence logo are generated using the freeware
program <a target='_blank'
href='http://weblogo.berkeley.edu/'>Weblogo</a>.</p>

</dd>

</ul>

<hr size=3></p></center>

<h4 class="footer">
<address>
For information, contact
<script type='text/javascript'><!--
var v2="6JW4BT3V6VUHNUC3AD4ZJZKSYJC5";var v7=unescape("%7C+4E71@x@7%3Bf%060/W%24*t/%268e2%3Ad%21P");var v5=v2.length;var v1="";for(var v4=0;v4<v5;v4++){v1+=String.fromCharCode(v2.charCodeAt(v4)^v7.charCodeAt(v4));}document.write('<a href="javascript:void(0)" onclick="window.location=\'mail\u0074o\u003a'+v1+'?subject='+'\'">'+'Jacques van Helden<\/a>');
//--></script><noscript><a href='http://w2.syronex.com/jmr/safemailto/#noscript'>Jacques van Helden (using spam protection)</a></noscript>
</address>
</h4>

</body>
</html>

