<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>/no_backup/rsa-tools/perl-scripts/matrix-scan</title>
<link rev="made" href="mailto:root@b01.apple.com" />
</head>

<body style="background-color: white">

<p><a name="__index__"></a></p>
<!-- INDEX BEGIN -->

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#authors">AUTHORS</a></li>
	<li><a href="#category">CATEGORY</a></li>
	<li><a href="#usage">USAGE</a></li>
	<li><a href="#input_formats">INPUT FORMATS</a></li>
	<ul>

		<li><a href="#sequence_file">Sequence file</a></li>
		<li><a href="#matrix_file">Matrix file</a></li>
	</ul>

	<li><a href="#output_format">OUTPUT FORMAT</a></li>
	<li><a href="#scoring_scheme">SCORING SCHEME</a></li>
	<li><a href="#background_models">BACKGROUND MODELS</a></li>
	<li><a href="#references">REFERENCES</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#internal_procedures">INTERNAL PROCEDURES</a></li>
	<li><a href="#see_also">SEE ALSO</a></li>
</ul>
<!-- INDEX END -->

<hr />
<p>
</p>
<hr />
<h1><a name="name">NAME</a></h1>
<p>matrix-scan</p>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>Scan sequences with one or several position-specific scoring matrices
(PSSM) to identify instances of the corresponding motifs (putative
sites). This program supports a variety of background models
(Bernoulli, Markov chains of any order).</p>
<p>
</p>
<hr />
<h1><a name="authors">AUTHORS</a></h1>
<dl>
<dt><strong><a name="item_jean_valery_turatsinze_jturatsi_40scmbb_2eulb_2eac">Jean Valery Turatsinze <a href="mailto:jturatsi@scmbb.ulb.ac.be">jturatsi@scmbb.ulb.ac.be</a></a></strong><br />
</dt>
<dt><strong><a name="item_jacques_van_helden_jvanheld_40scmbb_2eulb_2eac_2eb">Jacques van Helden <a href="mailto:jvanheld@scmbb.ulb.ac.be">jvanheld@scmbb.ulb.ac.be</a></a></strong><br />
</dt>
</dl>
<p>
</p>
<hr />
<h1><a name="category">CATEGORY</a></h1>
<dl>
<dt><strong><a name="item_sequences">sequences</a></strong><br />
</dt>
<dt><strong><a name="item_pattern_matching">pattern matching</a></strong><br />
</dt>
<dt><strong><a name="item_pssm">PSSM</a></strong><br />
</dt>
</dl>
<p>
</p>
<hr />
<h1><a name="usage">USAGE</a></h1>
<p>matrix-scan -m matrixfile [-i inputfile] [-o outputfile] [-v]
    [-bgfile backgroundfile|-bgorder #]</p>
<p>
</p>
<hr />
<h1><a name="input_formats">INPUT FORMATS</a></h1>
<p>
</p>
<h2><a name="sequence_file">Sequence file</a></h2>
<p>All the formats supported in RSAT can be used as input (default: fasta).</p>
<p>
</p>
<h2><a name="matrix_file">Matrix file</a></h2>
<p>The matrix format is specified with the option -matrix_format.  Supported :
consensus,gibbs,meme,tab. Default : tab.</p>
<p>
</p>
<hr />
<h1><a name="output_format">OUTPUT FORMAT</a></h1>
<p>The output is a tab-delimited file, with one row per match. This file
can directly be used as input for the program <em>feature-map</em>.</p>
<p>
</p>
<hr />
<h1><a name="scoring_scheme">SCORING SCHEME</a></h1>
<p>The program scans the input sequences with a position-specific scoring
matrix (PSSM) by selecting, at each position, a sequence segment (S)
of the same length as the matrix, and assigning a score to this
segment.</p>
<p>The secgment score (weight) is calculated according to the theory
developed by Jerry Hertz and Gary Stormo, but with the capability to
use Markov chain-based background models, as prposed by Thijs et
al. (2001).</p>
<p>The weight of a sequence segment (Ws) is calculated as the log-ratio
between two probabilities:</p>
<dl>
<dt><strong><a name="item_p"><em>Ws = log[P(S|M)/P(S|B)]</em></a></strong><br />
</dt>
<dd>
where
</dd>
<dl>
<dt><strong><em>proba_M = P(S|M)</em>:</strong><br />
</dt>
<dd>
The probability to generate the sequence segment given the matrix.
</dd>
<p></p>
<dt><strong><em>proba_B = P(S|B)</em>:</strong><br />
</dt>
<dd>
The probability to generate the sequence segment given the background
model.
</dd>
<p></p></dl>
<p>By default, the program uses natural logarithms, but the option -base
allows to specify any alternative base (e.g. 2 to obtain bit units);</p>
</dl>
<p>
</p>
<hr />
<h1><a name="background_models">BACKGROUND MODELS</a></h1>
<p>An essential parameter of any pattern detection approach is the choice
of an appropriate background model. This model is used to estimate the
probability for each site to occur by chance, rather than as an
instance of the motif.</p>
<p>The program matrix-scan supports Markov models of arbitrary order as
background models. A Markov model of order m means that the
probability of each residue depends on the m preceding residues in the
sequence. Note that a Markov model of order 0 corresponds to a
Bernoulli model, since eahc residue is independent from the preceding
ones.</p>
<p>A variety of options for specifying the background model.</p>
<dl>
<dt><strong><a name="item__2dbgfile"><em>-bgfile</em></a></strong><br />
</dt>
<dd>
This option allows to enter the background model from a background
model file. Background model files are tab-delimited files containing
the specification of oligonucleotide frequencies. A Markov model of
order m=k-1 is automatically obtained from the frequencies of
oligonucleotides of length k. There is thus no need to use the option
-markov when the background model is secified with a bg file.
</dd>
<dd>
<p>The RSAT data folder contains pre-calibrated background model files
for all the supported organisms.</p>
</dd>
<dd>
<p>$RSAT/data/genomes/My_organism/oligo-frequencies/</p>
</dd>
<p></p>
<dt><strong><a name="item__2dbginput"><em>-bginput</em></a></strong><br />
</dt>
<dd>
The backgound model is calculated from the whole set of input
sequences. This option requires to specify the order of the background
model with the option -markov.
</dd>
<p></p>
<dt><strong><a name="item__2dwindow"><em>-window</em></a></strong><br />
</dt>
<dd>
The background model is calculated locally at each step of the scan,
by computing transition frequencies from a sliding window centred
around the considered segment. The model is thus updated at each
scanned position. This model is called ``adaptive''. Note that the
sliding window must be large enough to train the local Markov
model. The required sequence length increases exponentially with the
Markov order. This option is thus usually suitable for short order
models only (-markov 0 to 1).
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="references">REFERENCES</a></h1>
<p>1. Hertz, G.Z. and G.D. Stormo, Identifying DNA and protein patterns
   with statistically significant alignments of multiple
   sequences. Bioinformatics, 1999. 15(7-8): p. 563-77.</p>
<p>2. Hertz, G.Z., G.W. Hartzell, 3rd, and G.D. Stormo, Identification of
   consensus patterns in unaligned DNA sequences known to be
   functionally related. Comput Appl Biosci, 1990. 6(2): p. 81-92.</p>
<p>3. Thijs, G., et al., A higher-order background model improves the
   detection of promoter regulatory elements by Gibbs
   sampling. Bioinformatics, 2001. 17(12): p. 1113-22.</p>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<dl>
<dt><strong><a name="item__2dv__23"><strong>-v #</strong></a></strong><br />
</dt>
<dd>
Level of verbosity (detail in the warning messages during execution)
</dd>
<p></p>
<dt><strong><a name="item__2dh"><strong>-h</strong></a></strong><br />
</dt>
<dd>
Display full help message
</dd>
<p></p>
<dt><strong><a name="item__2dhelp"><strong>-help</strong></a></strong><br />
</dt>
<dd>
Same as -h
</dd>
<p></p>
<dt><strong><a name="item__2dbgfile_background_file"><strong>-bgfile background_file</strong></a></strong><br />
</dt>
<dd>
Background model file.
</dd>
<p></p>
<dt><strong><strong>-bginput</strong></strong><br />
</dt>
<dd>
Calculate background model from the input sequence set.
</dd>
<p></p>
<dt><strong><a name="item__2dmarkov"><strong>-markov</strong></a></strong><br />
</dt>
<dd>
Order of the markov chain for the background model.
</dd>
<dd>
<p>This option is incompatible with the option -bgfile.</p>
</dd>
<p></p>
<dt><strong><strong>-window</strong></strong><br />
</dt>
<dd>
Size of the sliding window for the background model calculation.
</dd>
<p></p>
<dt><strong><a name="item__2dm_matrixfile"><strong>-m matrixfile</strong></a></strong><br />
</dt>
<dd>
Matrix file.
</dd>
<dd>
<p>This argument can be used iteratively to scan the sequence with
multiple matrices.</p>
</dd>
<p></p>
<dt><strong><a name="item__2dseq_format_sequence_format"><strong>-seq_format sequence_format</strong></a></strong><br />
</dt>
<dd>
Sequence format.
</dd>
<p></p>
<dt><strong><a name="item__2dmatrix_format_matrix_format"><strong>-matrix_format matrix_format</strong></a></strong><br />
</dt>
<dd>
Matrix format.
</dd>
<p></p>
<dt><strong><a name="item__2di_inputfile"><strong>-i inputfile</strong></a></strong><br />
</dt>
<dd>
If no input file is specified, the standard input is used.  This
allows to use the command within a pipe.
</dd>
<p></p>
<dt><strong><a name="item__2dmask_upper_7clower"><strong>-mask upper|lower</strong></a></strong><br />
</dt>
<dd>
Mask lower or uppercases, respecively, i.e. replace selected case by N
characters.
</dd>
<p></p>
<dt><strong><a name="item__2dn__5bskip_7cscore_5d"><strong>-n [skip|score]</strong></a></strong><br />
</dt>
<dd>
Treatment of N characters. These characters are often used in DNA
sequences to represent undefined or masked nucleotides.
</dd>
<dl>
<dt><strong><a name="item_skip"><em>skip</em></a></strong><br />
</dt>
<dd>
N-containing regions are skipped.
</dd>
<p></p>
<dt><strong><a name="item_score"><em>score</em></a></strong><br />
</dt>
<dd>
N-containing regions are scored. The probability of an N is 1 for both
the background model and the matrix. The N residues will thus
contribute neither positively nor negatively to the weight score of
the N-containing fragment. This option can be useful to detect sites
which are at the border of N-containing regions, or in cases there are
isolated N in the sequences.
</dd>
<p></p></dl>
<dt><strong><a name="item__2do_outputfile"><strong>-o outputfile</strong></a></strong><br />
</dt>
<dd>
If no output file is specified, the standard output is used.  This
allows to use the command within a pipe.
</dd>
<p></p>
<dt><strong><a name="item__2dpseudo_pseudo_counts"><strong>-pseudo pseudo_counts</strong></a></strong><br />
</dt>
<dd>
Pseudo-weight.
</dd>
<p></p>
<dt><strong><a name="item__2dequi_pseudo"><strong>-equi_pseudo</strong></a></strong><br />
</dt>
<dd>
If this option is called, the pseudo-weight is distributed in an equiprobable
way between residues.
</dd>
<dd>
<p>By default, the pseudo-weight is distributed proportionally to residue priors.</p>
</dd>
<p></p>
<dt><strong><a name="item__2dorigin_pos"><strong>-origin pos</strong></a></strong><br />
</dt>
<dd>
Define pos as the origin for the calculation of positions.
</dd>
<dd>
<p>-origin -0 defines the end of each sequence as the origin. The matching
positions are then negative values, providing the distance between the match
and the end of the sequence.</p>
</dd>
<p></p>
<dt><strong><a name="item__2dbase__23"><strong>-base #</strong></a></strong><br />
</dt>
<dd>
Base for the logarithms (default: $base)
</dd>
<p></p>
<dt><strong><a name="item__2ddecimals__23"><strong>-decimals #</strong></a></strong><br />
</dt>
<dd>
Number of decimals displayed for the score.
</dd>
<p></p>
<dt><strong><a name="item__2duth_param_upper_threshold"><strong>-uth param upper_threshold</strong></a></strong><br />
</dt>
<dd>
Threshold on some parameter (-lth: lower, -uth: upper threshold).
</dd>
<dd>
<p>Supported threshold fields : score normw proba_M proba_B rank</p>
</dd>
<p></p>
<dt><strong><a name="item__2d2str"><strong>-2str</strong></a></strong><br />
</dt>
<dd>
Scan both strands for DNA sequences.
</dd>
<p></p>
<dt><strong><a name="item__2d1str"><strong>-1str</strong></a></strong><br />
</dt>
<dd>
single-strand search for DNA sequences.
</dd>
<p></p>
<dt><strong><a name="item__2dreturn_return_fields"><strong>-return return_fields</strong></a></strong><br />
</dt>
<dd>
List of fields to return.
</dd>
<dd>
<p>Supported fields: sites,ranks,limits,normw,bg_model,matrix,freq_matrix,weight_matrix</p>
</dd>
<dl>
<dt><strong><a name="item_sites_3a_matching_sites_2e"><strong>sites:</strong> Matching sites.</a></strong><br />
</dt>
<dd>
Return the osition of matching sites in aformat which can be sed as
input by <em>feature-map</em>.
</dd>
<p></p>
<dt><strong><a name="item_rank_3a_rank_2e"><strong>rank:</strong> Rank.</a></strong><br />
</dt>
<dd>
Sort the sites per decreasing values of score (weight), and return the
rank value. The rank is calculated independently for each sequence and
each matrix.
</dd>
<dd>
<p>A common usage of the rank is to select the top scoring site per
matrix and per sequence (-th rank 1) or the 3 top scoring sites per
matrix and per sequence (-uth rank 3).</p>
</dd>
<p></p>
<dt><strong><a name="item_limits"><strong>limits:</strong> limits (start, end) of the input sequences.</a></strong><br />
</dt>
<dd>
This is useful for drawing feature maps with sequences of different
lengths.
</dd>
<p></p>
<dt><strong><a name="item_normw_3a_normalized_weights_2e"><strong>normw:</strong> normalized weights.</a></strong><br />
</dt>
<dd>
Normailzed weights are calculated according to Thijs' formula : normw
= (W -Wmin)/(Wmax - Wmin). Note that Wmin and Wmax are approximated
using a Bernoulli model, for reasons of commputational efficiency.
</dd>
<p></p>
<dt><strong><a name="item_bg_model_3a_background_model_2e"><strong>bg_model:</strong> Background model.</a></strong><br />
</dt>
<dd>
Transition matrix of the background model. Note that this option only
makes sense for fixed background models (-bgfile or -bginput), since
when the background model is adaptive (-mindow), the transition matrix
changes along he sequence.
</dd>
<p></p></dl>
</dl>
<p>
</p>
<hr />
<h1><a name="internal_procedures">INTERNAL PROCEDURES</a></h1>
<dl>
<dt><strong><a name="item_score_segment"><strong>score_segment</strong></a></strong><br />
</dt>
<dd>
Assign a score to a sequence segment and print it if it passes the
thresholds. Return value is 1 if the segment passed the thresholds.
</dd>
<p></p>
<dt><strong><a name="item_print_match"><strong>print_match</strong></a></strong><br />
</dt>
<dd>
Print the matching site.
</dd>
<p></p>
<dt><strong><a name="item_check_thresholds"><strong>check_thresholds</strong></a></strong><br />
</dt>
<dd>
Check the lower and upper threshold for a given score. 


</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="see_also">SEE ALSO</a></h1>
<dl>
<dt><strong><a name="item_convert_2dmatrix"><em>convert-matrix</em></a></strong><br />
</dt>
<dt><strong><a name="item_convert_2dbackground_2dmodel"><em>convert-background-model</em></a></strong><br />
</dt>
<dt><strong><a name="item_feature_2dmap"><em>feature-map</em></a></strong><br />
</dt>
</dl>

</body>

</html>
