<HTML>
<HEAD>
   <TITLE>RSA-tools: retrieve-seq manual</TITLE>
</HEAD>
<BODY TEXT="#000000" BGCOLOR="#FFEEDD" LINK="#0000EE" VLINK="#55008B" ALINK="#FF0000">

<H2 ALIGN=CENTER><A HREF="RSAT_home.cgi">RSA-tools</A> - retrieve-seq manual</H2>

<H2>Name</H2>
<UL>
retrieve-seq<P>
1997-98 by <script type='text/javascript'><!--
var v2="6JW4BT3V6VUHNUC3AD4ZJZKSYJC5";var v7=unescape("%7C+4E71@x@7%3Bf%060/W%24*t/%268e2%3Ad%21P");var v5=v2.length;var v1="";for(var v4=0;v4<v5;v4++){v1+=String.fromCharCode(v2.charCodeAt(v4)^v7.charCodeAt(v4));}document.write('<a href="javascript:void(0)" onclick="window.location=\'mail\u0074o\u003a'+v1+'?subject='+'\'">'+'Jacques van Helden<\/a>');
//--></script><noscript><a href='http://w2.syronex.com/jmr/safemailto/#noscript'>Jacques van Helden (using spam protection)</a></noscript>

</UL>

<H2>Description</H2>

<UL>
Returns upstream, downstream or ORF sequences for one or several genes. <P>
</UL>

<H2>Options</H2>

<UL>

<A NAME="genes"></A>
<h3>Genes</h3>
<ul>
Two options are proposed
<ul>
<li><b>All</b>: all genes of the selected organism are considered
<li><b>Selection</b>: user-specified list of genes 
<UL>
<li>gene list typed directly in the text area.
<li><B>Upload</B>: select a text file on your computer that contains
the list of genes
</UL>

</ul>
<P>
<ul>
<li>Several queries can be entered simultaneously, separated by carriage
returns. The first word of each line is the query; any following
information is ignored. </li>

<li>Valiad queries are ORF identifiers (eg: YFL021W) or gene names (eg:
GAL4, NIL1).</li>

<li>By default, synonyms are accepted (eg: NIL1 = GAT1), but only one gene
name is returned as sequence identifier (GAT1 in this case). </li>

</ul>

<A NAME="ids_only"></A>
</ul><h3>Query contains only IDs (no synonyms)</h3><ul>

By default, genes can be specified by their ID or by their
names. Several synonyms are recognized for some genes.

<p>The option <i>IDs only</i> indicates that the input queries contain
only IDs, no name or synonym. This avoids to load the table of
synonyms, and reduces thus the delay to obtain the result.

<A NAME="Organism"></A>
</ul><h3>Organism</h3><ul>

<A NAME="single_org"></A>
<h4>Single organism</h4><ul>
All the query genes belong to the same organism. 

When this option is checked, the query organism has to be specified among the
supported organisms, indicated by the pop-up menu <b>Organism</b>.  

<A NAME="multi_org"></A>
</ul><h4>Multiple organisms</h4><ul>
When this option is active, the input must contain two columns, separated by a
tab character. Each row of the input is considered as a query. 
<ol>
<li>The first column indicates the ID or the name of a query gene.</li>
<li>The second column incitaed the organism to which the gene belongs.</li>
</ol>

<h5>Example</h5>

<ul><pre>
NP_310394.1     Escherichia_coli_O157H7
NP_313053.1     Escherichia_coli_O157H7
NP_416175.1     Escherichia_coli_K12
NP_418467.1     Escherichia_coli_K12
NP_753947.1     Escherichia_coli_CFT073
NP_756866.1     Escherichia_coli_CFT073
NP_288094.1     Escherichia_coli_O157H7_EDL933
NP_290677.1     Escherichia_coli_O157H7_EDL933
</pre></ul>

<h5>Warning</h5>

For the organism names, all spaces must be replaced by the underscore
character, as whosn in the example.

</ul>

<A NAME="feattype"></A>
</ul><h3>Feature type</h3><ul>

Currently supported:
<ul>
<li><b>CDS:</b> coding sequences (from start to stop codon, unspliced)</li>
<li><b>mRNA:</b>messenger RNA</li>
<li><b>tRNA:</b>transfer RNA</li>
<li><b>rRNA:</b>ribosomial RNA</li>
<li><b>scRNA:</b></li>
</ul>

<P>
The availability of some sequence types depends on the genome. For
example, some Genbank flat files contain annotations about CDSs but no
mRNA (e.g. bacterial annotations from the NCBI). Some other genome
contain separate annotations for CDS and mRNA
(e.g. <i>A.thaliana</i>). When mRNAs are annotated in Genbank, their
coordinates are stored and can be used.

<P>
The advantage of using mRNA is that, if the mRNA is complete (which is
not always the case), the upstream regions aretrieved relative to the
transcription initiation site, rather than the start codon.

<P>
<b>Remarks</b>
<ol>
<li>One gene can be associated to multiple CDSs and to multiple mRNAs.
<li>Many annotated "mRNAs" seem to be actually CDS (e.g. in June 2003,
12,000 out of 27,000 mRNAs from <i>A.thaliana</i> start with ATG).
</ol>

</ul>
<A NAME="sequence_type"></A>
<h3>Sequence type</h3>
<ul>
Currentlly supported:
<ul>
<li><B>Upstream</B> sequences located upstream the coding region. The
origin is at the start codon.

<li><B>Downstream</B> sequences located downstream the coding
region. The origin is at the stop codon.

<li><B>Unspliced ORF</B> DNA sequences located between the start and
stop codons. WARNING: introns are not spliced out (this will be
implemented in further versions)

</ul>
</ul>

<A NAME="from_to"></A>
<h3>Sequence limits (from, to)</h3><ul>

Limits of the region to retrieve. Coordintates are calculated relative
to the ORF start.

<h4>Sign</h4><UL>

  <LI><B><FONT COLOR="#DD0000">negative</FONT></B> 
           values return sequence located <B><FONT COLOR="#DD0000">upstream</FONT></B> the origin 
    <LI><B><FONT COLOR="#00AA00">positive</FONT></B> 
           values return sequences <B><FONT COLOR="#00AA00">downstream</FONT></B> the origin 
      <br>
	   The origin itself depends on the sequence type, see above)

</ul><h4>Default values for upstream sequence retrieval</h4><ul>

<LI>For <B>yeast</B>, we generally obtain good results with upstream
regions from -800 to -1. About 99% of the known upstream elements are
comprized between these limits (source: Transfac).<BR> 

<LI>For <B>bacteria</B>, the distribution of regulatory sites depends
on the mode of regulation :
<UL>
<LI><B>transcriptional repressors</B> generally bind proximally, and
may overlap the transcription initiation or even be located downstream. A
good guess is from -200 to +50.  <LI>Binding sites for
<B>transcriptional activators</B> have a more distal distribution
(-400 to -1). <P>
</UL>
The default is from -400 to -1 from the start codon (since we
currently do not have annotations about transcription initiation
sites).

<li>The default values for each organism can be obtained with the
program <a
href=supported-organisms.cgi>supported-organisms</i>.
</UL>


</ul>

<A NAME="noorf"></A>
<h3>Prevent overlap with upstream ORFs:</h3>
<ul>
It is quite frequent to find a predicted ORF in close proximity
upstream from a query ORF. If you want to discard these sequences from
your analysis, you should make sure this option is active. 

<P>
When the option is checked, upstream sequences are automatically
clipped when a predicted ORF is located within the range defined by
the option <I>from</I>. The actual size retainedfor the upstream
sequence is indicated in the sequence comments.

<P> Note that in some cases a known regulatory element is located
upstream or within a predicted ORF. This means either that the
predited ORF is an artifact, or that the same sequence is bifunctional
(coding and regulatory). 

<P>
It is particularly important to activate this option when working with
bacteria, since many genes are located in operons, and have a very
close upstream neighbour.

<!--
<P> Note that the distance to the closest upstream neighbour can be
checked with the program <A
HREF="neighbour-orfs.html"><B>neighbour-orfs</B></A>.
-->

<P>
</ul>

<A NAME="formats"></A>
<h3>Output sequence format:</h3>
<ul>
The result can be displayed in various sequence formats (click on the links for more details). 
<UL>
<LI><A HREF="help.formats.html#raw"><B>raw</B></A>: the raw sequence without any identifier or comment.
<LI><A HREF="help.formats.html#multi"><B>multi</B></A>: several raw sequences concatenated.
<LI><A HREF="help.formats.html#ig"><B>IG</B></A>: IntelliGenetics format.
<LI><A HREF="help.formats.html#fasta"><B>FastA</B></A>: the sequence format used by FastA, BLAST, Gibbs sampler and a lot of other bioinformatic programs.
<LI><A HREF="help.formats.html#wconsensus"><B>Wconsensus</B></A>: the format defined by Jerry Hertz for his programs (patser, consensus, wconsensus).
</UL>
</UL>

<A NAME="seq_label"></A>
<h3>Sequence label</h3>
<ul>
Sequences can be labeled (named) in different ways:
<ul>
<li><b>ORF identifier</b>
<li><b>gene name</b>
<li><b>ORF id + gene name</b>
<li><b>full</b>: a concatenation of ORF identifier, gene name, sequence type, from, to and strand. This option gives a full description of the conditions of sequence retrieval
</ul> 
</ul>

</ul>

<H2>Availability</H2>
<UL>
The program can be used through its web interface at:
<A HREF="http://rsat.scmbb.ulb.ac.be/rsat/" target=_blank>
http://rsat.scmbb.ulb.ac.be/rsat/</A> <BR>

retrieve-seq is a perl script running on unix machines (SUN, SGI
have been tested). The web interface is a perl-cgi script.
</UL>

<HR SIZE=3>

</BODY>
</HTML>
