<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>get-orthologs-compara manual</title>
<link rel="stylesheet" href="main.css" type="text/css" />
</head>

<body class="info">


<center>
<h2><A HREF="RSAT_home.cgi">RSAT</A> - <a href=get-orthologsi-compara_form.cgi>get-orthologs-compara</a> manual</h2>
</center>

<h1><a name="description">DESCRIPTION</a></h1>
<p>
Returns orthologues, plus optionally paralogues and homoeologues, for a set of genes in one or more organisms. Relies on primary data from Ensembl Compara.
</p>
<p>
</p>
<hr />
<h1><a name="authors">AUTHORS</a></h1>
<dl>
<dt><strong><a href="mailto:bcontreras@eead.csic.es">Bruno Contreras Moreira</a></strong><br />
</dt>
<dt><strong><a href="mailto:Jacques.van-Helden\@univ-amu.fr">Jacques van Helden </a></a></strong><br />
</dt>
</dl>

<hr />
<h1><a name="genes">INPUT FORMAT</a></h1>
<p>
Query genes can be directly entered or uploaded from an input file (option -i). The first word of each row of such a file is handled as a gene. Any additional information on the same row is ignored.
</p>

<h1><a name="type" name="subtype">HOMOLOGY TYPES</a></h1>
<p>
Supported homology types and subtypes: 
<ul>

<li><a name="orth"><b>orth</b></a><ul>Returns only orthologous genes, including many2many, one2many & one2one.</ul><br></li>

<li><a name="para"><b>para</b></a><ul>Returns only paralogues, including within_species_paralog, other_paralog & gene_split.</ul><br></li>

<li><a name="hom"><b>hom</b></a><ul>Returns only homoeologues, including many2many, one2many & one2one.</ul><br></li>
</ul>
</p>

<hr />
<h1><a name="return">OUTPUT FORMAT and RETURN FIELDS</a></h1>
<p>
A tab-separated file with six columns. Each row of the output file describes a homology relation between a query gene and a target gene. Output contains the following columns:
</p>
<ul>
<li><a name="target_id"><b>target_id</b></a><ul>ID of target homologous gene (Ensembl gene_stable_id).</ul><br>
</li>
<li><a name="ref_org"><b>ref_org</b></a><ul>Name of the reference organism.</ul><br>
</li>
<li><a name="type"><b>type</b></a><ul>Type of homology relationship.</ul><br>
</li>
<li><a name="query_id"><b>query_id</b></a><ul>ID of the query gene.</ul><br>
</li>
<li><a name="ident_target"><b>ident_target</b></a><ul>%protein sequence identity with respect to target length.</ul><br>
</li>
<li><a name="ident_query"><b>ident_query</b></a><ul>%protein sequence identity with respect to query length.</ul><br>
</li>
</ul>

<hr />
<h1><a name="orthology_criteria">HOMOLOGY CRITERIA</a></h1>
<p>
Ensembl Compara gene orthology and paralogy predictions are generated by a pipeline where maximum likelihood phylogenetic gene trees play a central role. They aim to represent the evolutionary history of gene families, i.e. genes that diverged from a common ancestor. These gene trees reconciled with their species tree have their internal nodes annotated to distinguish duplication or speciation events, and thus support the annotation of orthologous and paralagous genes, which can be part of complex one-to-many and many-to-many relations. 

[Adapted from: <a href="http://www.ensembl.org/info/genome/compara/homology_method.html">http://www.ensembl.org/info/genome/compara/homology_method.html</a>].
</p>
<hr />

</body>

</html>
