<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
<HEAD>
   <TITLE>RSA-tools - supported sequence formats</TITLE>
   <META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
   <META NAME="GENERATOR" CONTENT="Mozilla/3.01Gold (X11; I; SunOS 5.5.1 sun4u) [Netscape]">
</HEAD>
<BODY TEXT="#000000" BGCOLOR="#FFEEDD" LINK="#0000EE" VLINK="#55008B" ALINK="#FF0000">

<H3 ALIGN=CENTER><A HREF="RSAT_home.cgi">RSA-tools</A> - supported sequence formats</H3>
RSA-tools are dedicted to the analysis of regulatory sequences and thus only deals with DNA sequences. The following formats are supported. <P>

<UL>
<A NAME="raw"><LI><B>raw</B><BR></A>
The text contains a single DNA sequence without any comment or other text. The sequence should only contain the letters corresponding to nucleotides (A, C, G, T). Uppercases and lowercases are onsidered equivalent. Tabs (\t), blank spaces and newline characters (\n) are accepted (they will be automatically removed before analysis). <P>

<UL>
<B>raw format example (a single 60 bp sequences)</B>
<PRE>GCGGTGCCCGGCCCAGCCACATATATATAGGTGTGTGCCA
CTCCCGGCCCCGGTATTAGC</PRE>
</UL>

<A NAME="multi"><LI><B>multi</B><BR></A>
This is a variant of the raw format, allowing to include several sequences within the same text. As in the raw format, there are neither comments nor sequence identifiers. Tabs and white spaces are ignored. The difference is that each new line is considered to contain a distinct sequence.<P>

<UL>
<B>multi format example (two distinct sequences, 60 bp each)</B>
<PRE>GCGGTGCCCGGCCCAGCCACATATATATAGGTGTGTGCCACTCCCGGCCCCGGTATTAGC
CCCTTCCAGTTTCTTTTATTCTTTGCTGTTTCGAAGAATCACACCATCAATGAATAAATC</PRE>
</UL>

<A NAME="ig"><LI><B>IG</B><BR></A>
IntelliGenetics format. All lines beginning with a semicolon ";" are considered as comments. The first non-comment line contains the sequence identifier (a single word without spaces). The sequence follows on the next lines. It can include spaces, tabs or newlines, that will be ignored for sequence analysis. The end of one sequence is indicated by termination character: 1 for linear, 2 for circular sequences. Several sequences can be concatenated within the same text. <P>

<UL>
<B>IG format example</B>
<PRE>;sequence of the region upstream from gat1
;Locus GAT1
;ORF YFL021W  coord:    VI 95964 97496
;upstream region size:  60
;upstream region coord: VI 95904 95963
GAT1_up
GCGGTGCCCGGCCCAGCCACATATATATAGGTGTGTGCCA
CTCCCGGCCCCGGTATTAGC1
;sequence of the region upstream from put4
;Locus PUT4
;ORF YOR348C  coord:    XV 988773 986890
;upstream region size:  60
;upstream region coord: XV 988833 988774
PUT4_up
CCCTTCCAGTTTCTTTTATTCTTTGCTGTTTCGAAGAATC
ACACCATCAATGAATAAATC1</PRE>
</UL>

<A NAME="fasta"><LI><B>fasta</B><BR></A>
The program defined for the program FastA, and used by a wide variety of programs (FastA, BLAST, gibbs sampler, ...). The first character of the first line is a ">". It is immediately followed by the sequence identifier, a single word. Further words on the same line are considered as comments. The sequence begins on the next line, and follows until the end of file is reached or another line is found which begins with a ">". White spaces, tabulations and newline characters can be inserted within the sequence, and will be ignored for analysis. Several sequence can be concatenated within the same text. <P>

<UL>
<B>FastA format example (2 sequences, 60 bp each)</B>
<PRE>>GAT1_up
GCGGTGCCCGGCCCAGCCACATATATATAGGTGTGTGCCA
CTCCCGGCCCCGGTATTAGC
>PUT4_up
CCCTTCCAGTTTCTTTTATTCTTTGCTGTTTCGAAGAATC
ACACCATCAATGAATAAATC</PRE>
</UL>


<A NAME="wconsensus"><LI><B>wconsensus</B><BR></A>
This is the format defined by Jerry Hertz for his programs patser, consensus and wconsensus. Lines beginning with a semiolon ";" or a "#" are considered as comments. The first word of the first non-comment sequence is the sequence identifier. The sequence follows, embraced in backslashes (\). <P>

<UL>
<B>Wconsensus format example (2 sequences, 60 bp each)</B>
<PRE>
;sequence of the region upstream from gat1
;Locus GAT1
;ORF YFL021W  coord:    VI 95964 97496
;upstream region size:  60
;upstream region coord: VI 95904 95963
GAT1_up \GCGGTGCCCGGCCCAGCCACATATATATAGGTGTGTGCCACTCCCGGCCCCGGTATTAGC\
;sequence of the region upstream from put4
;Locus PUT4
;ORF YOR348C  coord:    XV 988773 986890
;upstream region size:  60
;upstream region coord: XV 988833 988774
PUT4_up \CCCTTCCAGTTTCTTTTATTCTTTGCTGTTTCGAAGAATCACACCATCAATGAATAAATC\
</PRE>
</UL>

</UL>
Most formats (except raw) allow to include several sequences within the same text. This is convenient for analyzing a family of sequences (oligo analysis, pattern search). <P>
The program <A HREF="convert-seq.html"><B>convert sequence</B></A> performs interconversions between these sequence formats. <P>

</BODY>
</HTML>
