<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
<HEAD>
   <TITLE>RSA-tools: upstream-region manual</TITLE>
</HEAD>
<BODY TEXT="#000000" BGCOLOR="#FFEEDD" LINK="#0000EE" VLINK="#55008B" ALINK="#FF0000">

<H3 ALIGN=CENTER><A HREF="RSA_home.cgi">RSA-tools</A> - upstream region manual</H3>

<H4>Name</H4>
<UL>
upstream-region<P>
1997-98 by <a HREF="mailto:Jacques van Helden <jvanheld@ucmb.ulb.ac.be>">Jacques van Helden (jvanheld@ucmb.ulb.ac.be)</A>.
</UL>

<H4>Description</H4>

<UL>
Returns the upstream sequence for one or several genes. <P>
</UL>

<H4>Options</H4>

<UL>
<A NAME="genes"></A>
<B>Genes</B><BR>
Valid queries are ORF identifiers (eg: YFL021W) or gene names (eg: GAL4, NIL1). Synonyms are accepted (eg: NIL1 = GAT1), but only one gene name is returned as sequence identifier (GAT1 in this case). <BR>
Several queries can be entered simultaneously, separated by carriage returns. The first word of each line is the query; any following information is ignored. <P>

<A NAME="from_to"></A>
<B>From, to:</B><BR>
Limits of the region to retrieve. Coordintates are calculated relative to the ORF start. 
<UL>
<B>Sign</B> 
  <UL>
    <LI><B><FONT COLOR="#DD0000">negative</FONT></B> 
           values return <B><FONT COLOR="#DD0000">upstream</FONT></B> sequences
    <LI><B><FONT COLOR="#00DD00">positive</FONT></B> 
           values return sequences <B><FONT COLOR="#00DD00">downstream</FONT></B> from the start codon (e.g. from 0 to +2 returns the start codon itself)<P>
  </UL>

<B>Suggested values</B>
<UL>
<LI>For <B>yeast</B>, we generally obtain good results with upstream regions from -800 to -1. About 99% of the known upstream elements are comprized between these limits (source: Transfac).<BR>
<LI>For <B>bacteria</B>, the distribution of regulatory sites depends on the mode of regulation : 
<UL>
<LI><B>transcriptional repressors</B> generally bind proximally, and may overlap the transcription start or even be located downstream. A good guess is from -200 to +50. 
<LI>Binding sites for <B>transcriptional activators</B> have a more distal distribution (-400 to -1). <P> 
</UL>
</UL>
</UL>

<A NAME="noorf"></A>
<B>Allow overlap with upstream ORFs:</B><BR>
It is quite frequent to find a predicted ORF in close proximity upstream from a query ORF. If you want to discard these sequences from your analysis, inactivate this option, and the size of the upstream region will automatically be shortened when a predicted ORF is located within the range defined by the option <I>from</I>. The actual size selected for the upstream sequence is indicated in the sequence comments if you select the IG sequence format. <P>
Note that in some cases a known regulatory element is located upstream or within a predicted ORF. This means either that the predited ORF is an artifact, or that the same sequence is bifunctional (coding and regulatory). For this reason, this option is activated by default. <P>
Note that the distance to the closest upstream neighbour can be checked with the program <A HREF="neighbour-orfs.html"><B>neighbour-orfs</B></A>.
<P>

<A NAME="format"></A>
<B>Output sequence format:</B><BR>
The result can be displayed in various sequence formats (click on the links for more details). 
<UL>
<LI><A HREF="help.formats.html#raw"><B>raw</B></A>: the raw sequence without any identifier or comment.
<LI><A HREF="help.formats.html#multi"><B>multi</B></A>: several raw sequences concatenated.
<LI><A HREF="help.formats.html#ig"><B>IG</B></A>: IntelliGenetics format.
<LI><A HREF="help.formats.html#fasta"><B>FastA</B></A>: the sequence format used by FastA, BLAST, Gibbs sampler and a lot of other bioinformatic programs.
<LI><A HREF="help.formats.html#wconsensus"><B>Wconsensus</B></A>: the format defined by Jerry Hertz for his programs (patser, consensus, wconsensus).
</UL>
</UL>


<H4>Availability</H4>
<UL>
The program can be used through its web interface at:
<A HREF="http://copan.cifn.unam.mx/Computational_Biology/yeast-tools">http://copan.cifn.unam.mx/Computational_Biology/yeast-tools</A> <BR>
upstream-region is a perl script running on unix machines (SUN, SGI have been tested). The web interface is a perl-cgi script.
</UL>

<HR SIZE=3>

</BODY>
</HTML>
