<html>
<title>RSA-tools - Wargnings</title>
<body bgcolor='#ffffff'>
<blockquote>

<h1 align=center>Warning</h1>

<a name=human>
<h2>Problems for pattern discovery with mammalian genomes</h2>


<p>
Three mammalian genomes are currrently suported on RSAT: Homo sapiens,
Mus musculus and Rattus norvegicus.


<P>
Pattern discovery programs (e.g. oligo-analysis and dyad-analysis), it
have been proven to be highly efficient with the yeast genome, and
relatively efficient with A.thaliana and Drosophila melanogaster
melanogaster. 

<P>
My first tests with human and mouse genes suggest that with
mammalian genomes these programs return many false positive (i.e. if
you submit a random set of genes, you always get plenty of highly
'significant' motifs). This is likely to come from the heterogeneity
of human sequences, and I am currently working on an improvement of
the background model.

<P>
Consensus and the gibbs sampler are also likely to return many false
positives with human sequences, since they generally return more false
positives than oligo-analysis and dyad-analysis. I suspect that the
same problem will be encountered by most currently existing pattern
discovery programs : if you submit a random selection of human
promoters, they will always return plenty of motifs, although the
correct answer would be to return no motif (since these genes are not
likely to be co-regulated).  If you know any counter example, please
indicate it to me.

<p>
I am currently working on methodological improvements to model
heterogeneous sequences. This is an on-going research, and it may take
several months. Inbetween, I installed the genomes because you can
already run some tools (pattern matching for example), but I do not
recommend to trust pattern discovery results with mammalian genomes.

<P>
<a href=mailto:jvanheld@scmbb.ulb.ac.be>Jacques van Helden</a>

</blockquote>
</body>
</html>