<html>
<head>
   <meta HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
   <title>The analysis of regulatory sequences - course</title>
   <link rel="stylesheet" type="text/css" href = "course.css" media="screen">
</head>
<body class="info">

<h1 align=center>The Analysis of Regulatory Sequences</h1>



<a name=site><h3>Warning</h3>

This web site contains support material for the course on the Analysis
of Regulatory Sequences. 


<a name=site><h3>Site</h3>

<a target=_top href='http://rsat.bigre.ulb.ac.be/rsat/course/'>http://rsat.bigre.ulb.ac.be/rsat/course/</a>

<a name=author><h3>Author</h3>

<b>Jacques van Helden</b>
<br><a target=_top href='http://www.bigre.ulb.ac.be/Users/jvanheld/'>http://www.bigre.ulb.ac.be/Users/jvanheld/</a>
<br><a href='mailto:jvhelden@ulb.ac.be'>jvhelden@ulb.ac.be</a>

<a name=chapter><h3>Book chapter</h3>

Pre-print version of a book chapter for a course given in Les Houches
(2004).

<ul>
<a href='pdf_files/vanHelden_LesHouches_color.pdf'>The Analysis of Regulatory Sequences</a>
</ul>


<a name=slides><h3>Slides</h3>


<table bgcolor='#bbbbff' cellpadding=3 cellspacing=3>

<tr><td colspan=2><b><font face=Arial size=-1>
<a href='pdf_files/01.2_regulatory_sequences_intro.pdf'>
Transcriptional regulation
</a></font></b></td></tr>

<tr><td colspan=2><font face=Arial size=-1>
<a href='pdf_files/01.3.sequence_models.pdf'>
Sequence models
</a></font></td></tr>

<tr><td colspan=2><b><font face=Arial size=-1>
<a href='pdf_files/01.4.PSSM_theory.pdf'>
Position-specific scoring matrices
</a></font></b></td></tr>

<tr><td colspan=2><b><font face=Arial size=-1>
<a href='pdf_files/01.5.transcription_factor_databases.pdf'>
Transcription factor databases
</a></font></b></td></tr>

<tr><td colspan=2><b><font face=Arial size=-1>
<a href='pdf_files/02.1.pattern_matching_intro.pdf'>
Pattern matching
</a></font></b></td></tr>

<tr><td>&nbsp;&nbsp;</td><td><font face=Arial size=-1>
<a href='pdf_files/02.2.pattern_matching_strings.pdf'>
String-based pattern matching
</a></font></td></tr>

<tr><td>&nbsp;&nbsp;</td><td><font face=Arial size=-1>
<a href='pdf_files/02.3.pattern_matching_matrices.pdf'>
Matrix-based pattern matching
</a></font></td></tr>

<tr><td colspan=2><b><font face=Arial size=-1>
Pattern discovery
</font></b></td></tr><tr>

<tr><td>&nbsp;&nbsp;</td><td><font face=Arial size=-1>
<a href='pdf_files/03.1.pattern_discovery_intro.pdf'>
Introduction
</a></font></td></tr>

<tr><td>&nbsp;&nbsp;</td><td><font face=Arial size=-1>
<a href='pdf_files/03.2.1.pattern_discovery_strings.pdf'>
String-based pattern discovery
</a></font></td></tr>

<tr><td>&nbsp;&nbsp;</td><td><font face=Arial size=-1>
<a href='pdf_files/03.2.3.word_count_statistics.pdf'>
Word count statistics
</a></font></td></tr>

<tr><td>&nbsp;&nbsp;</td><td><font face=Arial size=-1>
<a href='pdf_files/03.3.pattern_discovery_matrices.pdf'>
Matrix-based pattern discovery
</a></font></td></tr>

<tr><td>&nbsp;&nbsp;</td><td><font face=Arial size=-1>
<a href='pdf_files/03.5.genome_scale_discovery.pdf'>
Genome-scale pattern discovery
</a></font></td></tr>

<tr><td colspan=2><b><font face=Arial size=-1>
Evaluation of pattern predictions
</font></b></td></tr>

<tr><td>&nbsp;&nbsp;</td><td><font face=Arial size=-1>
<a href='pdf_files/04.pattern_comparisons.pdf'>
Pattern comparison
</a>
</font></b></td></tr>

<tr><td>&nbsp;&nbsp;</td><td><font face=Arial size=-1>
<a href='pdf_files/05.pattern_prediction_eval.pdf'>
Pattern matching
</a></font></td></tr>

<tr><td>&nbsp;&nbsp;</td><td><font face=Arial size=-1>
<a href='pdf_files/05.pattern_prediction_eval.pdf'>
Pattern discovery
</a></font></td></tr>

<tr><td colspan=2><b><font face=Arial size=-1>
<a href='pdf_files/07.gene_classification.pdf'>Gene classification on the basis of upstream motifs</a>
</font></b></td></tr>

<tr><td colspan=2><b><font face=Arial size=-1>
Analysis of gene expression data
</font></b></td></tr>

<tr><td>&nbsp;&nbsp;</td><td><font face=Arial size=-1>
Introduction to DNA chips
</font></td></tr>

<tr><td>&nbsp;&nbsp;</td><td><font face=Arial size=-1>
Analysis of regulatory regions from DNA chip clusters
</font></td></tr>

<tr><td colspan=2><b><font face=Arial size=-1>
<a href='pdf_files/03.4.2.comparative_genomics.pdf'>
Comparative genomics
</a></font></b></td></tr>

<tr><td colspan=2><b><font face=Arial size=-1>
<a href='pdf_files/03.4.3.footpint_discovery.pdf'>
Discovering phylogenetic footprints in bacterial genomes
</a></font></b></td></tr>

<tr><td colspan=2><b><font face=Arial size=-1>
Inference of clusters of functionally related genes
</font></b></td></tr>

<tr><td colspan=2><b><font face=Arial size=-1>
<a href='pdf_files/acknowledgements.pdf'>
Acknowledgements
</a></font></b></td></tr>

<tr><td colspan=2><b><font face=Arial size=-1>
<a href='pdf_files/RSAT_short.pdf'>
Short presentation
</a></font></b></td></tr>

</table>


</body>
</html>
