<html>
  <head>
    <title>RSAT tutorial - First steps in the unix shell</title>
    <link rel="stylesheet" type="text/css" href = "../../../course.css" media="screen">

    <script type="text/javascript" src="../syntaxhighlight/shCore.js"></script>
    <script type="text/javascript" src="../syntaxhighlight/shBrushBash.js"></script>
    <script type="text/javascript" src="../syntaxhighlight/shBrushR.js"></script>
    <link type="text/css" rel="stylesheet" href="../syntaxhighlight/shCore.css"/>
    <link type="text/css" rel="stylesheet" href="../syntaxhighlight/shThemeDefault.css"/>
    <script type="text/javascript">
      SyntaxHighlighter.config.clipboardSwf = 'syntaxhighlight/clipboard.swf';
      SyntaxHighlighter.all();
      
    </script>
    
    
	<script language="JavaScript" type="application/javascript">
	  <!--
	      function sizeTbl2(h,i) {
	      var tbl = document.getElementById(i);
	      tbl.style.display = h;
	      }
	      // -->
	</script> 
	
  </head>
  <body class="info">

    <h1>RSAT tutorial - First steps in the unix shell</h1>
    
    <blockquote>
      <p align="right"><i>Less does more or less the same as more, but rather more than less.</i>
	<br>Jacques van Helden and Denis Puthier, June 24, 2014
      </p>
      
      <!--      <p align="right"><i>Time flies like an arrow, fruit flies like a banana.</i>
		<br>Uncertain original author (Groucho Marx?)
      </p>
      -->
    </blockquote>

    <a name='contents'></a>
    <h2>Contents</h2>
    <ul>
      <li><a href='#session'>Connection</a></li>
      <li><a href='#mkdir'>Creating a working directory</a></li>
      <li><a href="#ens" >Retrieving species list from Ensembl</a></li>
      <li><a href="#ins" >Inspecting the result</a></li>
      <li><a href='#exo'>Retrieving gene models</a></li>
    </ul>

    <hr>
    
    <h2>Introduction</h2>
    
    <p>In this tutorial, we will address the situation where we want
      to connect to a Unix server and use some command lines to
      performed specific tasks.</p>

    <p>This series of exercises aims at providing you with the basics
      of the Unix <b><i>Shell</i></b> environment. We will open a
      programmatic connection to the EnsemblGenomesq server and start
      to create a bioinformatics worflow.</p>
    
    <hr>		 
 
    <a name='session'></a>
    <h2>Connection</h2>
    <h3>Starting a terminal under Windows</h3>
    <p>
      <div class='protocol'>
	
	<ul>
	  <li>Start the <span class="program">putty</span>
	    software. You should get a windows as shown below.</li>
	  <li>Type <b>139.124.66.4</b> in the
	    option <span class='option'>Host Name (or IP
	      adress)</span>. </li>
	  <li>Press <b>Open</b>.</li>
	  <li>You will be prompted for a login. Enter your login (you
	    should have received one by mail). Press
	    the <span class="option">Return</span> key from your
	    keyboard.</li>
	  <li>You will then be prompted for a password, and have to
	    confirm that you trust the host key of the server. Accept
	    it by typing <b>'Y'</b>. </li>
	</ul>    
	
    	<center><img src='images/putty.png'></center>

      </div>
      
      <h3>Starting a terminal under Macintosh</h3>
      
      <div class='protocol'>
	
	<ul>
	  <li>Search for Applications > Utilities > Terminal.</li> 
	  <li>In the open window type the following command <b>changing the 'login' by the one you received by mail</b>:</li>
      	  <pre class="brush:bash;">	
	    ssh login@139.124.66.4
	  </pre>
	  <li>You will be asked for a password. Enter your  password.</li>
	  <li>you will be  prompted to trust the host key of the server. Accept it by typing <b>'Y'</b>. </li>
	</ul>    
	
	
      </div>
      
      <hr>
      <a name='mkdir'></a>

      <h2>Creating a working directory</h2>
    <p> After connection you will be able to work the remote server
      (whose name is "pedagogix"). To tell the server what you want it
      to do, you will need to write <b>instructions</b>. The basic
      blocks of these instructions will be <b>Unix commands</b> that are
      small programs dedicated to specific tasks.</p>

    <p>Here we will instruct the server to display our location in the
      file tree then to create a directory<p>
      
      <div class='protocol'>
	<ul>
	  
	  <li>Where are we located ? We can get this information using
	    the <b>pwd</b> directory
	    command. (<b>p</b>rint <b>w</b>orking <b>d</b>irectory)</li>

<pre class="brush:bash;">	
# check the current working directory
pwd 
</pre>

	  <li>Using the <i>mkdir command</i>
	    (<b>m</b>a<b>k</b>e <b>dir</b>ectory) create a working
	    directory named <i>ensembl_data</i> in your HOME folder. Use
	    the <b>ls</b> (<b>l</b>ist) command to get the list of file
	    and directories available in the current working
	    directory.</li>
      	  <pre class="brush:bash;">	
# create a directory
mkdir ensembl_data
# Get the list of available file and directories
ls
</pre>
		
	</ul>
      </div>
      
      <hr>
      <a name='ens'></a>
      
      <h2>Retrieving species list from Ensembl</h2>
      
      <h3>Getting information about ensembl_genomes</h3>
    <p>In the context of this training, we have developed a small
      python program (<b>ensembl_genomes</b>) that allows one to
      connect to the <b>Ensembl Genomes REST API</b>. This program
      fetches species-related data from ensemblGenomes DataBase, such
      as the list of supported species, or their annotations (genes,
      transcripts, proteins). Users can get informations about
      ensembl_genomes by using the <b>-h</b> argument
      (-<b>h</b>elp). At the moment, ensembl_genome implements two
      tasks: <b>retrieve_species</b>
      and <b>retrieve_features</b>. Users can get informations about
      retrieve_species and retrieve_features by using the <b>-h</b>
      argument. Note that ensembl_genomes is still under
      developement.</p>
    
    <div class='protocol'>
      <ol>
	
	
	
      	<pre class="brush:bash;">
# get information about the program
ensembl_genomes -h

# get information about the retrieve_species implemented task
ensembl_genomes retrieve_species -h

# get information about the retrieve_features implemented task
ensembl_genomes retrieve_features -h
        </pre>
	
    </div>
    
    
    <h3>Retrieving species information using ensembl_genomes</h3>
    
    <p>We will use retrieve_species from ensembl_genomes to get the
      list of available genomes. Below are the available arguments. Most
      generally, command arguments are used to define the input and
      output file or to modify the way the program works.</p>
    
<pre>
usage: ensembl_genomes.py retrieve_species [-h] [-o OUTPUT] [-v {1,2,3}]
                                           [-f FILE]

optional arguments:
  -h, --help            show this help message and exit
  -o OUTPUT, --output OUTPUT
                        Choose the output file name containing the
                        informations about all species
  -v {1,2,3}, --verbosity {1,2,3}
                        Choose the verbosity level
                        default: None 
                        1: few informations
                        2: detailed informations
                        3: complete
                         
  -f FILE, --file FILE  If enabled, this option will ensure the creation of a
                        separate file containing only species names (one
                        species per line).
</pre>
	
    <p> We will instruct the program to retrieve the species list
      (including for each of them division, name, taxon_id, release,
      display_name, common_name,...) but also to store in a separate
      file the list of species as a single column.</p>
    
    <pre class="brush:bash;">	
## Run the task "retrieve_species"
ensembl_genomes retrieve_species -v 1 -o ensembl_data/species_descriptions.tab -f ensembl_data/species_list.txt
ls -ltr ensembl_data ## should contain 2 files
</pre>


    <p><b>NB:</b> argument for the ls command: <b>-l</b> (<b>l</b>ong, get lots of information about file</p> 
    <p><b>NB:</b> argument for the ls command: <b>-t</b> (sort by modification <b>t</b>ime, newest first)</p> 
    <p><b>NB:</b> argument for the ls command: <b>-r</b> (<b>r</b>everse order while sorting)</p> 
    
    <hr>
    
    <a name='ins'></a>
    
    <h2>Inspecting the result</h2>
    
    <p> Unix-like system (Linux, Mac OSX, ...) propose lots of useful
      commands to work with files. We will see some of these basic
      commands that can be really helpful especially when one is
      working with large files.</p>
    
    <p>First we will go into the ensembl_data directory using the cd
    command (<b>c</b>hange <b>d</b>irectory).</p>

    <pre class="brush:bash;">	
# changing directory
cd ensembl_data

# print the path of the current working directory
pwd

# go up one level in the file tree (that is go back to the previous directory)
cd ..
pwd
        </pre>

    <p>Count the number of supported species at EnsemblGenomes
      (<i>i.e</i>; the number of line in the species_list.txt or
      species_descriptions.tab files)</p>

      	<pre class="brush:bash;">	
## Count the number of supported species at EnsemblGenomes
wc -l ensembl_data/species_list.txt
## Today: 11,169 species !
        </pre>

<p>Display the first n rows of a file.<p>
      	<pre class="brush:bash;">	
## Check the first 10 rows of the file
head -n 10 ensembl_data/species_list.txt
        </pre>

<p>Display the lst n rows of a file.<p>
      	<pre class="brush:bash;">	
## Check the last 20 rows of the file
tail -n 20 ensembl_data/species_list.txt
        </pre>



<p>Create a new file containing the species names in alphabetical order.</p>
      	<pre class="brush:bash;">	
## Species names are retrieved in an arbitrary order. Let us sort them alphabetically
sort ensembl_data/species_list.txt > ensembl_data/species_list_sorted.txt
## The ">" character redirects the output to a file (default to screen)
        ## Check species names are well sorted
head ensembl_data/species_list_sorted.txt
tail ensembl_data/species_list_sorted.txt
</pre>



<p>Let's have a look at the species_list_sorted.txt file using a pager.</p>

      	<pre class="brush:bash;">	
## Inspect the sorte species list in a page-per-page mode
more ensembl_data/species_list_sorted.txt
## Use 'enter' key to navigate
## Use 'q' key to quit
</pre>

<pre class="brush:bash;">
## "Less does more or less the same as more ... but rather more than less"
## A thought from JVH early this morning. :) DP
less ensembl_data/species_list_sorted.txt
## Use vertical arrow keys to navigate
## Use 'q' key to quit
</pre>



<p>Let's inspect the species description table. How many rows does it contains (<b>w</b>ord count <b>-l (lines)</b>. Let's have a look at the file using less.</p>
	<pre class="brush:bash;">
	## Inspect the species description table
	## How many rows
	wc -l ensembl_data/species_descriptions.tab
	## Inspect the content using less
	less ensembl_data/species_descriptions.tab # q to quit
	</pre>

<p>Cut columns 2,7 and 8 (name, Taxonomy ID, release number) and send the text stream to the 'less' command. Note that the pipe ("|") characters allows one to send the result of one command to another command.</p>
	<pre class="brush:bash;">
	## Cut the columns 2,7 and 8
	cut -f 2,3,5  ensembl_data/species_descriptions.tab | less ## select some columns of interest
	</pre>


<p>Cut the column 1 (division), then sort it, then get the list of non redundant character strings (<b>uniq</b>) and <b>c</b>ount their occurence (<b>-c</b>).</p>
	<pre class="brush:bash;">
	## The number of species in each division
	cut -f1   ensembl_data/species_descriptions.tab | sort | uniq -c 
	</pre>



	<hr>
    <a name='exo'></a>

	<h2>Retrieving gene models</h2>

	<div class="exo">
	  <p>Now that we have introduced the basics about command
	    lines, use them to get information about the gene models of
	    your favorite organism.</p>

	  <ul>
	    <li>Use the "retrieve_features" task from the
	      ensembl_genomes program to get the list of gene
	      models.</li>
	    <li>How many genes models are known in your choosen
	      organism ?</li>
	    <li>How many coding genes ?.</li>
	    <li>How many "hypothetical protein" ?.</li>
	  </ul>
	  <div>
	    
	    
	    
	    <a href="javascript:sizeTbl2('block','sol_mk')">View solution</a>| 
	    <a href="javascript:sizeTbl2('none','sol_mk')">Hide solution</a>
	    <div id=sol_mk  style="overflow:hidden;display:none">
	      <h4>Solution</h4>
	      
<pre class="brush:bash;">
## Retrieve gene models
ensembl_genomes retrieve_features -org campylobacter_jejuni_subsp_jejuni_bh_01_0142 --outputdir ensembl_data/ -v 1 -d 1

## How many genes
wc -l  ensembl_data/campylobacter_jejuni_subsp_jejuni_bh_01_0142/genes.tab # -1, as there is a header

## How many coding genes 
cut -f7 ensembl_data/campylobacter_jejuni_subsp_jejuni_bh_01_0142/genes.tab  | sort | uniq -c

## How many "hypothetical protein"
cut -f8  ensembl_data/campylobacter_jejuni_subsp_jejuni_bh_01_0142/genes.tab | grep "hypothetical protein" | wc -l
</pre>
	    </div>
	    
	    <hr>
  </body>
</html>

