<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" dir="ltr">

<head>
	<meta http-equiv="content-type" content="text/html; charset=UTF-8"/>
	<meta name="description" content=""/>
	<meta name="keywords" content="" />
	<meta name="author" content="" />
	<title>ECCB 2014 T01</title>
	<link rel="stylesheet" type="text/css" href="style.css" media="screen" />
<style type="text/css">

#wrapper{
  margin:0 auto;
  #padding:15px 15% 8em;
  text-align:left
}
#content {
  max-width:70em;
  width:100%;
  margin:0 auto;
  padding-bottom:20px;
  overflow:hidden
}
.demo {
  margin:1.5em 0;
  padding:1.5em 1.5em 0.75em;
  border:1px solid #ccc;
  position:relative;
  overflow:hidden
}

.post {
  position:relative;
  overflow:hidden
}
.collapse p {padding:0 10px 1em}

.switch {position:absolute; top:1.5em; right: 1.5em; padding:3px}

#.post .switch {position:static; text-align:right}

.post .main{margin-bottom:0; padding-bottom:0}

.other li, .summary {margin-bottom:.3em; padding:1em; border:1px solid #e8e7e8; background-color:#f8f7f8}

.other ul {list-style-type:none; text-align:center}


.expand{padding-bottom:.75em}

/* --- Links  --- */

#download {
  border-style:none;
  background:white;
}

.expand a {
  display:block;
  padding:3px 10px
}
.expand a:link, .expand a:visited {
  border-width:1px;
  background-image:url(img/arrow-down.gif);
  background-repeat:no-repeat;
  background-position:98% 50%;
}
.expand a:hover, .expand a:active, .expand a:focus {
}
.expand a.open:link, .expand a.open:visited {
  #border-style:solid;
  #background:#eee url(img/arrow-up.gif) no-repeat 98% 50%
  color:black;
}
</style>
<!--[if lte IE 6]>
<style type="text/css">
h3 a, .demo {position:relative; height:1%}
</style>
<![endif]-->

<!--[if lte IE 6]>
<script type="text/javascript">
   try { document.execCommand( "BackgroundImageCache", false, true); } catch(e) {};
</script>
<![endif]-->
<!--[if !lt IE 6]><!-->
<script type="text/javascript" src="http://ajax.googleapis.com/ajax/libs/jquery/1.4.2/jquery.min.js"></script>
<script type="text/javascript" src="scripts/expand.js"></script>
<script type="text/javascript">
<!--//--><![CDATA[//><!--
$(function() {
    $("#content h3.expand").toggler();
    $("#content h2.expand").toggler();
    $("#content div.demo").expandAll({trigger: "h3.expand", ref: "h3.expand"});
    $("#content div.other").expandAll({
      expTxt : "[Show]", 
      cllpsTxt : "[Hide]",
      ref : "ul.collapse",
      showMethod : "show",
      hideMethod : "hide"
    });
    $("#content div.post").expandAll({
      expTxt : "[Show tip]", 
      cllpsTxt : "[Hide tip]",
      ref : "div.collapse", 
      localLinks: "p.top a"    
    });    
});
//--><!]]>
</script>
<!--<![endif]-->


    <script type="text/javascript" src="syntaxhighlight/shCore.js"></script>
    <script type="text/javascript" src="syntaxhighlight/shBrushBash.js"></script>
    <script type="text/javascript" src="syntaxhighlight/shBrushR.js"></script>
    <link type="text/css" rel="stylesheet" href="syntaxhighlight/shCore.css"/>
    <link type="text/css" rel="stylesheet" href="syntaxhighlight/shThemeDefault.css"/>
    
    <script type="text/javascript">
      SyntaxHighlighter.config.clipboardSwf = 'syntaxhighlight/clipboard.swf';
      SyntaxHighlighter.all();
      
    </script>

</head>

<body>

  <div id="site-wrapper">

    <div id="header">

      <div id="top">

	<div class="left" id="logo">
	  <a href="#"><h2 class="label label-green">Session 2 : Command-line utilisation of the tools</h2></a>
	</div>



	<div class="clearer">&nbsp;</div>

      </div>

      <div class="navigation" id="sub-nav">

	<ul class="tabbed">
	  <li ><a href="index.html">Home</a></li>
	  <li class="current-tab"><a href="session1.html">Session1:Web site</a></li>
	  <li><a href="session2.html">Session2:Command line</a></li>
	  <li><a href="session3.html">Session3:SOAP Web services</a></li>
	</ul>

	<div class="clearer">&nbsp;</div>

      </div>

    </div>

    <div class="main" id="main-two-columns">

      <div class="left" id="main-content">

	<h2 id="contents"> Introduction </h2>
	<b>Goal</b><br/>

	<u>The aim is to :</u>
	<ul>
	  <li></li>
	</ul>
	
	<div id="wrapper"> 
	  <div id="content"> 
	    
	    <!-- SEQUENCES -->	 
            <div class="demo">
              <h2 id="download" class="expand">Retrieving sequences from your peaks</h2>
              <div class="collapse">
		<div class="notice">
		  <b>Goal:</b> Given a set of peaks from a ChIP-seq experiment in a  <a href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1" target='_blank'>bed format</a>, retrieve the sequences corresponding to those coordinates from the genome in fasta format.
		  <br/>
		</div>
		<h3 class="expand">1 - Example dataset1: CEBPa binding regions in dog liver </h3>
		<div class="collapse">
                  <p>Schmidt, Wilson and Ballester published a ChIP-seq experiment on liver tissue to identify binding regions for the transcription factor CEBPa <a href="http://www.sciencemag.org/content/328/5981/1036" target='_blank'>(PMID:20378774)</a> in five different species (human, mouse, dog, short-tailed opossum and chicken). This data set is publicly available through arrayexpess<a href="http://www.ebi.ac.uk/arrayexpress/experiments/E-TABM-722/" target='_blank'> (E-TABM-722)</a>.<br/>
		    As done by the authors, CEBPa binding regions (peaks) were called using SWEMBL with parameter R=0.005, merged reads from two biological replicates and their corresponding input controls. For this tutorial we will analyze CEBPa binding pattern in dog, peaks can be downloaded from <a href="../data/peaks/do61+do79_cfam_CEBPA_liver.bed.SWEMBL.3.3.bed">here</a>.<br/> 
		  </div>
		  
		  <h3 class="expand">2 - Fetch sequences from a bed file </h3>
		  <div class="collapse">
                    <ol>
		      <li>In a web browser window open the <a href="http://rsat.ulb.ac.be/">RSAT</a> web page</li>
		      <li>In the menu (left side) click on the NGS-ChIP-seq drop down menu, and select the tool: fetch-sequences from UCSC.</li>
		      <li>Select the genome of interest, in this case: canFam2.</li>
		      <li>There are several options to input a bed file: Paste the coordinates, input from a URL, and upload the file from your computer. In this case, we will use this URL   <a href="http://rsat.ulb.ac.be/dataset_cepa_peaks">http://rsat.ulb.ac.be/dataset_cepa_peaks</a></li>
		      <li>Introduce your email address to receive a mail once the job is done. </li>
		      <li>Click on Go once the form is complete.</li>
		    </br>
                    <img class="bordered" src="./figures/Screen1_c.png" alt="fetch_seq_SC1" width="550"/>
		  </br>
		  <li>When the tool is finished you will receive a link to the fasta file containing the sequences corresponding to the coordinates in the bed file.</li>
		</ol>
		
		<div class="post">
		  <div class="success"> <b>Selecting the correct genome version:</b> Genomes are constantly updated with the improvement of sequencing technologies, alignment tools and annotations.  Always verify you are selecting the correct version.<br/><br/>
		  </div>
		</div>
		
		<div class="error">At this point, you should have the URL link to the fasta file</div>
		
              </div>         
	      
            </div>         
          </div>
          
          
          
	  <!-- PEAK-MOTIFS-->	     
          
          
          <div class="demo">
            <h2 id="quality" class="expand" >Discovering motifs from peak sequences</h2>
            <div class="collapse">
              <div class="notice">
		<b>Goal:</b>Discover binding motifs or patterns from a fasta file containing ChIP-seq determined binding regions of a transcription factor. 
		<br/>
	      </div>
              <h3 class="expand">1 - Getting to know peak-motifs</h3>
              <div class="collapse">
		<i>peak-motifs</i> is a computational pipeline that discovers motifs in peak sequences, compares them with databases, exports putative binding sites for visualization in the UCSC genome browser and generates an extensive report. </br>
	      In this section we will get familiar with this tool and its general usage.</br>
            <ol>
              <li></li>
              
              
            </ol>
          </div>
          <h3 class="expand">2 - Fine-tuning peak-motifs parameters </h3>
          <div class="collapse">
            
          </div>
        </div>
      </div>
      
      
      <!--Visualisation -->	

      <div class="demo">
        <h2 id="mapping" class="expand">Visualizing the sites in the context of genome annotations</h2>
        <div class="collapse">
          <div class="notice">
	    <b>Goal:</b>
	    <br/>
	  </div>
          <h3 class="expand">1 </h3>
          <div class="collapse">
            
          </div>
          
          <h3 class="expand">2 - </h3>
          <div class="collapse">
            Now is the time to gather your courage and face the terminal. Don't worry, it's not as scary as it seems !
            
          </div>
          <h3 class="expand">3 </h3>
          <div class="collapse">
            
          </div>
          
        </div>
      </div>
      

      

      <!-- Epigenetic marks -->
      
      <div class="demo">
        <h2 id="peak" class="expand">Discovering motifs in epigenetic marks</h2>
        <div class="collapse">
          <div class="notice">
	    <b>Goal:</b> Define the peaks, i.e. the region with a high density of reads, where the studied factor was bound
	    <br/>
	  </div>
          <h3 class="expand">1 - </h3>
          <div class="collapse">
            

          </div>
          <h3 class="expand">2 -  </h3>
          <div class="collapse">
            
          </div>
          <h3 class="expand">3 -  </h3>
          <div class="collapse">
            
            
            
   	  </div>
   	  
        </div>
      </div>
      
      <!-- ENRICHMENT -->
      <div class="demo">
        <h2 id="vizu" class="expand">Measure the enrichment of your peak for the expected motif</h2>
        <div class="collapse">
          <div class="notice">
	    <b>Goal:</b> 
	    <br/>
	  </div>
          <h3 class="expand">1 -</h3>
          <div class="collapse">
            

          </div>
          <h3 class="expand">2 -  </h3>
          <div class="collapse">
            
          </div>
   	  
	  
	  
	  
	</div>	
      </div>
    </div>
    
  </div>
</div>

<!-- SIDE BAR -->		

<div class="right sidebar" id="sidebar">

  <h5>Slides:</h5>
  <ul>
    <li><a href="booklet/booklet_chip-seq.pdf" target='_blank'>Presentation</a></li>
  </ul>
  <h5>Dataset:</h5>
  Schmidt,Wilson and Ballester <i>Five-vertebrate ChIP-seq reveals the evolutionary dynamics of transcription factor binding.</i> Science, 328(5981), 1036–1040. 2010 <a href="http://www.sciencemag.org/content/328/5981/1036" target='_blank'>Article</a>
  <a href="articles/2013_Myers_PLoS Genet.pdf" target='_blank'> [PDF] </a><br/>
  <u>peaks:</u><a href="peaks_url.bed" target='_blank'>bed file</a><br/>		
</ul>

<h5>Suggested Reading:</h5>
<ul>
  <li>Bailey et al. <i>Practical Guidelines for the Comprehensive Analysis of ChIP-seq Data.</i> PLoS Comput Biol 9, e1003326 (2013).<a href="articles/2013_Bailey_PLoS Comput Biol.pdf" target='_blank'> [PDF] </a><br/>
  </li>
  <li>Thomas-Chollier et al. <i>A complete workflow for the analysis of full-size ChIP-seq (and similar) data sets using peak-motifs.</i> Nature Protocols 7, 1551–1568 (2012).<a href="articles/2012_Thomas-Chollier_Nature Protocols.pdf" target='_blank'> [PDF] </a><br/>
  </li>
</ul>
<p/>



</div>

<div class="clearer">&nbsp;</div>

</div>

<div id="footer">

  <div class="left" id="footer-left">
    
    
    
    <p>&copy; 2014 Morgane Thomas-Chollier. All rights Reserved</p>

    <p class="quiet"><a href="http://templates.arcsin.se/">Website template</a> by <a href="http://arcsin.se/">Arcsin</a></p>
    
    <div class="clearer">&nbsp;</div>

  </div>

  <div class="right" id="footer-right">

  </div>

  <div class="clearer">&nbsp;</div>

</div>

</div>

</body>
</html>
