<html>
<head>
   <title>RSA-tools - tutorials - Position-specific scoring matrices</title>
</head>
<body text="#000000" bgcolor="#ffffff" link="#0000ee" vlink="#55008b" alink="#ff0000">
<blockquote>

<hr size=6 width="100%">
<center>
<h2><a href="../RSAT_home.cgi">RSA-tools</A> - 
<a href="tutorials.html">Tutorials</a> - 
Position-specific scoring matrices
</h2>
</center>

<hr size=6 width="100%">

<h2>Contents</h2>

<p>
<ol>
<li><a href=#intro>Introduction</a>
<li><a href=#example>Example</a>
<li><a href=#convert_matrix>Converting and analyzing PSSM</a>
</ol>

<a name=prerequisite><h2>Introduction</h2>
<ul>

Position-specific scoring matrices (PSSM) offer a sensitive way to
represent the specificity of transcription factor/DNA interfaces. PSSM
are built on the basis of a set of known binding sites for the factor
of interest.  

<P>
Knowledge-based PSSM can be obtained from various transcription factor
databases (e.g. <a target=_blank
href=http://www.gene-regulation.com/pub/databases.html#transfac>TRANSFAC</a>).

<p>PSSM can be used to detect occurrences of a motif in
sequences. Matrix-based pattern matching will be introduced in the
tutorial on <a href=tut_patser.html>patser</a>.

<P>Several pattern-discovery programs allow to detect significant
motifs represented as PSSM from a set of unaligned sequences. These
programs have been used to predict regulatory elements in the
regulatory regions of sets of co-regulated genes. Matrix-based pattern
discovery will be discussed in the tutorials on the <a
href=tut_gibbs.html>gibbs sampler</a> and <a
href=tut_consensus.html>consensus</a>.

</ul><a name=example><h2>Example</h2><ul>

The matrix below was obtained from <a
href="http://cgsigma.cshl.org/jian" TARGET=_blank> SCPD</a>, the
Saccharomyces cerevisiae promoter database.  


<p>The PHO4 matrix has been built from an alignment of 12 binding sites
for the yeast transcription factor Pho4p.

 
<PRE>
A |  3   2   0  12   0   0   0   0   1   3
C |  5   2  12   0  12   0   1   0   2   1
G |  3   7   0   0   0  12   0   7   5   4
T |  1   1   0   0   0   0  11   5   4   4
</PRE>


<P>Each row represents one residue (A, C, G or T), and each column a
position in a set of aligned binding sites. Some positions are
perfectly conserved across all known binding sites (the motif CACGT
starting at the 3rd position), whereas some other positions present
two choices (e.g. G or T at position 8), and other position can
contain any letter, but with different frequencies (e..g first and
last position).

<P>When the matrix isused to scan sequences for putative Pho4p binding
sites, the more conserved positions qimpose stronger constraints than
those where any nucleotide can be found. This allows thus a more
quantitative evaluation of the binding affinity than <a
href=tut_strings.html>string-based representations</a>.

<p>One weakness of these matrices is that they do not take into
account higher order dependencie, i.e. influence between a residue and
the successive ones. Thus, it is not possible to specify "either
CACGTGGG or CACGTTTT" with a PSSM : if one builds a matrix where G and
T are allowed at the 3 last positions, any combination of them will be
allowed (e.g.  CACGTGTG, CACGTTGT). Higher order dependencies can be
represented with more elaborated methods, such as Hidden Markov Models
(HMM), which are ou of scope for this tutorial.


</ul><h2><a name=convert_matrix>Converting and analyzing PSSM</a></h2><ul>

The RSAT include a program called <a target=tools
href=../convert-matrix_form.cgi>convert-matrix</a>, which allows to
extract a PSSM from the output files of different programs (<a
target=tools href=../consensus_form.cgi>consensus</a>, <a
target=tools href=../gibbs_form.cgi>gibbs</a>, <a target=_blank
href=http://meme.sdsc.edu/meme/website/>MEME</a>, clustal).


<p>I still did not find the time to write a tutorial for this program,
but you can already read the <a target=tools
href=../help.convert-matrix.html>manual</a>, and do your own test with
the demo data set (click DEMO at the bottom of the form).

<h3>Matrix information</h3>

We will import a matrix representing the binding specificity of the
yeast transcription factor Abf1p, and analyze the different parameters
which characterize this PSSM.

<ol>
<p><li>Connect <a href="http://cgsigma.cshl.org/jian" target=scpd>
SCPD</a>, the Saccharomyces cerevisiae promoter database. To access
the list of sites by factor from the SCPD home page, click on the
Section "<a target=scpd
href=http://cgsigma.cshl.org/cgi-bin/jz/getfactorlist>Regulatory
elements and transcriptional factors</a>".

<li>Click on the link to <a target=scpd
href=http://cgsigma.cshl.org/cgi-bin/jz/getfactor?ABF1,BAF1>ABF1</a>,
and then on the button <b>Get matrix</b>. The following PSSM is
displayed.

<pre>
>ABF17121SCPD

A   0   0  11   2   3   7   3   5   5  14   0   1
T  14   0   2   1   6   2   7   3   0   0   0   0
G   0   0   0   0   2   3   1   2   3   0   1  13
C   0  14   1  11   3   2   3   4   6   0  13   0. 
</pre>

<p>Some basic questions

<ul>
<li>What is the alphabet size ?

<li>What is the matrix width ?

<li>How many inding sites were used to build this matrix ? 

<li>How many Abf1p binding sites are currently stored in SCPD (click on the button <b>Get sites</b>)?
</ul>

<p><li>We will now analyze the content of this matrix with the program <a
target=tools
href=../convert-matrix_form.cgi><b>convert-matrix</b></a>. Copy the
matrix from SCPD to the text area in the form.

<p><li>The current version of convert-matrix does not support the SCPD
format (this will be fixed soon), but you can easily enter it in
consensus format. For this, you need to perform 3 operations: 
<ul>
<li>replace the header line
(<tt>>ABF17121SCPD</tt>) by the words <tt>MATRIX 1</tt>. It should
give the following result:
<li>Add a pipe character after each residue (A, C, G, T).
<li>Suppress the dot at the end of the matrix.
</ul>

The result should look like this:
<pre>
MATRIX 1

A |   0   0  11   2   3   7   3   5   5  14   0   1
T |  14   0   2   1   6   2   7   3   0   0   0   0
G |   0   0   0   0   2   3   1   2   3   0   1  13
C |   0  14   1  11   3   2   3   4   6   0  13   0
</pre>

<p><li>Select the following options proposed as return fields:
<ul>
<li>alignment (counts)
<li>frequencies
<li>weights
<li>information
<li>margins
<li>parameters
</ul>
<p><li>click <b>GO</b>.
</ol>

<h4>Interpretation</h4><ul>

<blink><font color=red size=+1>In construction</font></blink>
</ul>

</ul>
<p>
You can now come back to the <a href=tutorials.html>tutorial main
page</a> and follow the next tutorials.

</ul>
<hr width="100%">
<center>
<address>
For suggestions or information request, please contact <a
href="mailto:jvanheld@ucmb.ulb.ac.be">Jacques van Helden
(jvanheld@ucmb.ulb.ac.be)</a>
</address></center>

</blockquote>
</body>
</html>
