<html>
<head>
   <title>RSA-tools - tutorials - Position-specific scoring matrices</title>
</head>
<body text="#000000" bgcolor="#ffffff" link="#0000ee" vlink="#55008b" alink="#ff0000">
<blockquote>

<hr size=6 width="100%">
<center>
<h2><a href="../RSA_home.cgi">RSA-tools</A> - 
<a href="tutorials.html">Tutorials</a> - 
Position-specific scoring matrices
</h2>
</center>

<hr size=6 width="100%">

<h2>Contents</h2>

<p>
<ol>
<li><a href=#intro>Introduction</a>
<li><a href=#example>Example</a>
</ol>

<a name=prerequisite><h2>Introduction</h2>
<ul>

Position-specific scoring matrices (PSSM) offer a sensitive way to
represent the specificity of trnscription factor/DNA interfaces. PSSM
are built on the basis of a set of known binding sites for the factor
of interest.  

<P>
Knowledge-based PSSM can be obtained from various transcription factor
databases (e.g. <a href=tut_transfac.html>TRANSFAC</a>). 

<P>PSSM can also be inferred by applying matrix-based pattern
discovery in regulatory regions of a set of co-regulated genes (see
tutorials on the <a href=tut_gibbs.html>gibbs sampler</a> and <a
href=tut_consensus.html>consensus</a>).

</ul>
<a name=example><h2>Example</h2>
<ul>

The matrix below was obtained from <a
href="http://cgsigma.cshl.org/jian" TARGET=_blank> SCPD</a>, promoter
database of the yeast <I>Saccharomyces cerevisiae</I>. It has been
built from an alignment of 12 binding sites for the yeast
transcription factor Pho4p.

 
<PRE>
A |  3   2   0  12   0   0   0   0   1   3
C |  5   2  12   0  12   0   1   0   2   1
G |  3   7   0   0   0  12   0   7   5   4
T |  1   1   0   0   0   0  11   5   4   4
</PRE>


<P>Each row represents one residue (A, C, G or T), and each column a
position in a set of aligned binding sites. Some positions are
perfectly conserved across all known binding sites (the motif CACGT
starting at the 3rd position), whereas some other positions present
two choices (e.g. G or T at position 8), and other position can
contain any letter, but with different frequencies (e..g first and
last position).

<P>When the matrix will used to scan sequences for putative Pho4p
binding sites, the more conserved positions will impose stronger
constraints. This allows thus a more quantitative evaluation of the
binding cappability than <a href=tut_strings.html>string-based
representations</a>. 

<p>One weakness of these matrices is that they do not take into
account higher order dependencie, i.e. influence between a residue and
the successive ones. Thus, it is not possible to specify "either
CACGTGGG or CACGTTTT" with a PSSM : if one builds a matrix where G and
T are allowed at the 3 last positions, any combination of them will be
allowed (e.g.  CACGTGTG, CACGTTGT).


</ul>

<p>
You can now come back to the <a href=tutorials.html>tutorial main
page</a> and follow the next tutorials.

</ul>
<hr width="100%">
<center>
<address>
For suggestions or information request, please contact <a
href="mailto:jvanheld@ucmb.ulb.ac.be">Jacques van Helden
(jvanheld@ucmb.ulb.ac.be)</a>
</address></center>

</blockquote>
</body>
</html>
