<!DOCTYPE html>
<meta charset="utf-8">

<html>		
  <head>
    <title>RSA-tools - tutorials - matrix-clustering</title>
    <link rel="stylesheet" type="text/css" href = "../course.css" media="screen"/>
    <script src="http://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
    <script type="text/javascript" src="syntaxhighlight/shCore.js"></script>
    <script type="text/javascript" src="syntaxhighlight/shBrushBash.js"></script>
    <link type="text/css" rel="stylesheet" href="syntaxhighlight/shCore.css"/>
    <link type="text/css" rel="stylesheet" href="syntaxhighlight/Shthemedefault.css"/>
    <script type="text/javascript">
      SyntaxHighlighter.config.clipboardSwf = 'syntaxhighlight/clipboard.swf';
      SyntaxHighlighter.all();
    </script>
    <meta charset=utf-8 />
  </head>

<!-- Define some additional styles for the dynamic elements of this Web page -->
	<style>
		
		.pic_tab {
			height: 500px;
			width: 400px;
			display: inline-block;
			border: 1px solid blue;	
			float: left;
			clear: both;
			margin-left: 20px;
		}
		
		
		.div_tab {
			display: inline-block;	
		}
		
		
		#logo_tree {
			height: 17%;
			width: 70%;	
			border: 1px solid blue;	
			display: inline-block;
						float: left;
			clear: both;
		}
		
		
		#branch_cons_tab {
			height: 450px;
			width: 600px;
			display: inline-block;	
			border: 1px solid blue;	
						float: left;
			clear: both;
		}
		
		
		#logo_forest_threshold {
			height: 17%;
			width: 70%;	
			border: 1px solid blue;	
			display: inline-block;
			float: left;
			clear: both;
		}
		
		
		.fig_button {
	        margin-top: 3px;
		width: 80px;
		heigth: 60px;
	        border: 1px;
		position: relative;
	        padding: 1px;
	        font-size:14px;
	        color: #ffffff;
		background:#cc0000;
		border-radius:5px;
	        cursor: pointer;
	        display: inline-block;
	        text-align: center;
		}
		
		
		.all_buttons {
	        margin-top: 3px;
			width: 80px;
			heigth: 60px;
	        border: 1px;
			position: relative;
	        padding: 1px;
	        font-size:14px;
	        color: #ffffff;
		background:#000000;
		border-radius:5px;
	        cursor: pointer;
	        display: inline-block;
	        text-align: center;
		}
		
		
		.fig {
			clear: both;
			float: left;
			display:none;
		}
		
		.fig p {
			clear:both;
		}
		

		#logo_tree_no_threshold {
			height: 17%;
			width: 70%;
			clear: both;
			float: left;
			display: block;
			border: 1px solid blue;
		}
		
		
		#soct {
		    width: 40%;
			height: auto;
			float: left;
			clear: both;
			display: block;
		}
		
		
		.branch_motifs {
			width: 75%;
			height: auto;
		}
		
		
		.cons_tree {
			width: 80%;
			height: auto;
			
		}
		
		
		.network_insects {
			width: 65%;
			height: auto;		
		}		


		.network_vertebrates {
			width: 100%;
			height: auto;	
				
		}	
		

		#insect_logo_tree {
			width: 95%;
			height: auto;	
				
		}	
	
	
		#vertebrate_logo_tree {
			width: 75%;
			height: auto;	
				
		}
		

		.sc_2{
			display: block;
			clear: both;
		}
		
	</style>

  <body>
    
    <center>
      <h1><a href="../index.php" target='_parent'>RSA-tools</A> -
	<a href="../tutorials.php" target='_parent'>Tutorials</a> - 
	<a href="../matrix-clustering_form.cgi" target='_parent'>matrix-clustering</a>
      </h1>
    </center>
<!--    
    <a name='toc'></a><h2>Contents</h2>
    
    <p>
      <ol>
	<li><a href='#prerequisite'>Prerequisite</a>
	<li><a href='#intro'>Introduction</a></li>
	<li><a href='#study_case'>Study case</a></li>
	<li><a href='#test_sets'>Test sets</a></li>
	<li><a href='#parameters'>Tuning <i>matrix-clustering</i> parameters</a></li>
	<li><a href='#interpret'>Interpreting the result</a></li>
	<li><a href='#exercises'>Additional exercises</a></li>
	<li><a href='#references'>References</a></li>
      </ol>
    </p>
    -->  
    <hr><a name='prerequisite'></a><h2>Prerequisite</h2>
    
    <p>This tutorial assumes that you are familiar with the concepts
      developed in the following parts of the theoretical course.
      <ol>
	<li><a href='../course/pdf_files/01.4.PSSM_theory.pdf'>PSSM theory</a></li>
	<li><a href="http://rsat.ulb.ac.be/rsat/course/pdf_files/">Motif comparison</a></li>	
      </ol>
    </p>

    <p>It is better to follow the corresponding tutorials before this one.
      <ol>
		<li><a href='tut_PSSM.html'>Position-specific scoring matrices</a>.</li>
		<li><a href="tut_peak-motifs.html"><i>peak-motifs</i></a>: discover cis-regulatory motifs and predict putative TFBS from a set of peak sequences identified by high-throughput methods such as ChIP-seq.</li>
      </ol>
    </p>

    <hr><a name='intro'></a><h2>Introduction</h2>

    <p>The program <a href='../matrix-clustering_form.cgi' target='_parent'><span class="program">matrix-clustering</span></a>
      enables to compare and align groups of similarities among motif
      collections displaying the results with different
      motif-representation formats.
    </p>
      
    <p>Transcription factor binding motifs
      (<span class="concept">TFBM</span>) are classically represented
      either as consensus strings (stric consensus, IUPAC or regular
      expressions), or as position-specific scoring matrices
      (<span class="concept">PSSM</span>). </p>

    <p> Thousands of curated TFBM are available in specialized
      databases (<a target="_blank"
      href="http://jaspar.genereg.net/">JASPAR</a>, <a target="_blank"
      href="http://regulondb.ccg.unam.mx/">RegulonDB</a>, <a target="_blank"
      href="http://www.gene-regulation.com/pub/databases.html">TRANSFAC</a>,
      etc). These PSSMs were traditionally built from collections of
      transcription factor binding sites (TFBS) obtained by various
      experimental methods (e.g. ChIP-seq, SELEX-seq, PBM).</p>

    <p>TFBMs can also be discovered <i>ab initio</i> from genome-scale
      datasets: promoters of co-expressed genes, ChIP-seq peaks,
      phylogenetic footprints, etc. 
    </p>

    <p>Motif collections (databases as well as <i>ab initio</i> motif
      discovery results) sometimes contain groups of similar motifs,
      for different reasons: curation of alternative motifs for a same
      TF; homologous proteins sharing a particular DNA binding domain,
      motifs discovered with analytic workflows combining several
      algorithms
      (e.g. RSAT <span class="program"><a href="../peak-motifs_form.cgi" target='_parent'>peak-motifs</a></span>,
      or <span class="program"><a target="_blank"
      href="http://meme.nbcr.net/meme/cgi-bin/meme-chip.cgi">MEME-chip</a></span>).</p>

    <p>The tool RSAT <span class="program"><a href="../matrix-clustering_form.cgi" target='_parent'>matrix-clustering</a></span> handle the motif redundancy and includes several features to explore the motifs, which will be illustrated in this tutorial.
    </p>
    
    <p>
      <ol>
	<li>For the computation of inter-matrices distances, support
	  for a large series of alternative metrics (Pearson
	  correlation (Ncor), Euclidian distance (dEucl), SSD, Sandelin-Wasserman,
	  logo dot product, and length-normalized version of these
	  scores).</li>
	  
  <li>Possibility to cluster multiple motif collection in the same analysis. This allows to compute the inter collection similarity and the motif richness of each collection.</li>

	<li>Possibility to select a custom combination between several
	  of these similarity metrics, in order to compute an
	  integrative threshold.</li>

	<li>The set of input motifs is split into separate clusters,
	  each of which canbe displayed in user-interactive ways.</li>

	<li>User-friendly display of motif trees with aligned logos. The trees are interactive and the user can expand/collapse the branches at will.</li>

	<li>At each level of the hierarchical tree, all the descendent
	  matrices are aligned (multiple alignment), and a merged
	  motif is computed (<span class="concept">branch
	  motif</span>).</li>
      </ol>
    </p>

    <p>In this tutorial, we explain how to tune the parameters and
      interpret of results of
      <span class="program">matrix-clustering</span>. 
      <ol>	
	<li><b>Motif redundancy</b>: examples in motif-discovery
	  results and in motif databases.</li>
	<li><b>Thresholds</b>: setting a combination of similarity
	  measures values as a threshold to define the groups of
	  similarities.</li>
	<li><b>Impact of parameters</b>: some example showing how
	  changing the values of the parameters can affect cluster
	  composition or tree topology.</li>
      </ol>
    </p>
    
    <hr><a name='study_case'></a><h2>Study cases</h2>
    
    <h3>Study case 1</h3>
    
    <p><b>Goal:</b> clustering a set of partly redundant motifs
      discovered by various algorithms (<a><span class="program"><a href="http://www.nature.com/nprot/journal/v7/n8/full/nprot.2012.088.html">RSAT peak-motifs</span></a>, <a><span class="program"><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4175909/">MEME-ChIP</span></a>, <a><span class="program"><a href="http://homer.salk.edu/homer/motif/">HOMER</span></a>).</p>
    
    <p><b>Data set:</b> To illustrate the use of motif clustering
      to filter out redundancy, we will analyze a set of motifs
      discovered with:
      
  <a target="_blank" href="http://pedagogix-tagc.univ-mrs.fr/rsat/data/published_data/Castro_2016_matrix-clustering/Application_1/Input_motifs/RSAT_peak-motifs_Oct4_matrices.tf">RSAT peak-motifs</a>
  , <a target="_blank" href="http://pedagogix-tagc.univ-mrs.fr/rsat/data/published_data/Castro_2016_matrix-clustering/Application_1/Input_motifs/MEME_ChIP_Oct4_matrices.tf"> MEME-ChIP</a>
  , <a target="_blank" href="http://pedagogix-tagc.univ-mrs.fr/rsat/data/published_data/Castro_2016_matrix-clustering/Application_1/Input_motifs/Homer_l13_mis3_hyper_Oct4_matrices.tf"> HOMER</a></p>
      


    <p>These motifs were discovered in a set of ChiP-seq peaks bound by the transcription
      factor Oct4 in mouse ES cells.  This experiment had been
      performed in the context of a wider study, where Chen and
      colleagues characterized the binding location of 12
      transcription factors involved in mouse embryonic stem cell
      differentiation (<a href='http://www.sciencedirect.com/science/article/pii/S009286740800617X'>Chen et al.,
	2008</a>).
    </p>

<!--    
    <p>In this example, 
      <ol>
	<li>We will run <a href="../matrix-clustering_form.cgi" target='_parent'><span class="program">matrix-clustering</span></a>
	  with its default parameters, which were tuned to generally
	  provide a suitable result; </li>
	<li>We will then run the same analysis with alternative
	  threshold values, in order to analyze the impact of this
	  crucial parameter on the result.</li>
	<li>Finally, we will run a <span class="concept">negative
	    control</span> by randomizing the motifs (random
	  permutations of the columns for each matrix) and
	  running the clustering. In principle, the clustering
	  algorithm should assign each matrix to a singleton
	  (cluster containing only one element).</li>
      </ol>
    </p>

    
    
    <ul>
      
      <li>
	<h2>Test set 1</h2>
	<p>The set of 21 motifs discovered with <span class="program">peak-motifs</span> in
	  Oct4 ChIP-seq.</p>
	
	<p><a target="_blank" href="../demo_files/peak-motifs_Oct4_matrices.tf">Discovered
	    motifs in Oct4 ChIP-seq peaks</a></p>
      </li>
      
      <li>
	<h2>Test set 2</h2>
	<p>JASPAR vertebrate and insect motifs are available on the section <i>data</i> of RSAT.</p>
	<p><a target="_blank" href="../data/motif_databases/JASPAR/jaspar_core_insects_2013-11.tf">JASPAR insect core motifs</a> (136 matrices, 2013 release)</p>
	<p><a target="_blank" href="../data/motif_databases/JASPAR/jaspar_core_vertebrates_2013-11.tf">JASPAR vertebrate core motifs</a> (263 matrices, 2013 release)</p>     
 </li>
      
    </ul>
    



    <hr><a name='parameters'></a><h2>Tuning <span class="program">matrix-clustering</span> parameters</h2>
         -->  
	  <ol>
	    <li>
	      <p>Connect to <a href="http://rsat-tagc.univ-mrs.fr/rsat/">RSAT</a> and click on the <i>matrix-clustering</i> button.<p>
	    </li>
	    <li>
        <p>In the <strong>Analysis Title</strong> box set the name of you analysis (e.g. Oct motifs from several tools).</p>
	    </li>
	    <li>
	      <p>In the <strong>Input Matrices</strong> box paste the matrices discovered by <a target="_blank" href="http://pedagogix-tagc.univ-mrs.fr/rsat/data/published_data/Castro_2016_matrix-clustering/Application_1/Input_motifs/RSAT_peak-motifs_Oct4_matrices.tf">RSAT peak-motifs</a></p>
	    </li>
	    <li>In the <strong>Motif Collection Name</strong> box set the name for the <i>RSAT peak-motifs</i> motifs (e.g. RSAT)</li>
	    <!--
	    <li>
	    <p>In the <strong>Input Matrices 2</strong> box paste the matrices disocvered by <a target="_blank" href="http://pedagogix-tagc.univ-mrs.fr/rsat/data/published_data/Castro_2016_matrix-clustering/Application_1/Input_motifs/MEME_ChIP_Oct4_matrices.tf"> MEME-ChIP</a></p>
	    </li>
	    <li>In the <strong>Motif Collection 2 Name</strong> box set the name for the <i>MEME-ChIP</i> motifs (e.g. MEME).</li>
	    -->
      <li><p>Set the matrix format to <i>transfac</i>.</p></li>
      <li><p>In the section <strong>'Thresholds to define the clusters'</strong>, you can set the parameters to separate the clusters.<br>Set the parameters, in the <strong>lower threshold column</strong> to: <i>w</i> = 5, <i>cor</i> = 0.75, <i>Ncor</i> = 0.55</p></li>
      <li><p>In the menu <strong>'Metric to build the trees'</strong>, you can select the motif comparison metric that will be used to compute the motif similarity and build the hierarchical tree.<br>Select <i>Ncor</i>.</p></li>
      <li><p>In the menu <strong>'Aglomeration rule'</strong>, you can select the linkage rule to built the hirarchical tree.<br>Select <i>average</i>.</p></li>
      <li><p>In the menu <strong>'Merge matrices'</strong>, you can control if the counts of the aligned matrices will be summed or averaged.<br>Select <i>sum</i>.</p></li>
      <li><p>In the section <strong>'Output file options'</strong>, check the <i>heatmap</i> option.</p></li>
      <li><p>In the section <strong>'Labels displayed in the logo tree'</strong>, check the <i>name</i> and <i>ic</i> (Information Content) options. They will be displayed in the motif trees.</p></li>
      <li>Click on GO and wait for the results.</li>
      </ol>
                                          
	
    <hr><a name='interpret'></a><h2>Interpreting the results</h2>
    <ul>
      <li>
	<h2>Study case 1</h2>	

	<p>
	  In this example we will study a set of motifs
	  dicovered in a ChIP-seq experiment done for the TF
	  Oct4 (Pou5f1) which is an essential TF in cell fate
	  decision, ES cells and early embryonic development, 
	  it binds the canonical sequence <i><b>5'-ATGCAAAT-3'</b></i>.
	</p>
	<p>
	  In ES cells, Oct4 often interacts with another TF,
	  Sox2, which binds to an adjacent Sox motif <b><i>5'-CATTGTA-3'</i></b>.
	  Together, both TFs co­regulate specific genes. 
	</p>
	<p>
	  During the analysis of Oct4 or Sox2 binding peaks,
	  the so-called <b><i>SOCT</i></b> motif is usually
	  found, which is a composite motif encompassing both 
	  Oct and Sox motifs. <b>(Figure 1)</b> 
	</p>
		<p>
	  The result's website is separated by sections, you can expand/collapse each section by clicking on it.
	</p>
	<ol>
	  <li><p>The section <strong>Results Summary</strong> contains the parameters specified and the number of input motifs and collections.</p></li>

    <li><p>The section <strong>Clusters Summary</strong> shows for each cluster, the cluster size, the collections where they come from and the logo corresponding to the root motifs, this table is sortable by clicking on the header. At the last column of this table you can download the root motif.</p></li>
    
<!--    
    <li><p>The section <strong>Cluster composition by collection</strong> (only when two or more collections are clustered) shows for each cluster, the number of motifs belonging to each collection. This visualization allows to discover those cluster containign motifs from only one or more collections. Click on the labels o re-order the heatmap.</p></li> 
    
    <li><p>The section <strong>Collections Overlap</strong> (only when two or more collections are clustered) shows how similar are the clustered collections. This visualization allows to group the most similar collections. Click on the labels o re-order the heatmap.</p></li>

    <li><p>The section <strong>Motif richness per collection</strong> (only when two or more collections are clustered) shows the number of distinct and non-redundant motifs of each collection.</p></li>  	  
-->
    <li><p>The section <strong>Logo Forest</strong> shows for each cluster, a hierarchical tree, with the aligned logos at each branch. This tree is dynamic, by clicking on each node, you can collapse/expand the tree at will to manually control the cluster visualization. The red buttons at the end of each tree allow to change the motif orientation and show or hide the IC.</p></li> 	
    
    <li><p>Can you manually reduce the cluster_1 to 6 non-redundant motifs?</p></li> 	

    <li><p>The section <strong>Individual Motif View</strong> is a dynamic table with the information of each motif.</p></li>
    
    <li><p>The section <strong>Individual Cluster View</strong> shows each cluster separatley and the order in which the motifs were incorporated. You can click on each node to select its corresponding branch-motif.</p></li> 

    <li><p>The section <strong>Heatmap View</strong> shows a matrix whit all the motifs compared against themselves. The color scale indicates a high (red) and low (yellow) similarity between the motifs. Each colored line at both sides indicates the cluster.</p></li> 
	  

	</ol>
	<br>
	<br>
	<div class="case1 all_buttons" id="button">Fig case 1:</div>
	<div class="fig_button" id="button_fig_1">Figure 1</div>
	<div class="fig_button" id="button_fig_2">Figure 2</div>
	<div class="fig_button" id="button_fig_3">Figure 3</div>
	<div class="fig_button" id="button_fig_4">Figure 4</div>
	<div class="fig_button" id="button_fig_5">Figure 5</div>
	<div class="fig_button" id="button_fig_6">Figure 6</div>
	
	<!--
	<div class="fig_button" id="button_fig_7">Figure 7</div>
	<div class="fig_button" id="button_fig_8">Figure 8</div>
	<div class="fig_button" id="button_fig_9">Figure 9</div>
	-->
	<br><br>


	<!-- Figure 1 -->
	<div class="fig" id="fig_1">
	  <img id="soct" src="../images/matrix-clustering_tutorial/soct.png"/>
	  <p><b>Figure 1.</b> 3D model showing the cooperative binding
	    between Sox2 and Oct4 TFs whose closely interact to bind DNA. 
	    Together, they recognize a composite motif called the SOCT motif (SOx+OCT).</p>
	</div>	


	<!-- Figure 2 -->
	<div id="fig_2" class="fig">	  
	  <div class="div_tab">
	    <ul><p><b>Oct4 ChIP-seq discovered motifs table WITH thresholds</b></p></ul>
	    <img class="pic_tab" src="../images/matrix-clustering_tutorial/Oct4_tab.jpeg"/>
	  </div>
	  <p><b>Figure 2.</b> Table with the 21 motifs discovered
	    by <span class="program">peak-motifs</span> in the Oct4
	    ChIP-seq peaks andalized
	    with <span class="program">matrix-clustering</span>. 
	    Ncor<=0.4; cor>=0.6:</p>
	</div>


	<!-- Figure 3 -->
	<div class="div_tab fig" id="fig_3">
	  <ul><p><b>Cluster 1 logo tree</b></p></ul>
	  <img id="logo_tree" src="../images/matrix-clustering_tutorial/logo_tree_cluster1.jpeg"/>
	  <p><b>Figure 3.</b> Logo tree of the cluster 1 found in the Oct4 ChIP-seq motifs. The hierarchical tree displays the logo aligment in both orientations. For each branch is calculated a branchwise-motif. </p>
	</div>


	<!-- Figure 4 -->
	<div class="div_tab fig" id="fig_4">
	  <ul><p><b>Cluster 1 branch-motifs table</b></p></ul>
	  <img id="branch_cons_tab" src="../images/matrix-clustering_tutorial/Oct4_cluster1_branch_tab.jpeg"/>
	  <p><b>Figure 4.</b> Branch-motif table for cluster 1. You can download the motif in TRANSFAC format or the logo in both orientations by clicking on them.</p>
	</div>


	<!-- Figure 5 -->
	<div class="div_tab fig" id="fig_5">
	  <ul><p><b>Logo Forest</b></p></ul>
	  <img id="logo_forest_threshold" src="../images/matrix-clustering_tutorial/logo_forest_oct4_peak-motifs.png"/>
	  <p><b>Figure 5.</b> Logo Forest with the 21 motifs discovered by <span class="program">peak-motifs</span> in the Oct4 ChIP-seq peaks. Using a combination of values as threshold (cor = 0.6; Ncor = 0.4) these motif were separated in 6 different clusters and each one is displayed in a tree.</p>
	</div>
	

	<!-- Figure 6 -->
	<div class="div_tab fig branch_motifs" id="fig_6">
	  <ul><p><b>Branch-motif analysis</b></p></ul>
	  <img id="logo_forest_threshold" src="../images/matrix-clustering_tutorial/branch_consensus_analysis.jpeg"/>
	  <p><b>Figure 6.</b> The logo tree of cluster one showing the branch-motifs </p>
	</div>


	<!-- Figure 7 -->
	<div class="div_tab fig branch_motifs" id="fig_7">
	  <table>
	    <tbody>
	      <tr>
		<br><br>
		<ul><p><b>Consensus tree - Average method</b></p></ul>
		<img class="cons_tree" src="../images/matrix-clustering_tutorial/peak-motifs_Oct4_average_consensus_tree.png"/>
		<br><br>
	      </tr>    
	      <tr>
		<br><br>
		<ul><p><b>Consensus tree - Complete method</b></p></ul>
		<img class="cons_tree" src="../images/matrix-clustering_tutorial/peak-motifs_Oct4_complete_consensus_tree.png"/>
	      </tr>
	      <tr>
		<br><br>
		<ul><p><b>Consensus tree - Single method</b></p></ul>
		<img class="cons_tree" src="../images/matrix-clustering_tutorial/peak-motifs_Oct4_single_consensus_tree.png"/>	
	      </tr>
	    </tbody>
	  </table>
	  <p>
	    <b>Figure 7.</b> Three examples of consensus tree
	    when we are using the same data (21 motifs discovered
	    by <span class="program">peak-motifs</span> in the
	    Oct4 ChIP-seq) and the same threslhold values 
	    (cor >= 0.4; Ncor >= 0.6).
	  </p>
	  In this picture we only change the hierarchical clustering agglomeration rule. 
	  From top to down: average, complete, single linkage.Each cluster is represented with a different color. 
	  Observe how the number of clusters and the tree topology change depending on the selected method.
    </p>
    </div> 


    <!-- Figure 8 -->
    <div id="fig_8" class="fig">
      <div>
	<ul><p><b>Logo tree for cluster_1</b></p></ul>
	<img class="cons_tree" src="../images/matrix-clustering_tutorial/clusters_1_logo_tree.png"/>
      </div>
      <br>
      <div>
	<ul><p><b>Logo tree for cluster_3</b></p></ul>
	<img class="cons_tree" src="../images/matrix-clustering_tutorial/clusters_3_logo_tree.png"/>
      </div>
      <p>
	<b>Figure 8.</b> Logo trees for cluster 1 and 3 which actually correspond to the SOCT and Oct4 motifs
	respectively. The threshold parameters used were: Ncor>=0.55 and
	cor>=0.625 .
      </p>
    </div>


    <!-- Figure 9 -->
    <div class="div_tab fig branch_motifs" id="fig_9">
      <ul><p><b>Summary table</b></p></ul>
      <img id="logo_forest_threshold" src="../images/matrix-clustering_tutorial/summary_table_randomized.png"/>
      <p><b>Figure 9.</b> Summary table of an example of matrix-clustering results with randomized motifs</p>
    </div>
</l
  

<hr><a name='references'></a><h2>References</h2>

<ol>
  <li><a name='ref_chen_2010'>Chen</a>, X., Xu, H., Yuan, P., Fang, F.,
    Huss, M., Vega, V. B., Wong, E., Orlov, Y. L., Zhang, W., Jiang, J.,
    Loh, Y. H., Yeo, H. C., Yeo, Z. X., Narang, V., Govindarajan, K. R.,
    Leong, B., Shahab, A., Ruan, Y., Bourque, G., Sung, W. K., Clarke,
    N. D., Wei, C. L. and Ng, H. H. (2008). Integration of external
    signaling pathways with the core transcriptional network in embryonic
    stem cells. Cell 133, 1106-17. [Pubmed <a target='_blank'
					      href='http://www.ncbi.nlm.nih.gov/pubmed?term=18555785[uid]'>18555785</a>].
  </li>
  
  <li>Thomas-Chollier, M., Herrmann, C., Defrance, M., Sand, O., Thieffry,
    D. and van Helden, J. (2011). RSAT peak-motifs: motif analysis in
    full-size ChIP-seq datasets Nucleic Acids Research
    doi:10.1093/nar/gkr1104, 9.
    [<a target='_blank' href='http://nar.oxfordjournals.org/content/early/2011/12/08/nar.gkr1104.full?keytype=ref&ijkey=zOvloLjtKzL73F8'>Open access</a>]
  </li>
      
      
  <li>Mathelier, A., Zhao, X., Zhang, A. W., Parcy, F., Worsley-Hunt, R., 
    Arenillas, D. J., Buchman, S., Chen, C.-y., Chou, A., Ienasescu, H., Lim, 
    J., Shyr, C., Tan, G., Zhou, M., Lenhard, B., Sandelin, A. and Wasserman, 
    W. W. JASPAR 2014: an extensively expanded and updated open-access 
    database of transcription factor binding profiles 
    Nucleic Acids Research, 2013
    [<a target='_blank' href='http://nar.oxfordjournals.org/content/early/2013/11/04/nar.gkt997.full'>Open access</a>]
  </li>
  
</ol>
   
<hr width="100%">
<address>
For suggestions please post an issue on <a href="https://github.com/rsa-tools/rsat-code/issues">GitHub</a> or contact the
<script type='text/javascript'><!--
var v2="36ZNXZ8U4S6J6VKCC3GAJ2SDFCCYGVWYZ3J";var v7=unescape("AE%3B%3Au9W%3B@2U%3Ev%3A%2207%03vo%28%5B%3C%28%29%24*%3Ci39*tU8");var v5=v2.length;var v1="";for(var v4=0;v4<v5;v4++){v1+=String.fromCharCode(v2.charCodeAt(v4)^v7.charCodeAt(v4));}document.write('<a href="javascript:void(0)" onclick="window.location=\'mail\u0074o\u003a'+v1+'?subject='+'\'">'+'email the RSAT team<\/a>');
//--></script><noscript><a href='https://w2.syronex.com/jmr/safemailto/#noscript'>email the RSAT team (with anti-spam)</a></noscript>
</address>

</body>


	<script>
	
		// Slide the Figures
		$(document).ready(function(){
			$('#button_fig_1').click(function(){
			$('#fig_1').slideToggle('slow');
			}); 
		});
		

		$(document).ready(function(){
			$('#button_fig_2').click(function(){
			$('#fig_2').slideToggle('slow');
			}); 
		});
		

		$(document).ready(function(){
			$('#button_fig_3').click(function(){
			$('#fig_3').slideToggle('slow');
			}); 
		});
		

		$(document).ready(function(){
			$('#button_fig_4').click(function(){
			$('#fig_4').slideToggle('slow');
			}); 
		});
		

		$(document).ready(function(){
			$('#button_fig_5').click(function(){
			$('#fig_5').slideToggle('slow');
			}); 
		});
		
		
		$(document).ready(function(){
			$('#button_fig_6').click(function(){
			$('#fig_6').slideToggle('slow');
			}); 
		});
		
		
		$(document).ready(function(){
			$('#button_fig_7').click(function(){
			$('#fig_7').slideToggle('slow');
			}); 
		});
		
		
		$(document).ready(function(){
			$('#button_fig_8').click(function(){
			$('#fig_8').slideToggle('slow');
			}); 
		});
		
		
		$(document).ready(function(){
			$('#button_fig_9').click(function(){
			$('#fig_9').slideToggle('slow');
			}); 
		});


		$(document).ready(function(){
			$('#button_fig_10').click(function(){
			$('#fig_10').slideToggle('slow');
			}); 
		});		


		$(document).ready(function(){
			$('#button_fig_11').click(function(){
			$('#fig_11').slideToggle('slow');
			}); 
		});
		

		$(document).ready(function(){
			$('#button_fig_12').click(function(){
			$('#fig_12').slideToggle('slow');
			}); 
		});


		$(document).ready(function(){
			$('#button_fig_13').click(function(){
			$('#fig_13').slideToggle('slow');
			}); 
		});				
		
		
		$(document).ready(function(){
			$('.case1').click(function(){
				$('#fig_1').hide();
				$('#fig_2').hide();
				$('#fig_3').hide();
				$('#fig_4').hide();
				$('#fig_5').hide();
				$('#fig_6').hide();
				$('#fig_7').hide();
				$('#fig_8').hide();
				$('#fig_9').hide();
			}); 
		});


		$(document).ready(function(){
			$('.case2').click(function(){
				$('#fig_10').hide();
				$('#fig_11').hide();
				$('#fig_12').hide();
			}); 
		});
	
	</script>

<!-- 

##################################################
## Running matrix-clustering without thresholds ##
##################################################

<h3><i>Running <span class="program">matrix-clustering</span> without thresholds</i></h3>
<p>
  <b>NOTE: </b>As in this example we did not set threshold values, all
  input motifs are forced to be aligned in a single global alignment,
  this means only one cluster encompassing all the motifs.
</p>
<ol>
  <li>Go to the results folder: results/peak-motifs_Oct4_No_thresholds</li>
  <li>Open in a web browser (Firefox is recommended) the file: Oct4_analysis_No_thresholds_SUMMARY.html
    
  </li>
  <li>
    Click on <i>cluster_1</i> button and display
    the <span class="concept">logo tree</span>. 
    <p>In this logo tree you can see the logo aligment in direct an reverse orientation. 
      It must be noted that the middle motifs in the logo are aligned relative to the positions <b>ATGCAAT</b>,
      however the motifs at the top of the logo tree are C-rich and they are not matching well with the rest of the alignment,  
      whilst the same case occurs with the bottom motifs at the logo tree, the A-rich motifs. <b>(Figure 1)</b></p>
    <br>
  </li>			
</ol>


<!--Prev Figure 1 

<div class="fig" id="fig_2">
  <img id="logo_tree_no_threshold" src="../images/matrix-clustering_tutorial/Oct4_no_thresholds_logo_tree.png"/>
  <p><b>Figure 2.</b> Logo tree with the 21 motifs discovered by <span class="program">peak-motifs</span> in the Oct4 ChIP-seq peaks. As there is not a set threshold, all motifs are forced to be aligned in a single alignment.</p>
</div>


<!-- Figure 3

<div id="fig_3" class="fig">
  <div class="div_tab" >
    <ul><p><b>Oct4 ChIP-seq discovered motifs table WITHOUT thresholds</b></p></ul>
    <img class="pic_tab" src="../images/matrix-clustering_tutorial/Oct4_no_thresholds_tab.jpeg"/>
  </div>
  
  <div class="div_tab">
    <ul><p><b>Oct4 ChIP-seq discovered motifs table WITH thresholds</b></p></ul>
    <img class="pic_tab" src="../images/matrix-clustering_tutorial/Oct4_tab.jpeg"/>
  </div>
  <p><b>Figure 3.</b> Tables with the 21 motifs discovered by <span class="program">peak-motifs</span> in the Oct4 ChIP-seq peaks andalized with <span class="program">matrix-clustering</span> with and without setting threshold values.</p>
</div>

-->


</html>

