<HTML>
<HEAD>
   <META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
   <TITLE>RSA-tools - tutorials</TITLE>
</HEAD>
<BODY TEXT="#000000" BGCOLOR="#FFFFFF" LINK="#0000EE" VLINK="#55008B" ALINK="#FF0000">
<blockquote>

<HR SIZE=6 WIDTH="100%">
<CENTER>
<H2><A HREF="../RSA_home.cgi">RSA-tools</A> - 
<a href="tutorials.html">Tutorials</a> - 
<a href="../retrieve-seq_form.cgi">retrieve-seq</a>
</H2>
</CENTER>

<HR SIZE=6 WIDTH="100%">

In the left pannel, under the title <B>Sequence retrieval</B>, click
on the tool "<B>Retrieve sequence</B>". A form appears, allowing you
to select the parameters of sequence retrieval.


<h3>Retrieving upstream sequences for a selected set of genes</h3>
<UL>

We will first retrieve upstream sequences for a set of 5 yeast genes
involved in yeast phosphate metabolism.

<OL>

<P><LI>
In the pop-up menu "<B>Organism</b>", select <I>Saccharomyces cerevisiae</I> 

<P><LI>
In the text area under "<B>Genes</B>", type the names of your genes</LI>

<PRE>
YBR093C
PHO8
pho11
PHO81
Pho84
</PRE>

<B>Warning:</B> gene names must be separated by carriage returns, because
only the first word of each line is considered as a query.


<P>
Notice that gene names are case-insensitive. 

<P>Genes can be specified either by the systematic ORF identifier or
by a common name. Synonyms are supported.

<P><LI>Leave all other options unchanged and click on the button <B>GO</B></LI>

</OL>

<P>
After a few seconds, the upstream sequence of the selected genes
will be displayed.

<P>

At the bottom of the sequences, you can see a series of buttons, which
will allow you to send the retrieved sequences to the next task (see
other tutorials).


</UL>
<h3>Retrieving all downstream sequencs</h3>
<UL>

As an illustration fo the retrieval of all sequences, we will retrieve
the complete set of downstream sequences for the yeast
<I>Saccharmoyces cerevisiae</I>

<OL>

<P><LI> Come back to the previous form by clicking <b>retrieve
sequence</b> in the left pannel.

<P><LI>Make sure that the selected organism is <I>Saccharomyces
cerevisiae</I>

<P><LI>Besides the label <B>Genes</B>, select "all".

<P><LI>Do not enter anything in the "Gene" box (it would be ignored
anyway).

<P><LI>Select "downstream" as <B>Sequence type</B>.

<P><LI>In the <B>From</b> box, type "1".

<P><LI>In the <B>to</b> box, type "200".

<P><LI>Select "ORF id + Gene name" in the pop-up menu <B>Sequence
label</B>

<P><LI>Click <B>GO</B>

</OL>

</ul>
<h3>Reference point</H3>
<ul>

<P>
The reference point (coordinate 0) depends on the type of sequences to
be retrieved : 
<P>

<ul>

<li>For <B>upstream sequences</B>, the reference point is the ORF start,
i.e. the first nucleotide of the start codon. Negative coordinates are
used to indicate sequences located upstream the start codon, i.e. in
the 5'end flank of the gene (e.g. from -1 to -800 for the yeast).

<li>On the contrary, when selecting <B>downstream sequences</B>, the
reference point (position 0) is the ORF end, i.e. the last nucleotide
of the stop codon. Positive coordinates are used to represent
sequences located downstream the stop codon, i.e. the 3'end flank of
the gene.

</UL>

</ul>
<h3>Positive/negative coordinates</H3>
<ul>

<P>
Notice that selecting upstream sequencs with positive coordinates
would amont to pick coding sequences on the 5' side of the gene.
For example, one can extract all start codons for a given organism by
retrieving all upstream sequences from position 0 to position +2. Try it.

<P>
Recprocally, for downstream sequences, negative coordinates could be
specified to retrieve coding sequences on the 3'en of the gene. One
could thus retrieve all downstream sequences from -2 to 0 to extract
all the stop codons found in a selected organism. 

<P>
You can now come back to the <a href=tutorials.html>tutorial main
page</a> and try the next tutorial.

</UL>
<HR WIDTH="100%">
<CENTER>
<ADDRESS>
For suggestions or information request, please contact :</ADDRESS></CENTER>

<CENTER><A HREF="mailto:jvanheld@ucmb.ulb.ac.be">Jacques van Helden (jvanheld@ucmb.ulb.ac.be)</A></CENTER>

</blockquote>
</BODY>
</HTML>
