<html>
<head>
   <title>RSA-tools - tutorials - microarrays</title>
<link rel="stylesheet" type="text/css" href = "../main.css" media="screen">
</head>
<body class="info">
<blockquote>


<center>
<h2><a href="../RSAT_home.cgi">RSA-tools</A> - 
<a href="tutorials.html">Tutorials</a> - 
<a href="../microarrays_form.cgi">microarrays</a> 
</h2>
</center>



<h2>Contents</h2>

<p>
<ol>
<li><a href=#prerequisite>Prerequisite</a>
<li><a href=#intro>Introduction</a>
<li><a href=#example>Example of utilization</a>
<li><a href=#interpret>Interpreting the result</a>
<li><a href=#exercises>Additional exercises</a>
</ol>

<a name=prerequisite></a><h2>Prerequisite</h2>
<ul>

This tutorial assumes that you already followed the tutorials on
pattern discovery.

</ul>
<a name=intro></a><h2>Introduction</h2>
<ul>

<p>
One of the most common utilizations of the tools proposed on this web
site is the prediction of regulatory motifs from clusters of
co-expressed genes.


For this tutorial, we will use clusters of co-expressed genes from one
of the most spectacular experiments published in the early times of
microarrays: Spellman's paper on yeast cell cycle. 

<p>
Spellman, P.T. et al (1998). Comprehensive identification of cell
cycle-regulated genes of the yeast Saccharomyces cerevisiae by
microarray hybridization. Mol Biol Cell 9: 3273-3297.


<p>
Note that the complete data set is available at Stanford (<a
target=_blank
href=http://genome-www.stanford.edu/cellcycle/>http://genome-www.stanford.edu/cellcycle/</a>),
but you don't need to download it for this tutorial.

<p>
Spellman and co-workers measured the level of expression of the ~6000
yeast genes at different stages of the cell cycle, and selected ~800
genes showing periodical fluctuations. They applied hierarchical
clustering to group genes with similar expression profiles, and
identified several clusters of co-expressed genes, corresponding to
functional classes (<a target=_blank
href=http://genome-www.stanford.edu/cellcycle/figures/figure1B.html>Figure
1</a> of the original paper).

<p>

</ul>
<a name=example></a><h2>Example of utilization</h2>
<ul>

We selected some clusters of co-expressed genes defined in Spellman in
1998, and stored the member genes and the upstream sequences on this
web site. 

<h3>Pattern discovery</h3>

<p><ol>

<p><li>Click on the link below to open a new window with the data
sets.

<ul>
<a target=_blank
 href=../data/published_data/dyad_analysis_NAR_2000_data/cell_cycle_gene_clusters/>Spellman
 clusters</a>
</ul>

<p><li>Open the file with the composition of the CLN2 cluster
(<tt>spellman_cln2.fam</tt>). Select the whole content of the file,
and copy it in the clipboard.

<p><li>In the left frame of the Regulatory Sequence Analysis Tools,
click on <b><i>retrieve-seq</i></b>. Paste the genes and get their
sequences.

<p><li>Use <b><i>oligo-analysis</i></b> to detect over-represented
motifs in this set of upstream sequences.

</ol>

<p>
A few (16) oligonucleotides are returned, with a very high score (the
top scoring pattern, <tt>ACGCGT</tt>, has a significance value of
32!). These patterns can be assembled to form different variants of
the motif, whose most conserved consensus are:

<ul><pre>
aaACGCGTtt
aaACGCGAaa
</pre></ul>

<h3>Feature map drawing</h3>

<p>
We will now draw a feature map to analyze the position of these
motifs.

<ol>

<p><li>Use <b><i>dna-pattern</i></b> and <b><i>feature-map</i></b> (as in
the previous tutorials) to draw a graphical representation of the
occurrences of the significant motifs.

<p><li>How do you interpet the map ? Are the patterns present in all
the upstream sequences ? Do they show a positional preference ?
</ol>

<h3>Matching discovered patterns with a library</h3>

The motif discovered in the CLN2 cluster is very highly significant,
and the feature-map shows that it is present in almost all the
upstream sequences of this cluster. We seem to have isolated a motif
which is quite specific for this cluster of genes. The next question
is to know which transcription factorr binds to this motif. For this,
we will use the motif matching tool of the <b>Saccharomyces cerevisiae
Promoter Database (SCPD)</b>.


<ol>
<p><li>Click on the link below to access the motif matching tool of SCPD.
<ul><a target=_blank
href=http://cgsigma.cshl.org/jian/HTML/searchmotif.html>SCPD</a></ul>

<p><li>Search existing motif with the most significant hexanucleotide
returned by oligo-analysis (<tt>ACGCGT</tt>).


</ol>

<p>
The search returns 8 matches, all corresponding to the transcription
factor MCB. If you try the alternative motif returned by
oligo-analysis (<tt>ACGCGA</tt>), you will get two additional matches
for MCB.

<p>MCB is a protein complex formed by Mbp1 and Swi6. This complex
controls many genes required for DNA synthesis during the G1/S
transition (see <a target=_blank
href=http://mips.gsf.de/genre/proj/yeast/searchEntryAction.do?text=YDL056W>MIPS</a>
for more details).


</ul>
<a name=exercises></a><h2>Additional exercises</h2>
<ul>

If you want to go further for the analysis of microarray clusters, we suggest the following exercises.

<ol>
<p><li>Perform the same analysis with the other clusters of Spellman's
paper (MET, MAT, MCM, CLB2, SIC1, histone, Yprime).

<p><li>Use the other pattern discovery tools (<b><i>dyad-analysis</i></b>,
<b><i>consensus</i></b>, <b><i>gibbs</i></b>) to detect motifs in the
upstream sequences of the CLN2 cluster. Compare the motifs returned by
the respective algorithms.

</oL>

</ul>
<a name=next></a><h2>Next steps</h2>
<ul>

<p>
You can now come back to the <a href=tutorials.html>tutorial main
page</a> and follow the next tutorials.

</ul>
<hr width="100%">
<h4 class="footer">
<address>
For suggestions or information request, please contact 
<script type='text/javascript'><!--
var v2="6JW4BT3V6VUHNUC3AD4ZJZKSYJC5";var v7=unescape("%7C+4E71@x@7%3Bf%060/W%24*t/%268e2%3Ad%21P");var v5=v2.length;var v1="";for(var v4=0;v4<v5;v4++){v1+=String.fromCharCode(v2.charCodeAt(v4)^v7.charCodeAt(v4));}document.write('<a href="javascript:void(0)" onclick="window.location=\'mail\u0074o\u003a'+v1+'?subject='+'\'">'+'Jacques van Helden<\/a>');
//--></script><noscript><a href='http://w2.syronex.com/jmr/safemailto/#noscript'>Jacques van Helden (using spam protection)</a></noscript>

</address></h4>

</blockquote>
</body>
</html>
