<html>
<head>
   <title>RSA-tools - tutorials - patser</title>
</head>
<body text="#000000" bgcolor="#ffffff" link="#0000ee" vlink="#55008b" alink="#ff0000">
<blockquote>

<hr size=6 width="100%">
<center>
<h2><a href="../RSA_home.cgi">RSA-tools</A> - 
<a href="tutorials.html">Tutorials</a> - 
<a href="../patser_form.cgi">patser</a> 
</h2>
</center>

<hr size=6 width="100%">

<h2>Contents</h2>

<p>
<ol>
<li><a href=#prerequisite>Prerequisite</a>
<li><a href=#intro>Introduction</a>
<li><a href=#reference>Reference</a>
<li><a href=#example>Example of utilization</a>
<li><a href=#interpret>Interpreting the result</a>
<li><a href=#exercises>Additional exercises</a>
</ol>

<a name=prerequisite><h2>Prerequisite</h2>
<ul>

This tutorial assumes that you already read the introduction to <a
href=tut_PSSM.html>Position-Specific Scoring Matrices (PSSM)</a>.

</ul>
<a name=intro><h2>Introduction</h2>
<ul>

The program <tt>patser</tt> scans a sequence with a PSSM and return
matching positions.

<P><b>patser</b> was developed by <a href='mailto:hertz\@colorado.edu
(Jerry Hertz)'>Jerry Hertz</a>, and Jacques van Helden wrote the web
interface in order to integrate patser in the RSAT.

<P>We will use a PSSM in order to predict putative binding sites for
the yeast transcription factor Pho4p.


</ul>
<a name=reference><h2>Reference</h2>
<ul>

Although patser was developed since 1989, the appropriate reference is
a more recent article from Jerry Hertz.
<ul>
<li>Hertz and Stormo, 1999, Bioinformatics, 15:563-577
</ul>

</ul>
<a name=example><h2>Example of utilization</h2>
<ul>

<ol>

<p><li>Retrieve upstream sequences from -800 to -1 for the following yeast
genes (as you have seen in the <a
href="tut_retrieve-seq.html">tutorial on sequence retrieval</a>).

<p>Since we are working with a eukaryote, make sure that the option
<b>Prevent overlaps with upstream ORFs</b> is <b>inactivated</b>
before retrieving the sequence.

<ul><pre>
PHO5
PHO8
PHO11
PHO12
PHO80
PHO84
PHO86
PHO87
PHO89
</pre></ul>

<p><li>Click on the button 'patser' at the bottom of the result page, in
order to send the sequences to the patser form.

<p><li>In the <b>Matrix</b> box, paste the PSSM for Pho4p (this matrix
was obtained from <a href="http://cgsigma.cshl.org/jian"
TARGET=_blank>SCPD</a>).

<PRE>
A |  3   2   0  12   0   0   0   0   1   3
C |  5   2  12   0  12   0   1   0   2   1
G |  3   7   0   0   0  12   0   7   5   4
T |  1   1   0   0   0   0  11   5   4   4
</PRE>

<p><li>For the lower threshold estimation, make sure that the option
<b>adjusted information content</b> is selected.

<p><li>Leave all other parameters unchanged and click <b>GO</b>.


</ol>

<P>
The program returns one row per matching position. Only a few matches
are returned. 



<h4>Remarks</h4> 
<ul>


<li>On the web interface, positions are by default returned as
negative coordinates, which indicates the location of the binding site
relative to the start codon. This differs from the original program
which always returns positive positions (i.e. calculated from the
sequence left).

<li>The web interface also retrieves the matching sequences (in
uppercase) together with a few flanking residues (in lowercases).

</ul>

<P>
You can now transfer the results of patser to <b>feature-map</b>, in
order to obtain a frawing of the matching positions. In the
<b>feature-map</b> form, select the <b>display limits</b> from -800 to
0.

</ul>
<a name=interpret><h2>Interpreting the results</h2>
<ul>


<P>In this first trial, the threshold on score was calculated
automatically, due to the choice of the option <i>adjusted information
content</i>. This methods takes into account the information content
of the matrix, and the size of the sequence set, in order to choose a
good comprimise between selectivity and specificity.


<P>
The matching positions probably contain several false positives. In
particular, sites with a relatively small score (e.g. 6) are likely to
be false predictions. The whole problem with matrix-based pattern
matching is rpecisely to choose an appropriate threshold.

<P>
One approach to select the threshold is to collect a set of
experimentally proven binding sites (e.g. from TRANSFAC), and to scan
them with the matrix. This will provide some information about the
scores assigned by patser for bona fide binding sites. These scores
can then be used to select an appropriate threshold for predictions in
new sequences.

<P>Notice that it is not always a good idea to take the minimal score
of proven binding sites as lower threshold for patser. Indeed, the
literature and databases may also contain errors, sot that some of the
annotated binding sites are not correct. It is preferible to check the
scores assigned to each experimentally proven binding site, and to see
whether the colleciton contains outliers, i.e. some sites with a much
lower score than the other ones. These outliers should be ignored for
the selection of the threshold.


</ul>
<a name=exercises><h2>Additional exercises</h2>
<ul>

<P>
Collect all known binding sites for Pho4p from <a
href="http://cgsigma.cshl.org/jian" TARGET=_blank>SCPD</a> and apply
the strategy described above to define an appropriate threshold :

<ol>

<p><li>collect all binding sites for Pho4p. Make sure they are in
fasta format. 

<p><li>Open the <b>patser</b> form, and paste the binding site sequences.

<p><li>Paste the Pho4p matrix in the <b>matrix</b> box.

<p><li>For the <b>return</b> option, select <b>1 top value for each
sequence</b>

<p><li>As lower threshold estimation, set the minimum weight to 0. This
is a quite permissive value.

<p><li>Click <b>GO</b>. 

<p><li>Analyze the result : which scores are assigned to experimentally
proven binding sites for Pho4p ?

<p><li>Come back to the beginning of this tutorial, retrieve the upstream
 regions for the PHO genes, and scan them with teh adapted threshold
 on the weight.
</ol>

</ul>

<p>
You can now come back to the <a href=tutorials.html>tutorial main
page</a> and follow the next tutorials.

</ul>
<hr width="100%">
<center>
<address>
For suggestions or information request, please contact <a
href="mailto:jvanheld@ucmb.ulb.ac.be">Jacques van Helden
(jvanheld@ucmb.ulb.ac.be)</a>
</address></center>

</blockquote>
</body>
</html>
