<html>
<head>
   <title>RSA-tools - tutorials - template</title>
<link rel="stylesheet" type="text/css" href = "../main.css" media="screen">
</head>
<body class="info">
<blockquote>


<center>
<h2><a href="../RSAT_home.cgi">RSA-tools</A> - 
<a href="tutorials.html">Tutorials</a> - 
<a href="../peak-motifs_form.cgi">peak-motif</a> 
</h2>
</center>



<h2>Contents</h2>

<p>
<ol>
<li><a href=#intro>Introduction</a>
<li><a href=#coord>Get ChIP-seq peak coordinates</a>
<li><a href=#seq>Extract peak sequences</a>
<li><a href=#peak_motif>Discover over-represented motifs</a>
<li><a href=#interpret>Interpreting the result</a>
</ol>


<a name=intro></a><h2>Introduction</h2>
We will follow a protocol which combines various Web resources (databases, bioinformatics analysis tools, genome browser) in order to analyze genome-wide binding profiles of some transcription factors, obtained with the ChIP-seq method.

<a name=coord></a><h2>Get ChIP-seq peak coordinates</h2>
 
    <p><b>Goal:</b> Identify the dataset corresponding to the article by Chen et
      al., 2008 (Pubmed ID: 18555785) in the <a target='_blank'
						href='http://www.ncbi.nlm.nih.gov/gds/'>GEO datasets</a> and
      Retrieve the data for the <b>Sox2</b> experiment.</p>

    
    <ol>
      <li><p>Open a connection to the <a target='_blank'
					 href='http://www.ncbi.nlm.nih.gov/pubmed/'>Pubmed</a>
	  database.</p></li>

      <li><p>In the text box, enter the title of the article:

	<p><ul><i>Integration of external signaling pathways with the core
	      transcriptional network in embryonic stem cells</i></ul></p> 

	<p>This should give a signel result. If this is not the case, you
	  can select the publication on the basis of
	  its PubMed
	  ID <a target='_blank'
		href='http://www.ncbi.nlm.nih.gov/pubmed?term=18555785[uid]'>18555785</a>

      </p></li>

      <li><p>On the right side of the Pubmed record, under the title <b>All links from this record
	  </b>, click the link <b>GEO DataSets</b>.</p></li>

      <li><p>This opens the record <a target='_blank'
				      href='http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE11431'>GSE11431</a>
	  with the title <i>Mapping of transcription factor binding sites
	    in mouse embryonic stem cells</i>. </p>

	<p><ul>Note: the title of the record differs from the title of
	    the article, which makes it somewhat difficult to identify
	    a record by browsing the GEO datasets alone. The easiest
	    way to go from an article to the corresponding records is
	    generally to use the direct link from PubMed to GEO
	    DataSets as we did.</ul></p>


      </p></li>
      
      <li><p>In the GEO database, the identifiers with prefix GSE
	  denote series of experiments. Indeed, Chen et al. (2008)
	  published ChIP-seq results for various transcripiton
	  factors, so that the series contains 16 samples in
	  total.</p></li>

      <li><p>The bottom of the GSE record provides the list of samples
	  (identifiers starting with GSM). Click on the link
	  corresponding to ES_Oct4 (sample ID <a target='_blank'
	  href='http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM288346'>GSM288346</a>.
      </p></li>

      <li><p> Read the information disponible about this sample. The
	  bottom of the record provides links to the data sets at
	  various processing stages:
	  <ul>
	    <li>The <b><i>SRA</i></b> files contain the raw sequences
	      produced by the massively parallel sequencer. For the Oct4
	      sample, there are 4 files each containing more than 5
	      million reads (~150 Megabases of sequence per file).</li>
	    <li>The <b><i>bed</i></b> file <a target='_blank'
	    href='ftp://ftp.ncbi.nih.gov/pub/geo/DATA/supplementary/samples/GSM288nnn/GSM288346/GSM288346%5FOct4%2Ebed%2Egz'>GSM288346_Oct4.bed.gz</a>

	    contains the location of the reads. Beware, this is a big
	    file (33.4Mb), it can take some time to download it.</li>
	    <li>The <b><i>txt</i></b> file <a target='_blank'
	    href='ftp://ftp.ncbi.nih.gov/pub/geo/DATA/supplementary/samples/GSM288nnn/GSM288346/GSM288346%5FES%5FOct4%2Etxt%2Egz'>SM288346_ES_Oct4.txt.gz</tt></a>
	    contains a set of peaks selected by Chen et
	    al. (2008). Those reads are described in a tab-delimited
	    text file, the first column indicates the chromosome, the
	    second column the start and the third column the end of
	    each peak.
	  </ul>
      </p></li>
      
    </ol>

<a name=seq></a><h2>Extract peak sequences</h2>

<p>The Galaxy Web server offers a wide variety of tools for
      analyzing genomic data. It allows to quickly retrieve sequences from a coordinate file (eg. BED file) </p>

<ol>
      <li><p>Open a connection to the Galaxy server (<a target='_blank'
          href='http://main.g2.bx.psu.edu/'>http://main.g2.bx.psu.edu/</a>).</p></li>
      <li><p>In the menu at the left of the window, click <b>Get Data > Upload File</b>.</p></li>
      
      <li><p>You can either upload a file from your computer
	(click <b>Browse</b> besides the <b>File</b> text box) or from
	a Web server (type a link to the file in the <b>URL/Text</b>
	box).
	<ul>For this exercise, we will get the "txt" file (actually in BED format) from GEO. In
	  the <b>URL/Text</b> box, paste the URL of the Oct4 sample:
	  <pre>ftp://ftp.ncbi.nih.gov/pub/geo/DATA/supplementary/samples/GSM288nnn/GSM288346/GSM288346%5FES%5FOct4%2Etxt%2Egz</pre>

	</ul>
      </p></li>
      
      <li><p>In the <b>Genome</b> pop-up menu, select <i>Mouse Feb 2006
      (NCBI36/mm8)</i>. Tip: he genome is selected if you simply
      type <i>mm8</i> when the menu is selected.</p></li>
      
      <li><p>Leave the other options to their default value, and
	click <b>Execute</b>. The upload may take several
	minutes. When the file will be uploaded, the yellow box on the
	right side will turn to green. </p></li>

      <li><p>Click on this box and make sure that the format is BED and
	the genome is mm8. How many ChIP-seq "regions" are present in this
	file ?</p></li>
      
      <li><p>Click on the disk icon in this box to download this sequence file. Save it on your computer.</p></li>
    </ol>


<a name=peak_motif></a><h2>Discover over-represented motifs</h2>
<p>The program peak-motifs performs a complete procedure of motif analysis in the Chip-seq peaks.
It automatically run through the following steps:</p>
<ul>
<li>Statistics on the input peak sequences (length, mono- and di-nucleotides composition, links to genome browser to vizualise the peaks)</li>
<li>Motif discovery with multiple algorithms (oligo-analysis, position-analysis, dyad-analysis, local-words)</li>
<li>Comparison of discovered motifs with collections of known motifs</li>
<li>Scan of input sequence peaks with discovered motifs, binding site positional profiles</li>
<li>Vizualisation of binding sites in a genome browser (UCSC)</li>
</ul>

<p>This part of the tutorial will guide you into using peak-motif on the sequences retrieved at the previous step.</p>

<ol>
<li>Open a new page to <a href="../peak-motifs_form.cgi" target="_blank">the peak-motif form</a> </li>
<b>Peak sequences box</b>
<li>Enter a <b>title</b> for this analysis: <i>Oct4 dataset Chen 2008</i></li>
<li>Click on the "browse" button to select your peak sequence file. You may alternatively paste sequences in the box.<br/>
Warning: make sure you upload a file containing <i><u>peaks</u></i> (usually less than 100 000 sequences) and <i><u>not the raw reads</u></i> (usually millions of sequences)</li>

<b>Reduce input peak sequences box</b>
<li>This section aims at facilitating the edition of input sequences. It allows to restrict the analysis to X top sequences or restrict the analysis to the center of the peaks. 
For this first try, we will not edit the input sequences, leave this section blank.</li>
<b>Change motif discovery parameters box</b>
<li>This section contain the parameters for the motif discovery step. We will first keep the default settings, 
using the programs <i>oligo-analysis</i> and <i>position-analysis</i> to discover over-represented motifs and motifs with positional biais.</li>

<b>Compare motifs box</b>
<li>The discovered motifs are compared to databases of known motifs. We directly support various databases like JASPAR, Uniprobe,...</li>
<li>Users may also upload here private collection of matrices e.g. transfac </li>
<li>Keep <i>JASPAR core Vertebrates</i> checked, and <b>also check </b><i>JASPAR PBM (UNIPROBE) Mouse</i>, since our dataset was obtained from mouse.</li>

<b>Locate motifs and export as UCSC custom track</b>
<li>Keep the <i>Search putative binding sites</i> checked</li>
<li><b>Select "Yes,</b> sequences from fetched with Galaxy". This will recalculate the genomic coordinates of the found binding sites, and make a custom UCSC track to vizualise the results in this genome browser.</li>
<li>Enter your <b>email adress</b> (or a fake adress like nobody@nowhere, but you will not be notified of the results) and click <b>GO</b></li>
</ol>


<a name=interpret></a><h2>Interpreting the result</h2>
<p>A link to the result should appear on the new page. The results appear progressively, to enable the users to analyse their results more quickly.
</p>
<ol>
<li>Click on this link to see the results.</li>
<li>TO BE CONTINUED...</li>


</ol>
<a name=exercises></a><h2>Additional exercises</h2>
<ul>
<li>Upload your own collection of motifs</li>
<li>Upload a reference motif</li>
<li>Change motif-discovery parameters by clicking "dyad-analysis"</li>

</ul>

</ul>

<hr width="100%">
<h4 class="footer">
<address>
For suggestions or information request, please contact 
<script type='text/javascript'><!--
var v2="6JW4BT3V6VUHNUC3AD4ZJZKSYJC5";var v7=unescape("%7C+4E71@x@7%3Bf%060/W%24*t/%268e2%3Ad%21P");var v5=v2.length;var v1="";for(var v4=0;v4<v5;v4++){v1+=String.fromCharCode(v2.charCodeAt(v4)^v7.charCodeAt(v4));}document.write('<a href="javascript:void(0)" onclick="window.location=\'mail\u0074o\u003a'+v1+'?subject='+'\'">'+'Jacques van Helden<\/a>');
//--></script><noscript><a href='http://w2.syronex.com/jmr/safemailto/#noscript'>Jacques van Helden (using spam protection)</a></noscript>
</address></h4>

</blockquote>
</body>
</html>
