<!DOCTYPE html>
<meta charset="utf-8">

<html>		
  <head>
    <title>RSA-tools - tutorials - peak-motifs</title>
    <link rel="stylesheet" type="text/css" href = "../course.css" media="screen"/>
    <script src="http://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
    <script type="text/javascript" src="syntaxhighlight/shCore.js"></script>
    <script type="text/javascript" src="syntaxhighlight/shBrushBash.js"></script>
    <link type="text/css" rel="stylesheet" href="syntaxhighlight/shCore.css"/>
    <link type="text/css" rel="stylesheet" href="syntaxhighlight/Shthemedefault.css"/>
    <script type="text/javascript">
      SyntaxHighlighter.config.clipboardSwf = 'syntaxhighlight/clipboard.swf';
      SyntaxHighlighter.all();
    </script>
    <meta charset=utf-8 />
  </head>

<!-- Define some additional styles for the dynamic elements of this Web page -->
	<style>
		
		.pic_tab {
			height: 500px;
			width: 400px;
			display: inline-block;
			border: 1px solid blue;	
			float: left;
			clear: both;
			margin-left: 20px;
		}
		
		
		.div_tab {
			display: inline-block;	
		}
		
		
		#logo_tree {
			height: 17%;
			width: 70%;	
			border: 1px solid blue;	
			display: inline-block;
						float: left;
			clear: both;
		}
		
		
		#branch_cons_tab {
			height: 450px;
			width: 600px;
			display: inline-block;	
			border: 1px solid blue;	
						float: left;
			clear: both;
		}
		
		
		#logo_forest_threshold {
			height: 17%;
			width: 70%;	
			border: 1px solid blue;	
			display: inline-block;
			float: left;
			clear: both;
		}
		
		
		.fig_button {
	        margin-top: 3px;
		width: 80px;
		heigth: 60px;
	        border: 1px;
		position: relative;
	        padding: 1px;
	        font-size:14px;
	        color: #ffffff;
		background:#cc0000;
		border-radius:5px;
	        cursor: pointer;
	        display: inline-block;
	        text-align: center;
		}
		
		
		.all_buttons {
	        margin-top: 3px;
			width: 80px;
			heigth: 60px;
	        border: 1px;
			position: relative;
	        padding: 1px;
	        font-size:14px;
	        color: #ffffff;
		background:#000000;
		border-radius:5px;
	        cursor: pointer;
	        display: inline-block;
	        text-align: center;
		}
		
		
		.fig {
			clear: both;
			float: left;
			display:none;
		}
		
		.fig p {
			clear:both;
		}
		

		#logo_tree_no_threshold {
			height: 17%;
			width: 70%;
			clear: both;
			float: left;
			display: block;
			border: 1px solid blue;
		}
		
		
		#soct {
		    width: 40%;
			height: auto;
			float: left;
			clear: both;
			display: block;
		}
		
		
		.branch_motifs {
			width: 75%;
			height: auto;
		}
		
		
		.cons_tree {
			width: 80%;
			height: auto;
			
		}
		
		
		.network_insects {
			width: 65%;
			height: auto;		
		}		


		.network_vertebrates {
			width: 100%;
			height: auto;	
				
		}	
		

		#insect_logo_tree {
			width: 95%;
			height: auto;	
				
		}	
	
	
		#vertebrate_logo_tree {
			width: 75%;
			height: auto;	
				
		}
		

		.sc_2{
			display: block;
			clear: both;
		}
		
	</style>

  <body>
    
    <center>
      <h1><a href="../RSAT_home.cgi">RSA-tools</A> - 
	<a href="tutorials.html">Tutorials</a> - 
	<a href="../matrix-clustering_form.cgi">matrix-clustering</a> 
      </h1>
    </center>
    
    <a name='toc'></a><h2>Contents</h2>
    
    <p>
      <ol>
	<li><a href='#prerequisite'>Prerequisite</a>
	<li><a href='#intro'>Introduction</a></li>
	<li><a href='#study_case'>Study case</a></li>
	<li><a href='#test_sets'>Test sets</a></li>
	<li><a href='#parameters'>Tuning <i>matrix-clustering</i> parameters</a></li>
	<li><a href='#interpret'>Interpreting the result</a></li>
	<li><a href='#exercises'>Additional exercises</a></li>
	<li><a href='#references'>References</a></li>
      </ol>
    </p>
      
    <hr><a name='prerequisite'></a><h2>Prerequisite</h2>
    
    <p>This tutorial assumes that you are familiar with the concepts
      developed in the following parts of the theoretical course.
      <ol>
	<li><a href='../course/pdf_files/01.4.PSSM_theory.pdf'>PSSM theory</a></li>
	<li><a href="http://rsat.ulb.ac.be/rsat/course/pdf_files/12.peak-motifs.pdf">Motif comparison</a></li>	
      </ol>
    </p>

    <p>It is better to follow the corresponding tutorials before this one.
      <ol>
		<li><a href='tut_PSSM.html'>Position-specific scoring matrices</a>.</li>
		<li><a href="tut_peak-motifs.html"><i>peak-motifs</i></a>: discover cis-regulatory motifs and predict putative TFBS from a set of peak sequences identified by high-throughput methods such as ChIP-seq.</li>
      </ol>
    </p>

    <hr><a name='intro'></a><h2>Introduction</h2>

    <p>The
      program <a href='../matrix-clustering_form.cgi'><span class="program">matrix-clustering</span></a>
      enables to discover and align groups of similarities among motif
      collections displaying the results with different
      motif-representation formats.
    </p>
      
    <p>Transcription factor binding motifs
      (<span class="concept">TFBM</span>) are classically represented
      either as consensus strings (stric consensus, IUPAC or regular
      expressions), or as position-specific scoring matrices
      (<span class="concept">PSSM</span>). </p>

    <p> Thousands of curated TFBM are available in specialized
      databases (<a target="_blank"
      href="http://jaspar.genereg.net/">JASPAR</a>, <a target="_blank"
      href="http://regulondb.ccg.unam.mx/">RegulonDB</a>, <a target="_blank"
      href="http://www.gene-regulation.com/pub/databases.html">TRANSFAC</a>,
      etc). These PSSMs were traditionally built from collections of
      transcription factor binding sites (TFBS) obtained by various
      experimental methods (e.g. EMSA, DNAse footprinting, SELEX).</p>

    <p>TFBMs can also be discovered <i>ab initio</i> from genome-scale
      datasets: promoters of co-expressed genes, ChIP-seq peaks,
      phylogenetic footprints, etc. 
    </p>

    <p>Motif collections (databases as well as ab initio motif
      discovery results) sometimes contain groups of similar motifs,
      for different reasons: curation of alternative motifs for a same
      TF; homologous proteins sharing a particular DNA binding domain,
      motifs discovered with analytic workflows combining several
      algorithms
      (e.g. RSAT <span class="program"><a href="../peak-motifs_form.cgi">peak-motifs</a></span>,
      or <span class="program"><a target="_blank"
      href="http://meme.nbcr.net/meme/cgi-bin/meme-chip.cgi">MEME-chip</a></span>).</p>

    <p>The RSAT
      tool <span class="program"><a href="../matrix-clustering_form.cgi">matrix-clustering</a></span>
      addresses these needs, and includes several interesting features
      which will be illustrated in this tutorial.
    </p>
    
    <p>
      <ol>
	<li>For the computation of inter-matrices distances, support
	  for a large series of alternative metrics (Pearson
	  correlation, Euclidian distance, SSD, Sandelin-Wasserman,
	  logo dot product, and length-normalized version of these
	  scores).</li>

	<li>Possibility to select a custom combination between several
	  of these similarity metrics, in order to compute an
	  integrative threshold.</li>

	<li>The set of input motifs is split into separate clusters,
	  each of which canbe displayed in user-interactive ways.</li>

	<li>User-friendly display of motif trees with aligned logos
	  and consensuses.</li>

	<li>At each level of the hierarchical tree, all the descendent
	  matrices are aligned (multiple alignement), and a merged
	  motif is computed (<span class="concept">branch
	  motif</span>), which can be represented schematically as
	  a <span class="concept">branch logo</span> or
	  a <span class="concept">branch consensus</span>.</li>
      </ol>
    </p>

    <p>In this tutorial, we explain how to tune the parameters and
      interpret of results of the
      <span class="program">matrix-clustering</span>. 
      <ol>	
	<li><b>Motif redundancy</b>: examples in motif-discovery
	  results and in motif databases.</li>
	<li><b>Thresholds</b>: setting a combination of similarity
	  measures values as a threshold to define the gruops of
	  similarities.</li>
	<li><b>Impact of parameters</b>: some example showing how
	  changing the values of the parameters can affect cluster
	  composition or tree topology.</li>
      </ol>
    </p>
    
    <hr><a name='study_case'></a><h2>Study cases</h2>
    
    <h3>Study case 1</h3>
    
    <p><b>Goal:</b> clustering a set of partly redundant motifs
      discovered by various algorithms.</p>
    
    <p><b>Data set:</b> To illustrate the use of motif clustering
      to filter out redundancy, we will analyze a set of motifs
      discovered with the
      workflow <span class="program"><a href="../peak-motifs_form.cgi">peak-motifs</span></a>,
      which combines several motif discovery approaches:
      <ol>
	<li><a href="../oligo-analysis_form.cgi"><span class="program">oligo-analysis</span></a></li>
	<li><a href="../position-analysis_form.cgi"><span class="program">position-analysis</span></a></li>
	<li><a href="../local-word-analysis_form.cgi"><span class="program">local-word-analysis</span></a></li>
	<li><a href="../dyad-analysis_form.cgi"><span class="program">dyad-analysis</span></a></li>
      </ol>

    <p>We ran these algorithms on a set of peak sequences obtained
      by pulling down genomic regions bound by the transcription
      factor Oct4 in mouse ES cells.  This experiment had been
      performed in the context of a wider study, where Chen and
      colleagues characterized the binding location of 12
      transcription factors involved in mouse embryonic stem cell
      differentiation (<a href='#ref_chen_2010'>Chen et al.,
	2008</a>).
    </p>
    
    <p>In this example, 
      <ol>
	<li>We will run <a href="../matrix-clustering_form.cgi"><span class="program">matrix-clustering</span></a>
	  with its default parameters, which were tuned to generally
	  provide a suitable result; </li>
	<li>We will then run the same analysis without any
	      threshold on matrix similarity, and with alternative
	  threshold values, in order to analyze the impact of this
	  crucial parameter on the result.</li>
	<li>Finally, we will run a <span class="concept">negative
	    control</span> by randomizing the motifs (random
	  permutations of the columns for each matrix) and
	  running the clustering. In principle, the clustering
	  algorithm should assign each matrix to a singleton
	  (cluster containing only one element).</li>
      </ol>
    </p>

    <h3>Study case 2</h3>

    <p class="attention"><b>Beware</b>: this tutorial requires to let
      the program running for a while, and mobilize important computer
      resources. It is thus only available for the command line
      version of the tools.</p>

    <p><b>Goal:</b> cluster all motifs from specific sections
      (Vertebtrates, Insects) of the JASPAR database
      (<a href=''>Sandelin et al., 2004</a>), in order to identify
      groups of similar motifs, which may be bound by TFs of the same
      family (e.g. several factors of the Hox family).</p>

    <p>Note that in some cases factors with unrelated DNA-binding
      domain may recognize partly similar motifs (e.g. yeast Pho4p and
      Met4p recognize a motif centered on CACGTG, but differing by the
      flanking residues).
    </p>	
    
    <hr><a name='test_sets'></a><h2>Test sets</h2>
    
    <ul>
      
      <li>
	<h2>Test set 1</h2>
	<p>The set of 21 motifs discovered with <i>peak-motifs</i> in
	  Oct4 ChIP-seq is available on the section <i>data</i> of
	  RSAT.</p>
	
	<p><a href="../demo_files/demo_files//peak-motifs_Oct4_matrices.tf">Discovered
	    motifs in Oct4 ChIP-seq peaks</a></p>
      </li>
      
      <li>
	<h2>Test set 2</h2>
	<p>JASPAR vertebrate and insect motifs are available on the section <i>data</i> of RSAT.</p>
	<p><a target="_blank" href="../data/motif_databases/JASPAR">JASPAR motifs</a></p>
      </li>
      
    </ul>


    <hr><a name='parameters'></a><h2>Tuning <i>matrix-clustering</i> parameters</h2>
    
    <ul>
      <br>
      <li>
	
	<h2>Study case 1</h2>
	<br>
	
	<h3><i>Running matrix-clustering without thresholds</i></h3>
	
	<div class='protocol'>
	  <p><b>NOTE:</b> This section will be done using the command line</p>
	  
	  <ol>
	    <li>
	      <p>Open a terminal.<p>
	    </li>
	    <li>
	      <p>Move to a directory where the result will be saved.</p>
	    </li>
	    <li>
	      <p>Execute this command: </p>
      <pre class="brush:bash;">
matrix-clustering -v 2 \
  -i $RSAT/public_html/demo_files/peak-motifs_Oct4_matrices.tf \
  -matrix_format tf -cons -hclust_method average \
  -label name,consensus \
  -o results/peak-motifs_Oct4_No_thresholds/Oct4_analysis_No_thresholds
</pre>
	      <p>You can read the help of the parameters used in matrix-clustering typing: <i>matrix-clustering -h</i></p>
	      <ul>
		<li>
		  <p><b>Note:</b> You can change the input motif with the parameter -i.  Example: <i> -i folder/motif_file.tf -format tf</i></p>
		</li>
		<li>
		  <p><b>Note:</b> You can change the output folder name with the parameter -o. Example: <i> -o folder/prefix </i></p>
		</li>
		<li>
		  <p><b>Note:</b> You can change the clustering method with the parameter -hclust_method [average|complete|single]. Example: <i> -hclust_method single </i></p>
		</li>
	      </ul>	
	    </li>
	  </ol>
	</div>  
      </li>  
      
      <br>
      
		<h3><i>Running matrix-clustering with thresholds</i></h3>
      
		<div class='protocol'>
			<p><b>NOTE:</b> This section will be done using the command line</p>
			
			<ol>
				<li>
					<p>Open a terminal.<p>
				</li>
				<li>
					<p>Move to a directory where the result will be saved.</p>
				</li>
				<li>
					<p>Execute this command: </p>
					<p><b>matrix-clustering -v 2 -i $RSAT/public_html/demo_files/peak-motifs_Oct4_matrices.tf -matrix_format tf -cons -hclust_method average -lth Ncor 0.4 -lth cor 0.6 -lth w 5 -label name,consensus -o results/peak-motifs_Oct4/Oct4_analysis</b></p>
					<p>You can read the help of the parameters used in matrix-clustering typing: <i>matrix-clustering -h</i></p>
					<ul>
						<li>
							<p><b>Note:</b> You can change the lower threshold values with the parameter -lth [cor|Ncor|w]. Example: <i> -lth Ncor 0.6 -lth cor 0.75 -lth w 5 </i></p>
						</li>
					</ul>	
				</li>
			</ol>
		</div>  

	<br>
		<li>
			
			<h2>Study case 2</h2>
			<br>
			
			<h3><i>Running matrix-clustering with thresholds to identify groups of TFs belonging to the same family</i></h3>
      
			<div class='protocol'>
				<p><b>NOTE:</b> This section will be done using the command line</p>
		
				<ol>
					<li>
						<p>Open a terminal.<p>
					</li>
					<li>
						<p>Move to a directory where the result will be saved.</p>
					</li>
					<li>
						<p>Execute the next commands: </p>
							<ul>
								<li><p>JASPAR insect core (136 motifs):</p></li>
								<p> <b>matrix-clustering -v 2 -i $RSAT/data/motif_databases/JASPAR/jaspar_core_insects_2013-11.tf -matrix_format tf -cons -hclust_method average -lth Ncor 0.5 -lth cor 0.7 -lth w 5 -label name,consensus -o results/JASPAR_insects/Jaspar_insects</b></p>
								<li><p>JASPAR vertebrate core (263 motifs):</p></li>
								<p><b>matrix-clustering -v 2 -i $RSAT/data/motif_databases/JASPAR/jaspar_core_vertebrates_2013-11.tf -matrix_format tf -cons -hclust_method average -lth Ncor 0.5 -lth cor 0.7 -lth w 5 -label name,consensus -o results/JASPAR_vertebrates/Jaspar_vertebrates</b></p>
							</ul>
						<p>Remember: you can read the help of the parameters used in matrix-clustering typing: <i>matrix-clustering -h</i></p>	
						<p><b>Note:</b>These are bigger dataset than the study case 1 and <i>matrix-clustering</i> will take much more time to analyse them.</p>
					</li>
				</ol>
			</div>  
		</li> 

 
	</ul>


<hr><a name='interpret'></a><h2>Interpreting the results</h2>

<ul>
	<br>
	<li>
			<h2>Study case 1</h2>
			<h3><i>Running matrix-clustering without thresholds</i></h3>
			<p><b>NOTE: </b>As in this example we did not set threshold values, all input motifs are forced to be aligned in a single global alignment, this means only one cluster encompassing all the motifs. </p>
			<ol>
				<li>Go to the results folder: results/peak-motifs_Oct4_No_thresholds</li>
				<li>Open in a web browser (Firefox is recommended) the file: Oct4_analysis_No_thresholds_SUMMARY.html
					<p>This file has the link to all the output files and show separately the clusters found in the analysis.</p>
					<p>You will find the next files: </p>
					<ol>
						<li><b>Summary table:</b> shows the parameters used for tha analysis, the number of input motifs and the number of clusters.</li>
						<li><b>General Alignment table:</b> shows the global alignment of all input motifs, separated on their corresponding clusters. </li>
						<li><b>Buttons with the results:</b> There is one button for each cluster  each clster analysis.
							<ul>
								<li><b>Cluster aligment table</b></li>
								<li><b>Cluster tree</b></li>
								<li><b>Cluster branch-motif table</b></li>
							</ul>
						</li>
						<li><b>List with additional files:</b> output files with general information of all motifs (not cluster specific)
							<ul>
								<li><b>Distance table</b></li>
								<li><b>Internal node attributes table</b></li>
								<li><b>Logo Forest</b></li>
								<li><b>Pairwise comparison table</b></li>
								<li><b>Motif description table</b></li>
								<li><b>Consensus tree</b></li><br>
							</ul>
						</li>
					</ol>
				</li>
				<li>
					Click on <i>cluster_1</i> button and display the <i>logo tree</i>. 
					<p>In this logo tree you can see the logo aligment in direct an reverse orientation. 
						It must be noted that the middle motifs in the logo are aligned relative to the positions <b>ATGCAAT</b>,<br>
						however the motifs at the top of the logo tree are C-rich and they are not matching well with the rest of the alignment,  
						whilst the same case occurs with the bottom motifs at the logo tree, the A-rich motifs. <b>(Figure 1)</b></p>
					<br>
				</li>			
			</ol>
		<br>	
		<h3><i>Running matrix-clustering with thresholds</i></h3>
		<p>In this example we will study a set of motifs dicovered in a ChIP-seq experiment done for the TF Oct4 (Pou5f1).<br>Oct4 is an   essential   TF   in   cell   fate   decision,   ES   cells   and   early   embryonic
        development,   it  binds  the   canonical  sequence  <i><b>5'-­ATGCAAAT­-3'</i></b>.  In  ES   cells,  Oct4  often interacts   with   another   TF,   Sox2,   which   binds   to   an   adjacent   Sox   motif   <b><i>5'-­CATTGTA­-3'</i></b>.
        <br>Together, both TFs co­regulate specific genes. During the analysis of Oct4 or Sox2 binding peaks, the so­ called <b><i>SOCT</i></b> motif is usually found, which is a composite motif encompassing both Oct and Sox motifs. <b>(Figure 2)</b> 
		</p>
		<ol>
			<li>Go to the results folder: results/peak-motifs_Oct4</li>
			<li>Open in a web browser (Firefox is recommended) the file: Oct4_analysis_SUMMARY.html</li>
			<li>Compare the <i>Summary</i> and <i>Alignment</i> table between Oct4 analysis with and without thresholds. <b>(Figure 3)</b>
				<ul>
					<li>Observe how the number of clusters changes when you set threshold values.</li>
					<li>In the aligment table you can observe how the motifs are separated in different clusters, observe that each aligment has a different width.</li>
					<li>Some motifs are not asociated to a cluster and remain as singleton.</li>
					<br>
				</ul>
			</li>
			<li>Click the cluster buttons and display the aligment tables and logo trees.</li>
			<li>In the logo trees you can observe (for cluster with 2 or more motifs) that each branch has in blue a consensus, we called it <i>branch-consensus</i> which summarizes all the descendant motifs. <b>(Figure 4)</b></li>
			<li>Display the branch-motifs table. Ypu can donwload the motifs corresponding to these branch-motifs by clicking on the links in this table. <b>(Figure 5)</b></li>
			<li>
				Go to the list of additional files and display the <i>Logo forest</i>. Observe how the tree displayed in the example when we are not using threshold is partitioned in many trees (Forest). <b>(Figure 6)</b>
				<ul>
					<li>Some motifs are not grouped and are considered as a cluster wit a single member.</li>
					<li>Let's focus on cluster 1. We can see two groups within this cluster.<br>One group with 6 members sharing the motif <b><i>ATGCAAAT</i></b> (which actually corresponds to Oct4 motif) and the second group with five members sharing the bigger motif <b><i>TTGATATGCAAAT</i></b> (corresponding to the <i>SOCT</i> motif).</li>
					<li>By clicking on the branch-consensuses for these groups you can observe the logo. We use the program <i>compare-matrices</i> to compare these branch-motifs against the JASPAR vertebrates motifs,<br>the short motif in fact correspond to Oct4 and the other to Sox2::Pou5f1 (SOCT). <b>(Figure 7)</b></li>
					<li>This means some peaks only contain Oct4 bound, however others have Oct4+Sox2, for these reason were founded the SOCT motif.</li>
				</ul>
			</li>
			<br>
			<li>Now we are going to show the differences of the trees when we change the hierarchical clustering agglomeration rule.</li> 
				<ul>
					<li>By default the rule used in <i>matrix-clustering</i> is the <i>average</i> method, but also can be used the <i>single</i> and <i>complete</i> methods.</li>
					<li>Remember you can change this option with the parameter <i>-hclust_method</i> .</li>
					<li>Run <i>matrix-clustering</i> with the same data for each hclustering agglomeration rule.</li>
					<li>In the <i>List of file</i> menu display the file <i>consensus tree</i> for each analysis.</li>
					<li>See <b>Figure 8</b> to observe the differences in the tree topologies using the same 21 motifs as input data depending on the rulse selected. </li>
				</ul>
		</ol>

			<br>
			<div class="case1 all_buttons" id="button">Fig case 1:</div>
			<div class="fig_button" id="button_fig_1">Figure 1</div>
			<div class="fig_button" id="button_fig_2">Figure 2</div>
			<div class="fig_button" id="button_fig_3">Figure 3</div>
			<div class="fig_button" id="button_fig_4">Figure 4</div>
			<div class="fig_button" id="button_fig_5">Figure 5</div>
			<div class="fig_button" id="button_fig_6">Figure 6</div>
			<div class="fig_button" id="button_fig_7">Figure 7</div>
			<div class="fig_button" id="button_fig_8">Figure 8</div>
			<br><br>
			
			
			<!-- Figure 1 -->
			<div class="fig" id="fig_1">
				<img id="logo_tree_no_threshold" src="../images/matrix-clustering_tutorial/Oct4_no_thresholds_logo_tree.png"/>
				<p><b>Figure 1.</b> Logo tree with the 21 motifs discovered by <i>peak-motifs</i> in the Oct4 ChIP-seq peaks. As there is not a set threshold, all motifs are forced to be aligned in a single alignment.</p>
			</div>
			
			
			<!-- Figure 2 -->
			<div class="fig" id="fig_2">
				<img id="soct" src="../images/matrix-clustering_tutorial/soct.png"/>
				<p><b>Figure 2.</b> 3D model showing the cooperative binding between Sox2 and Oct4 TFs whose closely interact to bind DNA. Their recognize a composite motif called the SOCT motif (SOx+OCT).</p>
			</div>

			
			<!-- Figure 3 -->
			<div id="fig_3" class="fig">
				<div class="div_tab" >
					<ul><p><b>Oct4 ChIP-seq discovered motifs table WITHOUT thresholds</b></p></ul>
					<img class="pic_tab" src="../images/matrix-clustering_tutorial/Oct4_no_thresholds_tab.jpeg"/>
				</div>
			
				<div class="div_tab">
					<ul><p><b>Oct4 ChIP-seq discovered motifs table WITH thresholds</b></p></ul>
					<img class="pic_tab" src="../images/matrix-clustering_tutorial/Oct4_tab.jpeg"/>
				</div>
				<p><b>Figure 3.</b> Tables with the 21 motifs discovered by <i>peak-motifs</i> in the Oct4 ChIP-seq peaks andalized with <i>matrix-clustering</i> with and without setting threshold values.</p>
			</div>


			<!-- Figure 4 -->
			<div class="div_tab fig" id="fig_4">
				<ul><p><b>Cluster 1 logo tree</b></p></ul>
				<img id="logo_tree" src="../images/matrix-clustering_tutorial/logo_tree_cluster1.jpeg"/>
				<p><b>Figure 4.</b> Logo tree of the cluster 1 found in the Oct4 ChIP-seq motifs. The hierarchical tree displays the logo aligment in both orientations. For each branch is calculated a branchwise-motif. </p>
			</div>
			
			
			<!-- Figure 5 -->
			<div class="div_tab fig" id="fig_5">
				<ul><p><b>Cluster 1 branch-motifs table</b></p></ul>
				<img id="branch_cons_tab" src="../images/matrix-clustering_tutorial/Oct4_cluster1_branch_tab.jpeg"/>
				<p><b>Figure 5.</b> Branch-motif table for cluster 1. You can download the motif in TRANSFAC format or the logo in both orientations by clicking on them.</p>
			</div>
			
			
			<!-- Figure 6 -->
			<div class="div_tab fig" id="fig_6">
				<ul><p><b>Logo Forest</b></p></ul>
				<img id="logo_forest_threshold" src="../images/matrix-clustering_tutorial/logo_forest_oct4_peak-motifs.png"/>
				<p><b>Figure 6.</b> Logo Forest with the 21 motifs discovered by <i>peak-motifs</i> in the Oct4 ChIP-seq peaks. Using a combination of values as threshold (cor = 0.6; Ncor = 0.4) these motif were separated in 6 different clusters and Each one is displayed in a tree (We called <i>Forest</i> to one set of trees).</p>
			</div>
			
			
			<!-- Figure 7 -->
			<div class="div_tab fig branch_motifs" id="fig_7">
				<ul><p><b>Branch-motif analysis</b></p></ul>
				<img id="logo_forest_threshold" src="../images/matrix-clustering_tutorial/branch_consensus_analysis.jpeg"/>
				<p><b>Figure 7.</b> The logo tree of cluster one showing the branch-motifs </p>
			</div>
			
			
			<!-- Figure 8 -->
			<div class="div_tab fig branch_motifs" id="fig_8">
				<table>
					<tbody>
						<tr>
							<br><br>
							<ul><p><b>Consensus tree - Average method</b></p></ul>
							<img class="cons_tree" src="../images/matrix-clustering_tutorial/peak-motifs_Oct4_average_consensus_tree.png"/>
							<br><br>
						</tr>

						<tr>
							<br><br>
							<ul><p><b>Consensus tree - Complete method</b></p></ul>
							<img class="cons_tree" src="../images/matrix-clustering_tutorial/peak-motifs_Oct4_complete_consensus_tree.png"/>
						</tr>
						<tr>
							<br><br>
							<ul><p><b>Consensus tree - Single method</b></p></ul>
							<img class="cons_tree" src="../images/matrix-clustering_tutorial/peak-motifs_Oct4_single_consensus_tree.png"/>
							
						</tr>
					</tbody>
				</table>
				<p><b>Figure 8.</b> Three examples of consensus tree when we are using the same data (21 motifs discovered by <i>peak-motifs</i> in the Oct4 ChIP-seq) and the same threslhold values (cor >= 0.4; Ncor >= 0.6).<br>
				In this picture we only change the hierarchical clustering agglomeration rule. Top to down: average, complete, single linkage.<br>Each cluster is represented with a different color. Observe how the number of clusters and the tree topology change depending on the selected method.
				</p>
			</div>

	</li>
	<li class="sc_2">
	

		<h2><br><br><br>Study case 2</h2>
		<h3>Identifying groups of motifs belonging to the same TF Family with <i>matrix-clustering</i></h3>
		<p><i>matrix-clustering</i> can be used to group motifs bound by TF belonging to the same TF family because usually they share a common DNA binding 
			domain. This study case shows how <i>matrix-clustering</i> results can simplify the interpretation of motifs networks.<br> 
			We will analyze separately two sections of the non-redundant “core” Jaspar database, corresponding to insect and vertebrate motifs.
			For both cases, all the motifs of the collection where compared to each other, using RSAT <i>compare-matrices</i>,<br>the resulting motif-to-motif 
			similarities were converted to a motif-to-motif network using <i>Cytoscape</i> where each motif is represented as a vertex and the edges denote pairs of similar motif,
			 with color and thickness reflecting the similarity scores.<b>(Figures 9 and 10). </b><br>This network representation allows a 
            visualization of the groups, however this visualization does not provide a partition in the data into proper clusters, so let's run 
            <i>matrix-clustering</i> to study the similar motifs.
         </p>
		<ol>
			<li>Go to the results folder
				<ul>
					<p>results/JASPAR_insects<br>results/JASPAR_vertebrates</p>
				</ul>
			</li>
			<li>For each analysis open the Summary file (We recommend to use Firefox)</li>
			<li>
				<b>Insects motifs:</b>
				<ul>
					<li>In the motif-to-motif network <i>(Figure 9)</i> you can see a big group of motifs encompassing most motif in this database, 
					as they are highly connected in the network this means they are very similar to each other.</li>
					<li>
						To visualize the high similarity you can observe either:
						<ul>
							<li>Alignment table</li>
							<li>Logo Forest</li>
							<li>Consensus tree</li>
							<li>Cluster 4 results <b>(Figure 11)</b></li>
							<br>
						</ul>
					</li>
					<li>Observe that the motifs in this big cluster have the sequence <b>5'-TTAATT-3'</b> which correspond to Hox family TFs.<br>
						We could think most insect TFs correspond to Hox family, however the topology observed in the motif-to-motif network can be 
						explained because there is a high enrichment of motifs belonging to this TF family in JASPAR,<br>many research groups study these TFs and upload
						to this database the motifs they found.
						<br>
					</li>
				</ul>
			</li>
			<li>
				<b>Vertebrate motifs:</b>
				<ul>
					<li>In the motif-to-motif network <i>(Figure 10)</i> you can see a many groups of motifs (i.e. GATA, SOX, FOX).</li>
					<li>Open the Logo Forest file note that in many clusters the motifs grouped belong to the same Family. <b>(Figure 12)</b></li>
				</ul>
			</li>
		</ol>
		
		<br>
		<div class="case2 all_buttons" id="button">Fig case 2:</div>
		<div class="fig_button" id="button_fig_9">Figure 9</div>
		<div class="fig_button" id="button_fig_10">Figure 10</div>
		<div class="fig_button" id="button_fig_11">Figure 11</div>
		<div class="fig_button" id="button_fig_12">Figure 12</div>
		<br><br>
		
		
		<!-- Figure 9 -->
		<div class="fig" id="fig_9">
			<img class="network_insects" src="../images/matrix-clustering_tutorial/insects_JASPAR.jpeg"/>
			<p><b>Figure 9.</b> Motif-to-motif network done with <i>Cytoscape</i> for all the JASPAR insect motifs. Notice the big group encompassing most of the motifs of the network.</p>
		</div>
			
			
		<!-- Figure 10 -->
		<div class="fig" id="fig_10">
			<img class="network_vertebrates" src="../images/matrix-clustering_tutorial/vertebrates_JASPAR.jpeg"/>
			<p><b>Figure 10.</b> Motif-to-motif network done with <i>Cytoscape</i> for all the JASPAR vertebrate motifs.</p>
		</div>
			
			
		<!-- Figure 11 -->
		<div class="fig" id="fig_11">
			<img id="insect_logo_tree" src="../images/matrix-clustering_tutorial/insects_motifs_logo_tree.jpeg"/>
			<p><b>Figure 11.</b> Logo tree showing the highly similar motif visualized in the motif-to-motif insects network.</p>
		</div>
			
			
		<!-- Figure 12 -->
		<div class="fig" id="fig_12">
			<img id="vertebrate_logo_tree" src="../images/matrix-clustering_tutorial/vertebrate_clusters.png"/>
			<p><b>Figure 12.</b> Some clusters resulting from <i>matrix-clustering</i> are signaled in the motif-to-motif network of vertebrate motifs.</p>
		</div>
		
	</li>

</ul>

<hr><a name='references'></a><h2>References</h2>

<ol>
<li><a name='ref_chen_2010'>Chen</a>, X., Xu, H., Yuan, P., Fang, F.,
Huss, M., Vega, V. B., Wong, E., Orlov, Y. L., Zhang, W., Jiang, J.,
Loh, Y. H., Yeo, H. C., Yeo, Z. X., Narang, V., Govindarajan, K. R.,
Leong, B., Shahab, A., Ruan, Y., Bourque, G., Sung, W. K., Clarke,
N. D., Wei, C. L. and Ng, H. H. (2008). Integration of external
signaling pathways with the core transcriptional network in embryonic
stem cells. Cell 133, 1106-17. [Pubmed <a target='_blank'
href='http://www.ncbi.nlm.nih.gov/pubmed?term=18555785[uid]'>18555785</a>].
</li>

<li>Thomas-Chollier, M., Herrmann, C., Defrance, M., Sand, O., Thieffry,
  D. and van Helden, J. (2011). RSAT peak-motifs: motif analysis in
  full-size ChIP-seq datasets Nucleic Acids Research
  doi:10.1093/nar/gkr1104, 9.
  [<a target='_blank' href='http://nar.oxfordjournals.org/content/early/2011/12/08/nar.gkr1104.full?keytype=ref&ijkey=zOvloLjtKzL73F8'>Open access</a>]
</li>


<li>Mathelier, A., Zhao, X., Zhang, A. W., Parcy, F., Worsley-Hunt, R., 
Arenillas, D. J., Buchman, S., Chen, C.-y., Chou, A., Ienasescu, H., Lim, 
J., Shyr, C., Tan, G., Zhou, M., Lenhard, B., Sandelin, A. and Wasserman, 
W. W. JASPAR 2014: an extensively expanded and updated open-access 
database of transcription factor binding profiles 
Nucleic Acids Research, 2013
  [<a target='_blank' href='http://nar.oxfordjournals.org/content/early/2013/11/04/nar.gkt997.full'>Open access</a>]
</li>

</ol>

<hr width="100%">

<address>
For suggestions or information request, please contact 
<a target='_blank' href='http://www.bigre.ulb.ac.be/Users/morgane/'>Morgane Thomas-Chollier</a> or 
<a target='_blank' href='http://www.bigre.ulb.ac.be/Users/jvanheld/'>Jacques van Helden</a>
</address>

</body>


	<script>
	
		// Slide the Figures
		$(document).ready(function(){
			$('#button_fig_1').click(function(){
			$('#fig_1').slideToggle('slow');
			}); 
		});
		

		$(document).ready(function(){
			$('#button_fig_2').click(function(){
			$('#fig_2').slideToggle('slow');
			}); 
		});
		

		$(document).ready(function(){
			$('#button_fig_3').click(function(){
			$('#fig_3').slideToggle('slow');
			}); 
		});
		

		$(document).ready(function(){
			$('#button_fig_4').click(function(){
			$('#fig_4').slideToggle('slow');
			}); 
		});
		

		$(document).ready(function(){
			$('#button_fig_5').click(function(){
			$('#fig_5').slideToggle('slow');
			}); 
		});
		
		
		$(document).ready(function(){
			$('#button_fig_6').click(function(){
			$('#fig_6').slideToggle('slow');
			}); 
		});
		
		
		$(document).ready(function(){
			$('#button_fig_7').click(function(){
			$('#fig_7').slideToggle('slow');
			}); 
		});
		
		
		$(document).ready(function(){
			$('#button_fig_8').click(function(){
			$('#fig_8').slideToggle('slow');
			}); 
		});
		
		
		$(document).ready(function(){
			$('#button_fig_9').click(function(){
			$('#fig_9').slideToggle('slow');
			}); 
		});


		$(document).ready(function(){
			$('#button_fig_10').click(function(){
			$('#fig_10').slideToggle('slow');
			}); 
		});		


		$(document).ready(function(){
			$('#button_fig_11').click(function(){
			$('#fig_11').slideToggle('slow');
			}); 
		});
		

		$(document).ready(function(){
			$('#button_fig_12').click(function(){
			$('#fig_12').slideToggle('slow');
			}); 
		});				
		
		
		$(document).ready(function(){
			$('.case1').click(function(){
				$('#fig_1').hide();
				$('#fig_2').hide();
				$('#fig_3').hide();
				$('#fig_4').hide();
				$('#fig_5').hide();
				$('#fig_6').hide();
				$('#fig_7').hide();
				$('#fig_8').hide();
			}); 
		});


		$(document).ready(function(){
			$('.case2').click(function(){
				$('#fig_9').hide();
				$('#fig_10').hide();
				$('#fig_11').hide();
				$('#fig_12').hide();
			}); 
		});
	
	</script>

</html>
