<html>
<head>
   <meta HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
   <title>RSA-tools - tutorials - retrieve-seq</title>
</head>
<body text="#000000" bgcolor="#ffffff" link="#0000ee" vlink="#55008b" alink="#ff0000">
<blockquote>

<hr size=6 width="100%">
<center>
<h2><a href="../RSA_home.cgi">RSA-tools</a> - 
<a href="tutorials.html">Tutorials</a> - 
<a href="../retrieve-seq_form.cgi">retrieve-seq</a>
</h2>
</center>

<hr size=6 width="100%">

In the left frame, under the title <B>Sequence retrieval</B>, click
on the tool "<B>retrieve sequence</B>". A form appears, allowing you
to select the parameters of sequence retrieval.


<h3>Retrieving sequences for a selected set of genes</h3>
<ul>

We will first retrieve upstream sequences for a set of 5 yeast genes
involved in yeast phosphate metabolism.

<OL>

<p><li> In the pop-up menu "<B>Organism</b>", select <I>Saccharomyces
cerevisiae</I>

<p><li> In the text area under "<B>Genes</B>", type the following list
of genes</LI>

<PRE>
YBR093C
PHO8
pho11
PHO81
Pho84
</PRE>

<p><li>Inactivate the option <b>Prevent overlap with upstream
ORFs</b>.

<p><li>Leave all other options unchanged and click on the button
<b>GO</b>

</OL>

<p><B>Remarks</B>
<ul>
<li>Gene names must be separated by carriage returns, because
only the first word of each line is considered as a query.

<li>Notice that gene names are case-insensitive. 

<li>Genes can be specified either by the systematic ORF identifier or
by a common name. Synonyms are supported.
</ul>


<p>
After a few seconds, the result appears, in the form of a link to the
sequences, followed by a series of Buttons. 

<P>
Click on the link, to see the upstream sequence of the selected
genes. Check the sizes of the sequences obtained. Click on the
<b>Back</b> button to come back to the previous result page.

<p>
Below the link, you can see a series of buttons, which
will allow you to send the retrieved sequences to the next task (as we
will see in the next tutorials). For the time being, do not click
these buttons.


</ul>
<h3>Preventing overlap with upstream ORFs</h3>
<ul>

<ol>
<p><li>Come back to the <b>retrieve sequence</b> form. 
<p><li>Perform the same operations as above, but this time, activate
the option <b>Prevent overlap with upstream ORFs</b>.
<p><li>Compare the size of the sequences with the previous result. 
</ol>

</ul>
<h3>Interpretation of the results</h3>
<ul>


<p>
The default size for yeast upstream sequences is 800bp. The reason for
this choice is that 99% of the regulatory elements found in TRANSFAC
(the Transcription Factor Database) are located between -1 and -800,
relative to the start codon.

<p>
However, the median value for the distance between a gene and its
upstream neighbour is 450 bp. Thus, with a distance of 800bp, most
upstream regions would include a fragment of coding sequence from the
upstream ORF, which might bias subsequent analyses.

<p>

The situation is even worst in bacteria, due to the organization of
genes into operons. In <i>Escherichia coli</i>, 25% of the genes have
an upstream neighbour closer than 50bp. When this neighbour is on the
same strand as the gene, it might indicate that they belong to the
same operon. 

<p>
For most organisms, it is thus preferible to prevent overlap with
upstream ORFs. A notable exception is the yeast <i>Saccharomyces
cerevisiae</i>. We evaluated the results obtained when this option is
activated and inactivated respectively, and it is clear that many
patterns escape detection when upstream ORFs are clipped. One possible
reason is that the annotations of <i>Saccharomyces cerevisiae</i>
contain many false ORFs (some auhors estimate that there are ~1000
false ORFs in the original annotation).

<p>
This is particularly well examplified with the yeast PHO genes : when
the option is activated, PHO8 and PHO64 upstream regions are truncated
at 180 and 237 base pairs, respectively. In both cases, the putative
upstream ORF is very short (~300bp), and there is no experimental
evidence for the existence of a gene. In addition, several cis-acting
elements have been shown to be located in the regions corresponding to
these questionable ORFs. When the sequences are clipped, the remaining
Pho4p are not sufficiently over-rerpesented for allowing their
detection in a pattern discovery programs.

<p>
The PHO family is an extreme case, and when one disposes of a larger
family, it is less likely for false ORFs to prevent detection of
cis-acting elements. However, even in these cases, we noticed a loss
in sensitivity when the option was activated.

<p>
<b>In summary</b>, we propose the following strategy :
<ul>
<li>When working with bacteria, systematically activate the option
<b>prevent overlap with upstream ORFs</b>, to prevent including too
many coding sequences when the genes are located in operons.
<li>With eukaryotes, inactivate this option, to circumvent the
weaknesses of gene boundary annotations.
</ul>


</ul>
<h3>Retrieving sequences for all the genes of a genome</h3>
<ul>

As an illustration fo the retrieval of all sequences, we will now use
the same program to retrieve sequences on the other side of the genes
: downstream sequences.

<OL>

<p><li>Open the <b>retrieve sequence</b> form by clicking
<b>retrieve sequence</b> in the left frame.

<p><li>Make sure that the selected organism is <I>Saccharomyces
cerevisiae</I>

<p><li>Besides the label <B>Genes</B>, select "all".

<p><li>Do not enter anything in the "Gene" box (it would be ignored
anyway).

<p><li>Select "downstream" as <B>Sequence type</B>.

<p><li>In the <B>From</b> box, type "1".

<p><li>In the <B>to</b> box, type "50".

<p><li>Select "ORF id + Gene name" in the pop-up menu <B>Sequence
label</B>

<p><li>Click <B>GO</B>

</OL>

</ul>
<h3>Remark about the reference position</H3>
<ul>

<p>
The reference position (coordinate 0) depends on the type of sequences
to be retrieved :
<p>

<ul>

<li>For <B>upstream sequences</B>, the reference position is the ORF
start, i.e. the first nucleotide of the start codon. Negative
coordinates are used to indicate sequences located upstream the start
codon, i.e. in the 5'end flank of the gene (e.g. from -1 to -800 for
the yeast).

<li>On the contrary, when selecting <B>downstream sequences</B>, the
reference position (position 0) is the ORF end, i.e. the last
nucleotide of the stop codon. Positive coordinates are used to
represent sequences located downstream the stop codon, i.e. the 3'end
flank of the gene.

</ul>

</ul>
<h3>Positive/negative coordinates</H3>
<ul>

<p>
Notice that selecting upstream sequences with positive coordinates
amounts to pick coding sequences on the 5' side of the gene. As an
exercise, try to extract all start codons for a given organism (say
<i>Saccharomyces cerevisiae</i>). Check that the codons correspond to
the expectation (ATG).

<p>
Reciprocally, for downstream sequences, negative coordinates could be
specified to retrieve coding sequences on the 3'end of the gene. As en
exercise, try to extract all the stop codons for the same
organism. Check that the stop codons correspond to the expectation
(TAA, TAG or TGA).

</ul>
<p>
You can now come back to the <a href=tutorials.html>tutorial main
page</a> and follow the next tutorials.

</ul>
<hr width="100%">
<center>
<address>
For suggestions or information request, please contact <a
href="mailto:jvanheld@ucmb.ulb.ac.be">Jacques van Helden
(jvanheld@ucmb.ulb.ac.be)</a>
</address></center>

</blockquote>
</body>
</html>
