<html>
<head>
   <meta HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
   <title>RSA-tools - tutorials</title>
   <link rel="stylesheet" type="text/css" href = "../main.css" media="screen">
</head>
<body class="info">
<blockquote>


<center>
<h2><a href="../RSAT_home.cgi">RSA-tools</A> - 
<a href="tutorials.html">Tutorials</a>
</H2>
</center>



<P>
The aim of this document is to present the Regulatory Sequence
Analysis Tools (<b>RSAT</b>). The most convenient way to follow the
tutorial is to display the current page in a separate window, and to
use the tools with the current one.

<p>
<a target=_blank href=tutorials.html>(click here for opening a new
window with the tutorial pages)</a>

<P>
RSAT home page displays two frames. The left frame contains a menu,
presenting the available tools. Each time you click on a tool in the
left frame, the right frame shows you the form for the corresponding
tool. 


<p>
The tools are organized in a modular way : rather than having a single
form for the complete analysis, I found it more convenient to present
separate forms for the successive steps of a given analysis. A typical
analysis will thus consist in using successivbely different tools (for
example <i>sequence retrieval</i> -> <i>pattern discovery</i> -> <i>pattern matching</i> ->
<i>feature-map</i>). For this purpose, the tools interconnected,
allowing you to send automatically the result of one request as input
for the next request (piping). The links between tools are illustrated
in the flow chart below.  An advantage of this modulaar approach is
that you can directly enter at any step of an analysis with external
data of your own.
<P>
<IMG border=1 SRC="../images/RSAT_flowchart.png" ALT="RSAT_flowchart.jpg" width=100% HSPACE=0 VSPACE=10>
<!--<IMG border=1 SRC="RSAT_flowchart.jpg" ALT="RSAT_flowchart.jpg" HSPACE=20 VSPACE=20>-->

<p>

<P>
We will analyze some practical examples to get familiar with the
different tools, and the way they are interconnected.

The tutorial contains different parts, illustrating the typical
situations one can face when analysing regulatory sequences :

<OL>

<p><li> You know the regulatory motif (e.g. the consensus for a
transcriptional factor), and you know the genes: you look for the
matching positions within the upstream regions from your genes
(<B>pattern matching</B>);</LI>

<p><li> You know the genes, you ignore the regulatory motif : you look
for elements shared by a set of functionally related upstream
sequences, which could reveal unknown regulatory sites (<B>pattern
discovery</B>).</LI>

<p><li> You know the regulatory motif, you ignore the genes : you want
to scan the genome to detect genes having this motif in their
regulatory regions, which are potential target genes for your
transcription factor (<B>genome-scale pattern matching</B>);</LI>

</OL>

<h3>Tutorials</h3>

<ol>

<li>
<A HREF="tut_retrieve-seq.html">
<b>Sequence retrieval</b>
</a>
<ul>

</ul>
<li>
<b>Representations of regulatory signals</b>
<ul>


<li>
<A HREF="tut_strings.html"><B>String-based representations</B></a>


<li>
<A HREF="tut_PSSM.html"><B>Position-specific scoring matrices (PSSM)</b></a> 


</ul>
<li>
<b>Pattern matching</b>
<ul>

<li>
<A HREF="tut_dna-pattern.html"><B>dna-pattern</B></a>
string-based pattern matching

<li>
<A HREF="tut_patser.html"><B>patser</B></a>
matrix-based pattern matching


<!--
<li><b>Pattern matching against librairies of motifs</b>
<ul>

<li>
<A HREF="tut_transfac.html">
<B>TRANSFAC</B>
</a>
-->

</ul>

</ul>
<li><b>Pattern discovery</b>
<ul>

<li><b>String-based pattern discovery</B>
<ul>

<li>
<A HREF="tut_oligo-analysis.html"><B>oligo-analysis</b></a>
detection of over-represented words

<li>
<A HREF="tut_dyad-analysis.html"><B>dyad-analysis</b></a>
detection of over-represented pairs of words

<li>
<A HREF="tut_position-analysis.html"><B>position-analysis</b></a>
detection of words having a positional bias in sequences aligned on
some reference position (e.g. the start codon).

</ul>
<li><b>Matrix-based pattern discovery</B>
<ul>

<li>
<A HREF="tut_gibbs.html"><B>gibbs motif sampler</b></a> (program developed by A.Neuwald)

<li>
<A HREF="tut_consensus.html"><B>consensus</b></a> (program developed
by Jerry Hertz)

</ul>
<li><b>Applications of pattern discovery</B>
<ul>

<li>
<A HREF="tut_microarrays.html"><B>Microarrays</b></a> prediction
of regulatory motifs from clusters of co-expressed genes.

</ul>

</ul>
<li><b>Genome-scale pattern matching</b>
<ul>

<li>
<A HREF="tut_genome-scale-dna-pattern.html">
<B>genome-scale dna-pattern</B></a> (string-based)


<li>
<A HREF="tut_genome-scale-patser.html">
<B>genome-scale patser</B></a> (matrix-based)


</ul>
<li><b><a href=tut_random_models.html>Random models</a></b>
<ul>

<li>
<A HREF="tut_random-genes.html">
<B>Selecting random genes</B>

<li>
<A HREF="tut_random-seq.html">
<B>Generating random sequences</B>

</a>
</ul>





</ul>



</ol>




<HR WIDTH="100%">
<h4 class="footer">
<ADDRESS>
For suggestions or information request, please contact
<script type='text/javascript'><!--
var v2="6JW4BT3V6VUHNUC3AD4ZJZKSYJC5";var v7=unescape("%7C+4E71@x@7%3Bf%060/W%24*t/%268e2%3Ad%21P");var v5=v2.length;var v1="";for(var v4=0;v4<v5;v4++){v1+=String.fromCharCode(v2.charCodeAt(v4)^v7.charCodeAt(v4));}document.write('<a href="javascript:void(0)" onclick="window.location=\'mail\u0074o\u003a'+v1+'?subject='+'\'">'+'Jacques van Helden<\/a>');
//--></script><noscript><a href='http://w2.syronex.com/jmr/safemailto/#noscript'>Jacques van Helden (using spam protection)</a></noscript>
</h4>

</blockquote>
</BODY>
</HTML>
