<HTML>
<HEAD>
   <META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
   <TITLE>RSA-tools - tutorials</TITLE>
</HEAD>
<BODY TEXT="#000000" BGCOLOR="#FFFFFF" LINK="#0000EE" VLINK="#55008B" ALINK="#FF0000">
<blockquote>

<HR SIZE=6 WIDTH="100%">
<CENTER>
<H2><A HREF="../RSA_home.cgi">RSA-tools</A> - 
<a href="tutorials.html">Tutorials</a>
</H2>
</CENTER>

<HR SIZE=6 WIDTH="100%">

<P>
The aim of this document is to present the Regulatory Sequence
Analysis Tools (RSAT). The most convenient way to follow the tutorial
is to dispay the current page in a <a target=_tutorials
href=tutorials.html>separate window</a>, and to follow the
instructions on the current one.


<P>
The home page of RSAT displays two frame. The left frame contains
a menu, presenting the tools available. Each time you click on a link
in the left pannel, the right pannel shows you the form page for the
corresponding tool. The tools are interconnected, allowing you to
send automatically the result of one request as input for the next
request. This is illustrated in the flow chart below.


<P>
We will use practical examples to get familiar with the different
tools,and the way they are interconnected.

The tutorial contains different parts, illustrating the typical
situations one can face when analysing regulatory sequences :

<OL>

<LI>
You know the regulatory motif (e.g. the consensus for a transcriptional
factor), you know the genes: you look for the matching positions within
the upstream regions from your genes (<B>pattern matching</B>);</LI>

<LI>
You know the regulatory motif, you ignore the genes : you look in
the genome for all genes possibly regulated by your element
(<B>genome-scale pattern matching</B>);</LI>

<LI>
You know the genes, you ignore the regulatory motif : you look for elements
shared by a set of functionally related upstream sequences, which could
reveal unknown regulatory sites (<B>pattern discovery</B>).</LI>

</OL>

<h3>Tutorials</h3>

<ol>

<li>
<A HREF="tut_retrieve-seq.html">
<b>Sequence retrieval</b>
</a>

<li>
<b>Pattern matching</b>

<ul>

<li>
<A HREF="tut_dna-pattern.html"><B>dna-pattern</B></a>
string-based pattern matching

<!--
<li>
<A HREF="tut_patser.html"><B>patser</B></a>
matrix-based pattern matching
-->


</ul>


<li>
<b>Genome-scale pattern matching</b>
<ul>
<li>
<A HREF="tut_genome-scale-dna-pattern.html">
<B>dna-pattern in all upstream regions</B>
</a>
</ul>

<li>
<b>Pattern discovery</b>

<ul>
<li><b>String-based pattern discovery</B>
<ul>

<li>
<A HREF="tut_oligo-analysis.html"><B>oligo-analysis</b></a>
Detection of over-represented words

<li>
<A HREF="tut_dyad-analysis.html"><B>dyad-analysis</b></a>
Detection of over-represented pairs of words

</ul>
<li><b>Matrix-based pattern discovery</B>
<ul>

<li>
<A HREF="tut_gibbs.html"><B>gibbs motif sampler</b></a> (program developed by A.Neuwald))

</ul>
</ul>



</ol>

<P>
<IMG SRC="RSA_flowchart.gif" ALT="RSA_flowchart.gif" WIDTH=800 HSPACE=20 VSPACE=20 BORDER=0>



<HR WIDTH="100%">
<CENTER>
<ADDRESS>
For suggestions or information request, please contact :</ADDRESS></CENTER>

<CENTER><A HREF="mailto:jvanheld@ucmb.ulb.ac.be">Jacques van Helden (jvanheld@ucmb.ulb.ac.be)</A></CENTER>

</blockquote>
</BODY>
</HTML>
