<html><head><meta content="text/html; charset=UTF-8" http-equiv="Content-Type"><title>Documentation generated from WSDL</title></head><body><table align="left" border="0" width="80%"><tbody><tr><th><table align="left" border="1" cellpadding="10" width="100%"><tbody><tr><th bgcolor="#000080"><font color="#FFFFFF" size="5">RSATWSPortType</font></th></tr><tr><th align="left"><table cellpadding="10" width="100%"><tbody><tr><th align="left" valign="top" width="200"><font color="#2200FF" size="4">Service Documentation</font></th><td>Web
 services for the Regulatory Sequence Analysis Tools (RSAT). Tools 
developed by Jacques van Helden (jvanheld@bigre.ulb.ac.be), SOAP/WSDL 
interface developed by Olivier Sand (oly@bigre.ulb.ac.be).</td></tr><tr><th align="left" valign="top"><font color="#2200FF" size="4">Server Address</font></th><td>http://rsat.ulb.ac.be/rsat/web_services/RSATWS.cgi</td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">retrieve_seq</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Returns upstream, downstream or coding DNA sequences for list of query genes.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">organism</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Organism. Words need to be underscore separated (example: Escherichia_coli_K12).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">query</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A list of query genes.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">all</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return sequences for all the genes of the organism if value = 1. Incompatible with query.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">noorf</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Prevent overlap with upstream open reading frames (ORF) if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">from</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Inferior limit of the region to retrieve.
Default is organism dependant (example: Saccharomyces cerevisiae = -800).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">to</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Superior limit of the region to retrieve. Default is '-1'.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">feattype</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Type of genome features to load. Supported: CDS, mRNA, tRNA, rRNA. </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">type</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black"> Sequence type. Supported: upstream, downstream, ORF (unspliced open reading frame).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sequence format. Supported: IG (Intelligenetics), WC (wconsensus), raw, FastA</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">lw</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Line width (0 for whole sequence on one line).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">label</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Field(s) to be used in the sequence label. Multiple fields can be specified, separated by commas. 
Supported: id, name, organism_name, sequence_type, current_from, current_to, ctg, orf_strand, reg_left, reg_right. 
Default: name.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">label_sep</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Separator between the label fields. Default: | (pipe character).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">nocom</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">No comments if value = 1. Only the identifier and the sequence are returned. 
By default, the comment indicates the ORF and upstream sequence coordinates.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">repeat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Use the repeat masked version of the genome if value = 1.
Warning: repeated regions are annotated for some genomes only.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">imp_pos</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Admit imprecise positions if value = 1.</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">retrieve_seq_multigenome</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Returns upstream, downstream or coding DNA sequencesfor list of query genes and organisms.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">input</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The input file is a tab-delimited text files with (at least) the two following columns:
    1. gene ID or name
      Identifier oor synonyms are supported.
    2. Organism name
      For the organism name, spaces must be replaced by underscore character
      (exactly as for retrieve-seq).
    If additional columns are included in the input file, they are ignored.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_input_file</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input file on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">all</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return sequences for all the genes of the organism if value = 1. Incompatible with query.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">noorf</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Prevent overlap with upstream open reading frames (ORF) if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">from</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Inferior limit of the region to retrieve. Default is organism dependant (example: Saccharomyces cerevisiae = -800).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">to</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Superior limit of the region to retrieve. Default is '-1'.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">feattype</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Type of genome features to load. Supported: CDS, mRNA, tRNA, rRNA. </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">type</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black"> Sequence type. Supported: upstream, downstream, ORF (unspliced open reading frame).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sequence format. Supported: IG (Intelligenetics), WC (wconsensus), raw, FastA</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">lw</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Line width (0 for whole sequence on one line).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">label</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Field(s) to be used in the sequence label. Multiple fields can be specified, separated by commas. 
Supported: id, name, organism_name, sequence_type, current_from, current_to, ctg, orf_strand, reg_left, reg_right. 
Default: name.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">label_sep</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Separator between the label fields. Default: | (pipe character).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">nocom</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">No comments if value = 1. Only the identifier and the sequence are returned. 
By default, the comment indicates the ORF and upstream sequence coordinates.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">repeat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Use the repeat masked version of the genome if value = 1. Attention: repeated regions are annotated for some genomes only.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">imp_pos</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Admit imprecise positions if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">gene_col</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of the column containing the gene names/identifiers (default: 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">org_col</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of the column containing the organisms (default: 2).</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">retrieve_ensembl_seq</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Returns upstream, downstream or coding DNA sequences for list of query genes (in EnsEMBL database).<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">organism</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Organism. Words need to be underscore separated (example: Escherichia_coli_K12).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">ensembl_host</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Address of ensembl database server (default is EBI server).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">db_name</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of EnsEMBL database (alternative to organism).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">query</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A list of query genes. You need to supply either this parameter or the next one (tmp_infile).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black"> Name of the file with list of genes on the server. You need to supply either this parameter or the previous one (query).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">all</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return sequences for all the genes of the organism if value = 1. Incompatible with query.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">noorf</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Prevent overlap with upstream open reading frames (ORF) if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">nogene</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Prevent overlap with upstream gene (extreme transcripts limits) if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">from</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Inferior limit of the region to retrieve. Default is organism dependant (example: Saccharomyces cerevisiae = -800).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">to</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Superior limit of the region to retrieve. Default is '-1'.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">feattype</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Type of genome features to load. Supported: Gene, CDS, mRNA, intron, exon. </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">type</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black"> Sequence type. Supported: upstream, downstream.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">chromosome</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Chromosome name or number (to use with -left and -right).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">left</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Left limit of sequence to retrieve.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">right</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Right limit of sequence to retrieve.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">strand</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Strand of sequence to retrieve when using -left and -right. Values: 1, -1</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">features</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Features.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">feat_format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Features format. Supported: ft, gft</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">mask_coding</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">All coding sequence is replaced by N in the retrieved sequence if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">repeat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Use the repeat masked version of the genome if value = 1. Attention: repeated regions are annotated for some genomes only.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">all_transcripts</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Get sequences for all transcript of genes if value = 1. Combine with unique_sequences option if you do pattern discovery afterwards.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">unique_sequences</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">When getting sequences for all transcripts of genes, keep only non-redundant fragments if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">first_intron</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">With feattype intron, get only first intron sequence if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">non_coding</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">With feattype exon, get only non-coding (part of) exons if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">utr</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">With feattype UTR, get only 5prime or 3prime UTR (default is all UTRs).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">line_width</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A newline character will be inserted in the sequence every ## bases. 0 will prevent newline insertion. This is the default value</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">ortho</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Get orthologuous sequences if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">taxon</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Filter on taxonomic level when collecting orthologs (e.g. Murinae)</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">homology_type</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Filter on homology type when collecting orthologs (e.g. ortholog_one2one)</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">header_organism</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Type of organism name to use in the fasta header (scientific, common or none).
 Default is scientific. Common name is only accessible with -ortho.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">purge_seq</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Mask repeated fragments of an input sequence.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">sequence</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sequence to purge. You need to supply either this parameter or the next one (tmp_infile).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black"> Name of the file with input sequence on the server. You need to supply either this parameter or the previous one (sequence).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sequence format. Supported: IG (Intelligenetics), WC (wconsensus), raw, FastA.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">match_length</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Minimal match length. Default is 40.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">mismatch</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of mismatches allowed. Default is 3.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">str</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Discard duplications on the direct strand only (1) or on the reverse complement as well (2). Default is 2.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">delete</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Delete repeats instead of masking them if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">mask_short</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Mask (replace by N characters) sequences shorter than the specified length.</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">oligo_analysis</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Analysis
 of the statistical significance of all the oligomers of a given size in
 a sequence. Commonly used to detect over-represented oligonucleotides 
in a set of promoter sequences.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">verbosity</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Verbosity.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">sequence</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input sequence. You need to supply either this parameter or the next one (tmp_infile).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with input sequence on the server. You need to supply either this parameter or the previous one (sequence).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input sequence format. Supported: IG (Intelligenetics), WC (wconsensus), raw, fasta. Default is fasta.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">length</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Oligomer length.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">organism</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Organism. Words need to be underscore separated (example: Escherichia_coli_K12).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">background</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Background model: Type of sequences used as background model for estimating expected oligonucleotide frequencies. 
Supported: upstream, upstream-noorf, upstream-noorf-rm, intergenic, input.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">stats</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">List of statistics to return. Supported:occ, mseq, freq, proba, ratio, zscore, like, pos, rank.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">noov</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">No overlapping of oligos allowed if value = 1.
Disable the detection of overlapping matches for self-overlapping patterns (ex TATATA, GATAGA).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">str</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Oligonucleotide occurrences found on both stands are summed (2) or not (1). Default is 2.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">sort</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sort oligomers according to overrepresentation if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">lth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Lower threshold on some parameters. Format=list of'parameter value'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">uth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Upper threshold on some parameters. Format=list of 'parameter value'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">pseudo</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Pseudo-weight. Must be a real value between 0 and 1.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">oligo_diff</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Compare
 oligonucleotide occurrences between two input sequence files (test and 
control), and return oligos that are significantly enriched in one of 
the files respective to the other one.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">verbosity</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Verbosity.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">test</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input test sequence in fasta format. You need to supply either this parameter or the next one (tmp_test_infile).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_test_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with input test sequence on the server. You need to supply either this parameter or the previous one (test).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">control</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input control sequence in fasta format. You need to supply either this parameter or the next one (tmp_control_infile).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_control_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with input control sequence on the server. You need to supply either this parameter or the previous one (control).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">side</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Side of the significance test (values: test, ctrl, both). In practice, the side is converted into a threshold on the ratio test/control occurences.
	      test: only tests over-representation in the test sequences. This is converted into a lower threshold of 1 for the test/control ratio
	      both: test over-representation in eitherr the test or the control set
	      ctrl: only tests over-representation in the control sequences. This is converted into a upper thrshold of 1 for the test/control ratio.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">nopurge</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Do not purge input sequences before counting oligonucleotide occurences if value = 1.
Input sequences are purged by default, as this is highly recommended since redundant sequence fragments bias the overrepresentation statistics and create false positives</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">length</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Oligomer length.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">noov</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">No overlapping of oligos allowed if value = 1.
Disable the detection of overlapping matches for self-overlapping patterns (ex TATATA, GATAGA).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">str</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Oligonucleotide occurrences found on both stands are summed (2) or not (1). Default is 2.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">lth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Lower threshold on some parameters. Format=list of'parameter value'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">uth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Upper threshold on some parameters. Format=list of 'parameter value'.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">peak_motifs</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Pipeline for discovering motifs from ChIP-seq (or ChIP-chip) peak sequences.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">verbosity</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Verbosity.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">test</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input test peak sequence in fasta format. You need to supply either this parameter or the next one (tmp_test_infile).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_test_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with input test peak sequence on the server. You need to supply either this parameter or the previous one (test).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">control</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input control peak sequence in fasta format. You can supply either this parameter or the next one (tmp_control_infile) or none.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_control_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with input control peak sequence on the server. You can supply either this parameter or the previous one (control) or none.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">max_seq_length</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximal sequence length.
Larger sequences  are truncated  at the specified length around the sequence center (from -value/2 to value/2)</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">max_motif_number</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximal number of motifs (matrices) to return for pattern discovery algorithms.
Note the distinction between the maximal number of motifs (matrices) and the maximum number of patterns (words, dyads): a motif generally corresponds to mutually overlapping several patterns (dyads, words).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">ref_motif</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Reference motif.
In some cases, we already dispose of a reference motif, for example the motif annotated in some transcription factor database (e.g. RegulonDB, Jaspar, TRANSFAC) for the transcription factor of interest. These annotations may come from low-throughput experiments, and rely on a poor number of sites, but the reference motif may nevertheless be informative, because it is based on several independent studies.
Each discovered motif can be compared to the reference motif, in order to evaluate its correspondence with the binding motif of the factor of interest.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">top_peaks</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Restrict the analysis to the N peaks at the top of the input sequence file. Some peak calling programs return the peaks sorted by
score. In such case, the -top_peaks option allows to restrict the analysis to the highest scoring peaks. In some cases, the top-scoring peaks might contain a higher density of binding sites,
allowing to detect motifs with a higher significance.
This option can also be convenient for performing quick tests, parameter selection and debugging before running the full analysis of large sequence sets.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">min_length</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Minimal oligonucleotide length. Use in combination with the next option (max_length).
If those options are used, the program iterates over the specified range of oligonucleotide lengths.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">max_length</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximal oligonucleotide length. Use in combination with the previous option (min_length).
If those options are used, the program iterates over the specified range of oligonucleotide lengths.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">markov</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Order of the Markov model used to estimatd expected oligonucleotide frequencies for oligo-analysis and local-word-analysis.
Higher order Markov models are more stringent, lower order are more sensitive, but tend to return a large number of false positives.
Markov models can be specified with either a positive or a negative value. Positive value indicate the length of the prefix in the transition matrix. Negative value indicate the order of the Markov model relative to the oligonucleotide length. For example, the option -markov -2 gives a model of order m=k-2 (thus, an order 5 for heptanucleotides, an order 4 for hexanucleotides).
The optimal Markov order depends on the number of sequences in the test set. Since ChIP-seq data typically contain hundreds to thoursands of peaks, high Markov orders are generally good, because they are stringent and still sensitive enough. In our experience, motifs are well detected with the most stringent Markov order (-markov -2).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">min_markov</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Minimal value for markov order. Use in combination with the next option (max_markov).
If those options are used, the program iterates over the specified range of markov orders.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">max_markov</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximal value for markov order. Use in combination with the previous option (min_markov).
If those options are used, the program iterates over the specified range of markov orders.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">noov</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">No overlapping of oligos allowed if value = 1.
Disable the detection of overlapping matches for self-overlapping patterns (ex TATATA, GATAGA).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">class_int</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Class interval for position-analysis. The width of the position classes, in number of bases (default: 20)</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">str</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Oligonucleotide occurrences found on both stands are summed (2) or not (1). Default is 2.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">graph_title</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Title displayed on top of the graphs.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">image_format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Image format.
All the formats supported by XYgraph can be used.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">disco</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">oligos|dyads|positions|local_words|merged_words|meme|chipmunk
Specify the software tool(s) that will be used for motif discovery.
Several algorithms can be specified either by using the option iteratively:
	-disco oligos -disco dyads
or by entering a comma-separated list of algorithms:
	-disco oligos,dyads

Default motif discovery algorithms
  oligos: Run oligo-analysis to detect over-represented oligonucleotides of a given length (k, specified with option -l) in the test set (van Helden et al., 1998). Prior frequencies of oligonucleotides are taken from Markov model of order m (see option -markov) estimated from the test set sequences themselves.
  dyads: Run dyad-analysis to detect over-represented dyads, i.e. pairs of short oligonucleotides (monads) spaced by a region of fixed width but variable content (van Helden et al., 2000). Spaced motifs are typical of certain classes of transcription factors forming homo- or heterodimers. By default, chip-seq-analysis analyzes pairs of trinucleotides with any spacing between 0 and 20. The expected frequency of each dyad is estimated as the product of its monad frequencies in the input sequences (option -bg monads of dyad-analysis).
  positions: Run position-analysis to detect oligonucleotides showing a positional bias, i.e. have a non-homogeneous distribution in the peak sequence set.
	This method was initially developed to analyze termination and poly-adenylation signals in downstream sequences (van Helden et al., 2001), and it turns out to be very efficient for detecting motifs centred on the ChIP-seq peaks. For ChIP-seq analysis, the reference position is the center of each sequence.
	Note that chip-seq-analysis also uses position-analysis for the task profiles, in order to detect compositional biases (residues, dinucleotides) in the test sequence set.
  local_words: Run local-word-analysis to detect locally over-represented oligonucleotides and dyads.
	The program local-word-analysis (Matthieu Defrance,unpublished) tests the over-representation of each possible word (oligo, dyad) in positional windows in the input sequence set.
	Two types of background models are supported: (i) Markov model of order m estimated locally (within the window under consideration; (ii) the frequency observed for a word in the whole sequence set is used as estimator of the prior probability of this word in the window.
	After our first trials, this program gives excellent results in ChIP-seq datasets, because its senstivitity increases with large number of sequences (several hundreds/thousands), and its background model is more stringent than for programs computing the global over-representation (oligo-analysis, dyad-analysis).
  merged_words: Extract a position-specific scoring matrix (using matrix-from-patterns) from all the words discovered by the selected string-based motif disovery algorithms (oigos,dyads,positions and/or local_words).
	      </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">source</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Enter the source of the fasta sequence file.
Supported source: galaxy
When the sequence file comes from Galaxy, peak coordinates embedded in the fasta headers are extracted and used to convert predicted site coordinates (relative to peak center) to genomic coordinates (in the form of a bed file), which can then be uploaded to the UCSC genome browser as an annotation track.
This option is incompatible with -coord.
	      </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">task</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a subset of tasks to be executed.
By default, the program runs all necessary tasks. However, in some cases, it can be useful to select one or several tasks to be executed separately.
Beware: task selection requires expertise, because most tasks depends on the prior execution of some other tasks in the workflow. Selecting tasks before their prerequisite tasks have been completed will provoke fatal errors.
Available Tasks:
  all (default): Run all supported tasks.
  purge: Purge input sequences (test set and, if specified, control set) to mask redundant fragments before applying pattern discovery algorithms. Sequence purging is necessary because redundant fragments would violate the hypothesis of independence underlying the binomial significance test, resulting in a large number of false positive patterns.
  seqlen: Compute sequence lengths and their distribution. Sequence lengths are useful for the negative control (selection of random genome fragments). Sequence length distribution is informative to get an idea about the variability of peak lengths.
  composition: Compute compositional profiles, i.e. distributions of residues and dinucleotide frequencies per position (using position-analysis).
	Residue profiles may reveal composition biases in the neighborhood of the peak sequences. Dinucleotide profiles can reveal (for example) an enrichment in CpG island.
	Note that peak-motifs also runs position-analysis with larger oligonucleotide length (see option -l) to detect motifs on the basis of positionally biased oligonucleotides (see task positions).
  ref_motifs: This task combines various operations.
	Formating of the reference motif
          Perform various format conversion for the reference motif (compute parameters, consensus, logo).
	Motif enrichment
          Generate an enriched motif by scanning the peak sequence set with the reference motif.
	Motif comparison
	  Compare all discovered motifs with the reference motif.
  disco: Run the motif discovery algorithms. See option -disco for the selection of motif discovery algorithm(s).
  merge_words: Merge the words (oligos or dyads) discovered by the different string-based motif discovery algorithms.
	The table of merged words has one row per word (oligo or dyad) and one column per motif discovery program. This table is convenient to analyze the consistency between the words detected by different approaches, e.g. show that a word is both over-represented (oligo-analysis, dyad-analysis) and positionally biased (position-analysis, local-words). A heatmap is also exported to provide a graphical representation of the significance of each word (row) for each algorthm (column).
	The merged words can optionally be used as seeds for extracting position-specific scoring matrices from the sequences, using the program matrix-from-patterns (see option -disco merged_words).
  motif_compa: Motifs are compared in three ways.
	Discovered versus discovered (task cluster_motifs)
        Perform pairwise comparisons between all motifs (matrices) discovered by the different algorithms, to assess their consistency.
  motifs_vs_ref:
	Compare each discovered motif to the reference motif.
  motifs_vs_db:
	Compare each discovered motif to a database of known motifs (e.g. Jaspar, TRANSFAC, RegulonDB, UniProbe, ...)
  timelog: Generate a log file summarizing the time spent in the different tasks.
  synthesis: Generate the HTML file providing a synthesis of the results and pointing towards the individual result files.
  clean_seq: Delete the purged sequence files after the analysis, in order to save space. This task is executed only when it is called explicitly. It is not part of the tasks running with the option "-task all".</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">dyad_analysis</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Analysis
 of the statistical significance of all the spaced dyads of a given size
 in a sequence. Commonly used to detect over-represented spaced dyads in
 a set of promoter sequences.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">verbosity</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Verbosity.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">sequence</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input sequence. You need to supply either this parameter or the next one (tmp_infile).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with input sequence on the server. 
You need to supply either this parameter or the previous one (sequence).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input sequence format. Supported: IG (Intelligenetics), WC (wconsensus), raw, fasta. Default is fasta.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">length</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Dyad length.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">spacing</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Spacing between elements of the dyads.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">organism</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Organism. Words need to be underscore separated (example: Escherichia_coli_K12).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">background</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Background model: Type of sequences used as background model for estimating expected oligonucleotide frequencies. 
Supported: upstream, upstreamL, upstream-noorf, intergenic, input.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">stats</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">List of statistics to return. Supported: occ, mseq, freq, proba, ratio, zscore, like, pos, rank.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">type</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">dyad_type (dr | ir | rep | any) 
In order to fasten execution, the program can be asked to restrict its analysis to symmetric dyads. 
Four types are accepted: 
	dr - direct repeats: the second element is the same as the first one; 
	ir - inverted repeats: the second element is the revers complement of the first one; 
	rep - repeats: direct and inverted repeats are evaluated. 
	any - (default) 
When selecting the option any, the analysis is performed on all non-symmetric dyads as well.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">noov</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">No overlapping of dyads allowed if value = 1. 
Disable the detection of overlapping matches for self-overlapping patterns (ex TATATA, GATAGA).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">str</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Dyad occurrences found on both stands are summed (2) or not (1). Default is 2.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">sort</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sort dyads according to overrepresentation if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">under</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Detect under-represented instead of over-represented dyads (left tail significance test) if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">two_tails</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Detect under-represented and over-represented dyads (two-fail significance test) if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">zeroocc</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Report also dyads with zero occurrences (provided they fit the other thresholds) if value = 1. 
By default, the program reports only patterns present in the sequence. 
If the left tail or two-tail test is applied, patterns with zero occurrences are automatically taken into account. 
In some other cases, one would also like to detect patterns absent from the sequence. 
This is the function of the option -zeroocc.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">lth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Lower threshold on some parameters. Format=list of 'parameter value'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">uth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Upper threshold on some parameter. Format=list of 'parameter value'.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">position_analysis</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Calculates
 the positional distribution of oligonucleotides in a set of sequences, 
and detects those which significantly discard from a homogeneous 
distribution.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">verbosity</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Verbosity.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">sequence</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input sequence. You need to supply either this parameter or the next one (tmp_infile).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with input sequence on the server. You need to supply either this parameter or the previous one (sequence).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input sequence format. Supported: IG (Intelligenetics), WC (wconsensus), raw, fasta. Default is fasta.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">length</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Oligomer length.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">seq_type</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sequence type (values: dna, any).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">last</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Stop after a given number of sequences (for quick testing).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">mask</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Mask lower or uppercases by replacing the selected case by N (values: lower, upper).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">noov</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">No overlapping of oligos allowed if value = 1.
Disable the detection of overlapping matches for self-overlapping patterns (ex TATATA, GATAGA).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">str</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Oligonucleotide occurrences found on both stands are summed (2) or not (1). Default is 2.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">class_int</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Class interval. The width of the position classes, in number of bases (default: 20)</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">origin</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Reference for computing positions. The value should be chosen according to the sequence type (start, for downstream sequences; end, for promoter sequences; center, for ChIP-seq peaks).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">offset</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Add an offset to site position (positive or negative integer). Allows to select an arbitrary position as origin.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">group_rc</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">group reverse complement pairs if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">sort</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sort oligomers according to the bias in distribution profile if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">return</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Fields to return. Several fields can be entered, comma separated (values: distrib, for occurences found in each position class; exp, for expected occurences in each class; graph, for a graph file per oligo profile; chi, for chi-square value; rank).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">lth_chi</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Lower threshold on chi2.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">lth_sig</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Lower threshold on significance.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">lth_occ</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Lower threshold on the number of occurences.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">uth_rank</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Upper threshold on rank.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">max_graphs</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximum number of graphs to export.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">pattern</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A selection of patterns you want the analysis to be restricted to. Newline separated. A score can be associated to each pattern with the option score_column</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_pattern_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with input patterns on the server. You need to supply either this parameter or the previous one (pattern).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">score_column</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The column containing a score value for each pattern supplied with options pattern or tmp_pattern_infile. Only valid in combination with one of these options.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">min_pos</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Minimal position to take into account for the chi2 computation. This value must be a multiple of the class interval.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">max_pos</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximal position to take into account for the chi2 computation. This value must be a multiple of the class interval.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">no_check</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Do not check the applicability condition on the chi2 if set to 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">no_filter</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Do not discard oligos which do not fit the condition of applicability. Instead, mark them by including the chi2 value between curly brackets if set to 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">image_format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Image format passed to XYgraph.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">title</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Title for the index table and position profile plots.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">pattern_assembly</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Assemble a set of oligonucleotides or dyads into groups of overlapping patterns (assemblies).<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">input</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input data</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input file on the server (workflow usage)</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">verbosity</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Verbosity.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">score_col</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Score column. Column of the input file that contains the scores.
If not specified, patterns are incorporated according to their order in the input file.
However, if the input file has been generated with oligo-analysis or dyad-analysis with a verbosity &gt;= 1,
pattern-assembly detects the occ_sig column in the file header and uses this column as score column.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">str</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Strands for the assembly: 1 for single-strand; 2 for two-strand assembly. Default is 2.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">maxfl</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximum flanking segment size. Default is 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">subst</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximum allowed substitutions. Default is 0.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">max_asmb_nb</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximum number of assemblies. Default is 0 (no limit).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">max_asmb_size</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximum assembly size, i.e. number of patterns per assembly. Default is 50.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">maxpat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximum number of allowed patterns. If the number of patterns exceeds this value, the program issues a fatal error.
Default is 0 (no limit). Not compatible with toppat option.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">toppat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximum number of patterns to analyze.
If the number of patterns in the input exceeds this value, the assembly is restricted with the top patterns only.
Default: 100. Not compatible with maxpat option.</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">dna_pattern</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Searches all occurrences of a pattern within DNA sequences.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">sequence</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input sequence. You need to supply either this parameter or the next one (tmp_infile).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with input sequence on the server. You need to supply either this parameter or the previous one (sequence).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input sequence format. Supported: IG (Intelligenetics), WC (wconsensus), raw, fasta. Default is fasta.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">subst</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of substitutions allowed.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">pattern</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Pattern to match. Use this option or the 'pattern_file' option</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">pattern_file</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">File with patterns to match. Use this option or the 'pattern' option</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_pattern_file</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">File located on the server with patterns to match (workflow usage).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">id</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Pattern identifier.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">origin</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Origin for the calculation of positions (0 for end of sequence).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">noov</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">No overlapping of oligos allowed if value = 1.
Disable the detection of overlapping matches for self-overlapping patterns (ex TATATA, GATAGA).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">score</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Score column. Column of the pattern file which contains the score</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">str</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Oligonucleotide occurrences found on both stands are summed (2) or not (1). Default is 2.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">sort</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sort oligomers according to overrepresentation if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">th</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Threshold on match count.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">return</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">List of fields to return. Multiple fields can be entered separated by commas.
Supported fields: colsum,counts,ct,limits,profiles,rank,rowsum,scores,sites,stats,table,total.
    sites: return match positions (default)
    limits: return start and end positions for each input sequence
    counts: return the count of matches per sequence
    rank: return the rank of the sequence (this is especially useful in combination with the option -sort)
    score: return a score per sequence, computed by summing the scores of the matching patterns
    ct: same as '-return counts', except that it returns the sum of matches in all the files of the sequence file list, instead of the count within each separate file
    table: return the count of pattern matches per sequence in the form of a table.  (one line per sequence, one column per pattern)
    colsum: (together with -return table) prints an extra column with the total occurrences per sequence
    rowsum: (together with -return table) prints an extra row with total occurrences per pattern
    total: (together with -return table) prints an extra column with the total occurrences per sequence and 
          an extra row with total occurrences per pattern.  (amounts to combine -colsum and -rowsum)
    stats: return matching statistics
    profiles: return matching profiles with sliding windows.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">convert_features</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Interconversions between various formats of feature description.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">input</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input data</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input file on the server (workflow usage)</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">from</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input format. Supported: dnapat,ft,gft,gff3,gff.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">to</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Output format. Supported: dnapat,ft,gft,gff3,gff.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">feature_map</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Draws a graphical map of features (e.g. results of pattern matching) in a set of sequences.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">features</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">a list of features (ORFs, regulatory sites). Each feature is represented by a single line, which should	provide the following information:
Input file columns:
	1. map label (eg gene name)
	2. feature type
	3. feature identifier (ex: GATAbox, Abf1_site)
	4. strand (D for Direct, R for Reverse),
	5. feature start position
	6. feature end position
	7. (optional) description 
	8. (optional) score
The standard input format assumes that these topics are provided in this order, separated by tabs. Start and end positions can be positive or negative.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with input features on the server (workflow usage). You need to supply either this parameter or the previous one (features).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">sequence</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input sequence(s).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_sequence_file</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Reference sequence file on the server (workflow usage)</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">sequence_format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Format of reference sequence file on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Output image format. Supported: png,jpg,ps,gif (default = jpg)</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">from</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Lower limit of the positions represented on the graph.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">to</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Upper limit of the positions represented on the graph.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">title</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Generic Title for the feature map.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">label</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Define the info to display for each feature. Valid keys are: id, strand, descr (feature descritption), pos (feature start and end positions).
Several keys can be entered separated by commas without space. ex: -label pos,id. (default = id).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">symbol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Associates a graphical symbol (i.e. rectangle, circle, buterfly, ...) to each feature. This is convenient to distinguish the features on black and white printings.
Mutually exclusive with the -dot option.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">dot</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A color dot is associated to each feature. This allows to distinguish overlapping structures on a color screen. Mutually exclusive with the -symbol option.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">mlen</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">map length (in pixels).Default is 600.Length refers to either height (for vertical maps) of width (for horizontal maps).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">mapthick</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Map thickness.Thickness refers to either width (for vertical maps) or height (horizintal maps). This parameter allows to change the thickness allocated to each map.
This is useful when labels are too large. Default is 150.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">mspacing</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Map spacing.The size of the border between maps (in pixel).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">origin</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">All coordinates are recalculated relative to this origin.This allows to display all coordinates with respect to the ORF start or transcription start site.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">legend</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Draws a legend on the graph, showing the symbol associated to each distinct feature.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">scalebar</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Draws a scale bar on the left of the graph.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">scalestep</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Step between annotations of the scale bar. If not specified, a reasonable step is calculated on basis of the scale bar range.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">scorethick</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Each feature is displayed with a thickness proportional to its score. Only positive scores are represented.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">maxscore</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">(only valid when -scorethick is active) Maximal allowed score value. Higher score values are clipped for the drawing.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">minscore</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">(only valid when -scorethick is active) Minimal allowed score value. Features with smaller score are not displayed.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">maxfthick</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Max feature thickness</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">minfthick</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Min feature thickness</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">htmap</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">HTML map. An HTML document is automatically generated, which includes the feature map GIF file as an HTML map.
 In other words, this document displays a figure with sensitive areas. Each time the mouse is positioned above a feature, 
information about this particuliar feature is displayed at the bottom of the browser window.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">mono</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Monochrome palette (for printing on black/white printer).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">orientation</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Orientation of the map. Valid values are "horiz" for an horizontal map (default) and "vertic" for a vertical map.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">select</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black"> Only display the features whose ID is in the provided id_list.
The id_list contains one or several IDs, separated by commas. IDs may be embraced in single quotes to allow multiple words within the IDs. 
Commas and single quotes are not allowed within an ID.
Example: -select 'gataag','gattag' only displays features identified by gataag or gattag.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">footprint_discovery</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Detect phylogenetic footprints by applying dyad-analysis in promoters of a set of orthologous genes.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">verbosity</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Level of detail for comments</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">genes</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A list of genes (one by line).
	      </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with a list of genes on the server. You need to supply either this parameter or the previous one (genes).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">all_genes</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Automatically analyze all the genes of a query genome, and store each result in a separate folder (the folder name is defined automatically).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">max_genes</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximum number of genes to analyze.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">output_prefix</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Prefix for the output files. If the prefix is not specified, the program can guess a default prefix, but this is working only if there is a single query gene or query file.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">query</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A list of genes (referenced array).
	      </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">sep_genes</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Search footprints for each query gene separately. The results arestored in a separate folder for each	gene. The folder name is defined automatically.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">organism</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Query organism, to which the query genes belong.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">taxon</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Reference taxon, in which orthologous genes have to be collected.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">index</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Generate an HTML index with links to the result files.
This option is used for the web interface, but can also be convenient to index results, especially when several genes or taxa are analyzed (options -genes, -all_genes, -all_taxa).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">lth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Lower threshold on some parameters. Format: parameter value.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">uth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Upper threshold on some parameters. Format: parameter value.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">return</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return fields for dyad-analysis. See dyad_analysis for a list</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">to_matrix</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Convert assembled patterns into position-specific scoring matrices (PSSM).
Caution ! This conversion can take time if the sequence set is large and if there are many assemblies.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">bg_model</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Allow the user to choose among alternative background model.
                                 - taxfreq
                                      Taxon-wide background model, computed by counting dyad frequencies in all the promoters of all the genes of the reference taxon.
                                 - monads
                                      Expected dyad frequencies are the product of monad frequencies observed in the input sequences.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">no_filter</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Accept all dyads, even if they are not found in the promoter of the query gene, in the query organism.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">infer_operons</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Infer operons in order to retrieve the promoters of the predicted operon leader genes rather than those located immediately upstream of the orthologs.
This method uses a threshold on the intergenic distance.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">dist_thr</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify here the intergenic distance threshold in base pairs.
Pair of adjacent genes with intergenic distance equal or less than this value are predicted to be within operon. (default : 55).</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">get_orthologs</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Get orthologuous genes.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">organism</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Organism. Words need to be underscore separated (example: Escherichia_coli_K12).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">taxon</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Reference taxon. Orthologs are returned for each supported organism belonging to the reference taxon.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">query</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">List of gene(s) for which you want to get orthologs.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">all</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Get orthologs for all the genes of the query organism if value = 1. This option is particularly convenient to extract phylogenetic profiles.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">nogrep</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Disable the pre-filtering with grep if value = 1.
This pre-filtering accelerates the selection of hits, but some grep versions do not support the -E option.
If this is the case on your operating system, you can still obtain the correct results by inactivating the grep filter.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">return</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Output field(s).
By default, the output is a two-column file indicating the ID of the gene identified as similar to the query gene, and the name of the reference organism.
The option -return can be used to specify additional output fields.
Supported fields:
       - ref_id: ID of the reference (target) gene
       - ref_organism: Name of the reference (target) organism
       - query_id: ID of the query gene
       - query_organism: Name of the query organism
       - ident: Percent of identity (a number between 0 and 100)
       - ali_len: Alignment lengths (in residues)
       - mismat: Number of mismatches
       - gap_open: Number of gap openings
       - e_value: E-value (expected number of false positives)
       - bit_sc: Bit score
       - rank: Rank
       - s_rank: Source rank (rank of the hit for the query organism).
Several output fields can be entered separated by commas.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">lth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Lower threshold for dyad-analysis. Format: list of 'parameter value'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">uth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Upper threshold for dyad-analysis. Format: list of 'parameter value'.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">infer_operon</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Infer operon.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">organism</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Organism. Words need to be underscore separated (example: Escherichia_coli_K12).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">query</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">List of gene(s) for which you want to infer the operon.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with query genes on the server. You need to supply either this parameter or the previous one (query).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">all</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Infer operons for all the genes of the query organism if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">distance</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Distance threshold.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">min_gene_nb</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Min number of genes to report the operon.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">return</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">List of fields to return.
Supported fields: leader,trailer,operon,query,q_info,up_info,down_info
       - leader: Predicted operon leader.
       - trailer: Predicted operon trailer.
       - operon: Full composition of the operon. The names of member genes are separated by a semi-column ";".
       - q_info: Detailed info on the query gene(s).
       - up_info: Detailed info on the upstream gene.
       - down_info: Detailed info on the downstream gene.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">gene_info</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Get information about genes.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">organism</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Organism. Words need to be underscore separated (example: Escherichia_coli_K12).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">query</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">List of gene(s) for which you want info on or list of keywords to search for (can be regular expressions).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">full</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Full match only (no substring matching) if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">noquery</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Do not print the query at the begining of each line if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">descr</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Match query against the description, too, not just against gene ID and name if value = 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">feattype</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Feature type (CDS, mRNA, tRNA, rRNA, scRNA).</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">supported_organisms</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">List RSAT suppported organisms.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">return</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return fields. Supported: ID,name,data,last_update,taxonomy,up_from,up_to,genome,seq_format,source.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Output format. Supported: html_list, html_table, array, text, keys, names, sizes, full, tree, html_tree.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">taxon</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Root taxon.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">text_to_html</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Converts a tab-delimited file into a HTML table<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputfile</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The text input file that will be converted in html</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">chunk</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Chunk size (when there are many rows, the program splits the table into several HTML tables, to reduce the waiting time on the browser.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">no_sort</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">make the output HTML table not sortable.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">font</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Use fixed or variable fonts in html. Supported : variable,fixed</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">parse_psi_xml</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Converts a psi xml file in a tab delimited file<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputfile</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The psi XML input file that will be converted in a tab file</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">channels</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A list of evidence channels to filter separated by a comma</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">interactor_type</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A list of interactor_type to filter (separated by comas) : protein, 'small molecule'</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">uth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Upper threshold on the value</span></pre><i>Type = Float</i></td></tr><tr align="left"><td height="35"><font color="#228888">lth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Lower threshold on the value</span></pre><i>Type = Float</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">roc_stats</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Computes,
 from a set of scored results associated with validation labels, the 
derived statistics (Sn, PPV, FPR), which can be further used to draw a 
ROC curve.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputfile</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The input file</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">scol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Column of the input ile containing the score value (default 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">lcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Column of the input file containing the status label (default 2).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">status</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">This option allows to use different labels as synonyms for the pre-defined status: pos and neg.
It can be useful to rename these labels, for compatibility with other programs.
For example, it your input file contains annotations of "site" and "non-site", you can use it directly as input with the options.
-status site pos -status non-site neg indicates that the label "site" has to be understood as positive, and "non-site" as negative.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">total</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Total number of elements in the universe (neg + pos).
This option allows to manually specify the total number of elements, in case the input file would not contain the complete data set.</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">xygraph</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Plot a graph and export it.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputFile</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The input file should contain columns with numeric data. Each line contains info about one point of the graph.
By default, the first column is considered to contain X data, and the second column Y data. X and Y columns can be changed with -xcol and -ycol options.
Columns should be separated by tabs.
              </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">format</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Supported : png,pdf,jpg,eps,gif</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">title1</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">first graph title</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">title2</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">second graph title</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">lines</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">points are jointed by lines</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">legend</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Use the content of the first line from input file as legend for Y data.
	      </span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">header</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">First line of the data file contains a column header. If option -legend is active, this header is used as legend, else it is ignored.
	      </span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">xleg1</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">first x legend</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">xleg2</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">second x legend</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">yleg1</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">first y legend</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">yleg2</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">second y legend</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">xmax</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">maximal value represented on X axis.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">ymax</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">maximal value represented on Y axis.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">xmin</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">minimal value represented on X axis.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">ymin</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">minimal value represented on Y axis.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">ylog</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Y data are displayed on a logarithmic scale. If the next argument is a number, it provides the log base. Default log base is 10.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">xlog</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">X data are displayed on a logarithmic scale. If the next argument is a number, it provides the log base. Default log base is 10.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">xcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Column containing data for the X axis.
A zero value indicates that there is no column with X va lues. In this case, X values are ordinal.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">ycol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Column containing data for the Y axis.
Several columns can be specified by -ycol #,#,#. A range of columns can be specified by -ycol #-#. They have to be separated by commas without spaces.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">classfreq</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">This script takes a group of numbers (real or integers) and outputs their distribution among classes.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputFile</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input file</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">classinterval</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">class interval. If not specified, takes the value (max - min)/20</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">col</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">column to which apply the program. This option can be used iteratively.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">min</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">numbers strictly smaller than this value are not taken into account</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">max</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">numbers strictly higher than this value are not taken into account</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">from</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Inferior limit for the classes to display. Values lower than this limit are however taken into account in the calculation of statistics (avg, variance, ...) 
and of class frequencies (In contrast with the -min option).
	      </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">to</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Superior limit for the classes to display. Values higher than this limit are however taken into account in the calculation of statistics (avg, variance, ...) 
and of class frequencies (In contrast with the -min option).
	      </span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">convert_seq</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Converts a sequence between two formats (e.g. fasta -&gt; raw).<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">sequence</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sequence to convert.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with input sequence on the server. You need to supply either this parameter or the previous one (sequence).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">from</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Format of input sequence (embl, fasta, filelist, gcg, genbank, ig, maf, multi, ncbi, raw, tab, wc, wconsensus).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">to</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Format of output sequence (fasta, filelist, ft, ig, multi, raw, tab, wc, wconsensus).</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">compare_classes</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Compare
 two class files(the query file and the reference file). Each class of 
the query file is compared to each class of the reference file. The 
number of common elements is reported, as well as the probability to 
observe at least this number of common elements by chance alone.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">ref_classes</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A tab-delimited text file containing the description of reference classes</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">query_classes</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A tab-delimited text file containing the description of query classes</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">return_fields</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">List of fields to return separated by a comma. Supported: dotprod, entropy, freq, jac_sim, members, occ, proba, rank</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">score_column</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input file containing a score associated to each member</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">input_classes</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">This file will be used as both reference and query.This is equivalent to -q input_file -r input_file.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">upper_threshold_field</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Supported fields: E(QR), E_val, F(!Q!R), F(Q!R), F(Q), F(QR), F(R!Q), F(R), H(Q), H(Q,R), H(Q|R), H(R), H(R|Q), 
I(Q,R), IC, P(QR), P(Q|R), P(R|Q), P_val, Q, QR, QvR, R, U(Q|R), U(R|Q), dH(Q,R), dotprod, jac_sim, rank, sigE(QR), 
E_val, F(!Q!R), F(Q!R), F(Q), F(QR), F(R!Q), F(R), H(Q), H(Q,R), H(Q|R), H(R), H(R|Q), I(Q,R), IC, P(QR), P(Q|R), 
P(R|Q), P_val, Q, QR, QvR, R, U(Q|R), U(R|Q), dH(Q,R), dotprod, dp_bits, jac_sim, log2_dp, names, prodrts, sig, sqrt_dp.
Separate with semicolumn (":") as there are fields containing commas.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">upper_threshold_value</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Upper threshold value for a supported field. There must be as many thresholds values as thresholds fields in the same order as the list of threshold fields were given. </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">lower_threshold_field</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Supported fields: same fields as upper_threshold_field.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">lower_threshold_value</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Lower threshold value for a given field.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">population</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Population size. If not specified, the population size is estimated as the number of distinct elemenst in the whole set of reference classes.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">sort</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sort on the basis of the specified key.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">distinct</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Prevent to compare each class with itself (when the reference and query files contain the same classes)</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">triangle</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">(only valid if query file and reference file are the same) Do not perform the reciprocal comparisons.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">matrix</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return a pairwise matrix, where each row corresponds to a reference class, each column to a query class, and each cell contains a comparison between the two classes. 
The next argument indicates which statistics has to be return in the matrix (default = sig) 
Supported: E(QR), E_val, F(!Q!R), F(Q!R), F(Q), F(QR), F(R!Q), F(R), H(Q), H(Q,R), H(Q|R), H(R), H(R|Q), I(Q,R), IC, P(QR), P(Q|R), P(R|Q),
P_val, Q, QR, QvR, R, U(Q|R), U(R|Q), dH(Q,R), dotprod, jac_sim, rank, sig</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">multi_correction</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Factor used for the multi-testing correction.
Supported values:
  nt  number of significance tests (default)
  nq  number of query classes
  nr  number of reference classes
  nc  number of comparisons (nc = nq * nr)</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">convert_classes</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Interconversions between different formats of cluster files.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">informat</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input format : supported tab, profiles, mcl</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">outformat</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Output format : supported tab, profiles, mcl</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">member_col</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Member column. Column containing the member names in the tab format (default 1).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">class_col</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Class column. Column containing the class names in the tab format  (default 2).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">score_col</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Score column. Column containing the score in the tab format : if not specified, scores are not defined..</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">min_score</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Minimal score value for member to class assignation.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputclasses</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input classes in the format defined by the input_format tag</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">names</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Two tab-delimited columns specifying the labels of the members of the classification given in the classification file.
First column contains the identifier and second column the corresponding label.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">contingency_stats</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">This
 programs takes as input a contingency table, and calculates various    
 matching statistics between the rows and columns. The description of 
these statistics can be found in Brohee and van Helden (2006).<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputfile</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A contingency table : a N*M table used to compare the contents of two classifications. 
Rows represent the clusters of the first classification (considered as reference), and columns the clusters of the second classification (query).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">decimals</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of decimals to display for the computed statistics.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">return</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">List of fields to return.
    stats : table-wise statistics
    rowstats : row-wise statistics (one line per row of the contingency table)
    colstats : column-wise statistics (one line per column of the contingency table)
    tables : full tables for each statistics (counts, Sn, PPV, separation).
    margins : marginal statistics besides the tables (requires to return tables).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">rsizes</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify row group sizes in a separate file.
This option can be used in particular cases where the marginal sum of the contingency table does not correspond to the group sizes 
(for example if a classification supports the same elements assigned to multiple groups, or on the contrary if some elements can be unassigned).
The row size file must contain one row per row of the contingency table, and two columns. The first column indicated the name of the row 
(the same name as in the contingency table), and the second the size of the corresponding group.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">csizes</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify column group sizes in a separate file. Same description as for -rsizes tag.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">contingency_table</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3"> Create a contingency table from a two-column file.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputfile</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A tab delimited file</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">col1</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">First column to use for the contingency table</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">col2</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Second column to use for the contingency table</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">margin</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Calculate the marginal sums</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">null</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Value for the null character (default: 0).</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">compare_matrices</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Compare
 two collections of position-specific scoring matrices (PSSM), and 
return various similarity statistics + matrix alignments (pairwise, 
one-to-n).<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">matrix_1</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The first input containing one or several matrices. Matrix format is specified with the option format1 or format (see below).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">matrix_2</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The second input containing one or several matrices. Matrix format is specified with the option format2 or format (see below).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">matrix</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Single input containing one or several matrices. Each matrix of this file is compared to each other. Matrix format is specified with the option format (see below).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_matrix1_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the first file containing matrix/ces on the server. You need to supply either this parameter or matrix_1 (or matrix or tmp_matrix_infile).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_matrix2_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the second file containing matrix/ces on the server. You need to supply either this parameter or matrix_2 (or matrix or tmp_matrix_infile).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_matrix_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the single file containing matrix/ces on the server. You need to supply either this parameter or matrix (or use matrix_1 + matrix_2).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">format1</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Martix format for the first input. Supported fields: tab, cb, consensus, gibbs, meme, assembly.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">format2</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Matrix format for the second input. Supported fields: tab, cb, consensus, gibbs, meme, assembly.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Matrix format for both inputs. Supported fields: tab, cb, consensus, gibbs, meme, assembly.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">background_model</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Background model is tab-delimited and contains the specification of oligonucleotide frequencies.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_background_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Background model file on the server. It is a tab-delimited file containing the specification of oligonucleotide frequencies.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">background_format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Format for the background model file. Supported formats: all the input formats supported by convert-background-model.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">top1</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Only analyze the first X motifs of the first file. This options is convenient for quick testing before starting the full analysis.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">top2</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Only analyze the first X motifs of the second file. This options is convenient for quick testing before starting the full analysis.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">output_prefix</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Prefix for the output files. The output prefix is mandatory for some return fields (alignments, graphs, ...). This prefix will be appended with a series of suffixes for the different output types (see section OUTPUT FORMATS above for the detail).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">mode</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">-format matches (default)
            Return matches between any matrix of file1 and any matrix of file2.

            This is the typical use of compare-matrices: comparing one or several query motifs (e.g. obtained from pattern discovery) with a collection of reference motifs (e.f. a database of experimentally characterized transcription factor binding motifs, such as JASPAR, TRANSFAC, RegulonDB, ...).

            For a given pair of matrices (one from file1 and one from file2), the program tests all possible offsets, and measures one or several matching scores (see section "(Dis)similarity metrics" above). The program only returns the sore of the best alignemnt between the two matrices. The "best" alignement is the combination of offset and strand (with the option -strand DR) that maximizes the default score (Ncor). Alternative scores can be used as optimality criteria with the option -sort.

        -format profiles
            Return a table with one row for each possible alignment offset between two matrices, and various columns indicating the matching parameters (offset, strand, aligned width,...), the matching scores, and the consensus of the aligned columns of the matrices.

            Matching profiles are convenient for drawing the similarity profiles, or for analyzing the correlations between various similarity metrics, but they are too verbosy for the typical use of compare-matrices (detect matches between a query matrix and a database of reference matrices). The formats "matches" and "table" are more convenient for basic use.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">distinct</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Skip comparison between a matrix and itself.

        This option is useful when the program is sused to compare all matrices of a given file to all matrices of the same file, to avoid comparing each matrix to itself.

        Beware: the criterion for considering two matrices identical is that they have the same identifier. If two matrices have exactly the same content (in terms of occurrences per position) but different identifiers, they will be compared.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">strand</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Perform matrix comparisons in direct (D) reverse complementary (R) or both orientations (DR, default option).

        When the R or DR options are activated, all matrices of the second matrix file are converted to the reverse complementary matrix.

        This option is useful to answer very particular questions, for example

        Comparing motifs in a strand-insensitive way (-strand DR)
            DNA-binding motifs are usually strand-insensitive. A motif may be detected in one given orientation by a motif-discovery algorithm, but annotated in the reverse complementary orientation in a motif database. For DNA binding motifs, we thus recomment the DR option.

            On the contrary, RNA-related signals (termination, poly-adenylation, miRNA) are strand-sensitive, and should be compared in a single orientation (-strand D).

        Detecting reverse complementary palindromic motifs
            An example of reverse complementary palindromic motif is tCAGswwsGTGa. When a motif is reverse complementary palindromic, the matrix is correlated to its own reverse complement.

            Remark about a frequent misconception of biological palindromes

            Reverse complementary palindroms are frequent in DNA signals (e.g. transcription factor binding sites, restriction sites, ...) because they correspond to a rotational symmetry in the 3D structure. Such symmetrical motifs are often characteristic of sites recognized by homodimeric complexes.

            By contrast, simple string-based palindromes (e.g. CAGTTGAC) do absolutely not correspond to any symmetry on the biochemical point of view, because the 3D structure of the corresponding double helix is not symmetrical. The apparent symmetry is an artifact of the string-based representation, but the corresponding molecule has neither rotational nor translational symmetry.

            DNA signals can either be symmetrical (reverse complementary palindromes, tandem repeats) or asymmetrical.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">matrix_id</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Obsolete option for returning matrix names, Replaced by -return matrix_name. Maintained for backward compatibility.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">return</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">List of fields to return (only valid for the formats "profiles" and "matches").

        Supported return fields:

        offset
            Offset (shift) between the two compared matrices.

        cor
            Pearsons coefficient of correlatiojn.

        Ncor
            Normalized correlation (default sorting criterion for the matching mode).

        cov
            Covariance.

        SSD
            Sum of squared distances.

        NSW
            Normalized Sandelin-Wasserman similarity.

        SW
            Sandelin-Wasserman similarity.

        dEucl
            Euclidian distance

        NdEucl
            Normalized Euclidian distance.

        NsEucl
            Normalized Euclidian similarity.

        dKL
            Kullback-Leibler distance

        matrix_number
            Number of the matrices in the input files

        matrix_id
            Identifiers of the matrices

        matrix_name
            Names of the matrices

        matrix_ac
            Accession number of the matrices (TRANSFAC format makes a distinction between IDs and accession numbers).

        width
            Width of the matrices and the alignment

        strand
            Direct (D) or Reverse complementary (R) comparison

        offset
            Offset between the positions of the first and second matrix
        pos
            Relative positions the aligned matrices (start, end, strand, width)

        consensus
            Aligned consensus. The residues of the consensus corresponding to aligned columns are displaye, non-aligned columns are replaced by dots.

        offset_rank
            During pairwise alignment, scores are computed for each offset and offsets are sorted according to the sorting criterion. The offset_rank indicates the rank of an offset in this sorted list. This is a "within-alignment" rank, which is useful in profile mode.

        match_rank
            In matching mode, ranks can be computed for all the selected metrics, and a mean rank is computed.

        alignments_pairwise
            Shifted matrices resulting from the pairwise alignments.

        alignments_1ton
            Shifted matrices resulting from the 1-to-N alignments.

        alignments
            Shifted matrices resulting from the alignments (pairwise and 1-to-N).

        all
            All supported output fields, including all metrics.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">sort</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Field to sort the results. The sorting direction depends on the metric: ascending for dissimilarity metrics, decreasing for similarity metrics.

        Supported sort fields:

        offset, ascending (default sorting criterion for the profile mode)
        Ncor, decreasing (default sorting criterion for the matching mode)
        cor, decreasing
        cov, decreasing
        SSD, ascending
        SW, decreasing
        NSW, decreasing
        dEucl, ascending
        NdEucl, ascending
        NsEucl, decreasing
        dKL, ascending</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">lth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Lower threshold on some parameter. Format=list of 'parameter value'.
Supported fields: rank, dEucl, cor, cov, ali_len, offset</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">uth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Upper threshold on some parameter. Format=list of 'param value'.
Supported parameters: same as lth.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">matrix_scan</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Scan
 sequences with one or several position-specific scoring matrices (PSSM)
 to identify instances of the corresponding motifs(putative sites). This
 program supports a variety of background models (Bernoulli, Markov 
chains of any order).<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">sequence</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sequence(s) to scan - all the formats supported in RSAT can be used as input (default: fasta)</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_sequence_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with input sequence(s) on the server. You need to supply either this parameter or the previous one (sequence).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">matrix</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Matrix/ces to scan with. The matrix format is specified with the option "matrix_format" (see below) Default format: tab.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_matrix_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with input matrix/ces on the server. You need to supply either this parameter or the previous one (matrix).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">sequence_format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Supported fields: fasta (default), IG (Intelligenetics), WC (wconsensus), raw.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">matrix_format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Supported fields: tab, cb, consensus, gibbs, meme, assembly.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">n_treatment</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Treatment of N characters. These characters are often used in DNA sequences to represent undefined or masked nucleotides.
        skip
            N-containing regions are skipped.
        score
            N-containing regions are scored. The probability of an N is 1 for both the background model and the matrix. The N residues
            will thus contribute neither positively nor negatively to the weight score of the N-containing fragment. This option can be
            useful to detect sites which are at the border of N-containing regions, or in cases there are isolated N in the sequences.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">consensus_name</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Use the motif (degenerate) consensus as matrix name.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">pseudo</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Pseudo-count for the matrix (default: 1).
The pseudo-count reflects the possibility that residues that were not (yet) observed in the model might however be valid for future observations.
The pseudo-count is used to compute the corrected residue frequencies..</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">equi_pseudo</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">If this option is called, the pseudo-weight is distributed in an equiprobable way between residues.
By default, the pseudo-weight is distributed proportionally to residue priors, except for the -window option where equipseudo is default.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">top_matrices</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Only scan with the top # matrices per matrix file.
This option is valid for some file formats containing multiple matrices where top matrices are generally more informative.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">background_model</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Background model is a tab-delimited specification of oligonucleotide frequencies.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_background_infile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Background model file (tab-delimited specification of oligonucleotide frequencies) on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">organism</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">To use a precalculated background model from RSAT, choose the organism corresponding to the background model.
Works with background and markov options.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">background</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">To use a precalculated background model from RSAT. Works with organism and markov options.
Type of sequences used as background model for estimating expected oligonucleotide frequencies. 
Supported: upstream, upstream-noorf, upstream-noorf-rm, intergenic</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">background_input</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Calculate background model from the input sequence set.
This option requires to specify the order of the background model with the option markov.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">background_window</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Size of the sliding window for the background model calculation.
This option requires to specify the order of the background model with the option markov (suitable for short order model only markov 0 or 1)</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">markov</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Order of the markov chain for the background model. This option is incompatible with the option background.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">background_pseudo</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Pseudo frequency for the background models. Value must be a real between 0 and 1.
If this option is not specified, the pseudo-frequency value depends on the background calculation.
For -bginput and -window, the pseudo frequency is automatically calculated with the length (L) of the sequence following this formula:
    square-root of L divided by L+squareroot of L
For -bgfile, default value is 0.01.
If the training sequence length (L) is known, the value can be set by -bg_pseudo option to square-root of L divided by L+squareroot of L.</span></pre><i>Type = Float</i></td></tr><tr align="left"><td height="35"><font color="#228888">return_fields</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">List of fields to return. Supported fields: sites, rank, limits, normw, bg_model, matrix, freq_matrix, weight_matrix, distrib .</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">sort_distrib</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sort score distribution by decreasing value of significance, if value = 1.
By default, the score distributions are sorted by score (weight).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">lth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Lower threshold on some parameter. Format=list of 'parameter value'.
Supported fields: score, pval, eval, sig, normw, proba_M, proba_B, rank, crer_sites, crer_size, occ, occ_sum, inv_cum, 
exp_occ, occ_pval, occ_eval, occ_sig, occ_sig_rank</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">uth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Upper threshold on some parameter. Format=list of 'param value'.
Supported parameters: same as lth.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">str</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Scan 1 or 2 strands for DNA sequences.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">verbosity</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Level of verbosity (detail in the warning messages during execution)</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">origin</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Define the origin for the calculation of positions.
-origin -0 defines the end of each sequence as the origin.
The matching positions are then negative values, providing the distance between the match and the end of the sequence.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">decimals</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of decimals displayed for the weight score.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">crer_ids</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Assign one separate feature ID per CRER. This option is convenient to distinguish separate CRERs.</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">convert_matrix</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Performs
 inter-conversions between various formats of position-specific scoring 
matrices (PSSM). The program also performs a statistical analysis of the
 original matrix to provide different position-specific scores (weight, 
frequencies, information contents), general statistics (E-value, total 
information content), and synthetic descriptions (consensus).<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">matrix</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Matrix (or assembly or features) you want to convert.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">background_format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Format for the background model (prior) files. Supported: oligo-analysis, MotifSampler, meme, dyads.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">background_pseudo</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Pseudo frequency for the background models. Value must be a real between 0 and 1 (default: 0.01).</span></pre><i>Type = Float</i></td></tr><tr align="left"><td height="35"><font color="#228888">from</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input matrix format. Supported: alignace, assembly, cb, clustal, consensus, feature, gibbs, infogibbs, meme, motifsampler, tab, transfac.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">to</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Output matrix format. Supported: consensus, patser, tab, transfac.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">return</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Result type (matrix content). Supported: consensus, counts, frequencies, info, information, logo, margins, parameters, profile, sites, wdistrib, weights.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">sort</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">desc | asc | alpha sort_key.
Sort matrices according to the specified attribute (sort_key). The sorting can be done on numerical values, either in descending (desc) or ascending (asc) order.
It can also be done in alphabetical order (alpha).
The key must be one of the numeric parameters of the matrices (e.g. information.content, E-value, ...).
This option is convenient, for example, to sort matrices from MotifSampler according to their information content:
        -sort desc MS.ic.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">top</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximal number of matrices to return. 
Some of the input formats can contain several matrices in a single file (e.g. consensus, meme, MotifSampler).
By default, all the matrices are parsed and exported. The option -top allows to restrict the number of matrices to be exported.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">pseudo</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">pseudo-weight used for the calculation of the weight matrix (default: 1).</span></pre><i>Type = Float</i></td></tr><tr align="left"><td height="35"><font color="#228888">equi_pseudo</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">If value is 1, the pseudo-weight is distributed in an equiprobable way between residues.
By default, the pseudo-weight is distributed proportionally to residue priors.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">base</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Base for the logarithms used in the scores involving a log-likelihood (weight and information content). Default: exp(1) (natural logarithms).
A common alternative to natural logarithms is to use logarithms in base 2, in which case the information content is computed in bits..</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">decimals</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of decimals to print for real matrices (frequencies, weights, information) or to compute score distributions.
Warning: for the computation of score distributions, the computing time increases exponentially with the number of decimals.
We recommend to restrict the precision to 2 decimals for the weight, this is generally more than sufficient.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">perm</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of permuted matrices to return.
Matrix columns are permuted so that the total information content remains identical to the original matrix.
Note that the output format for permuted matrix is tab.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">max_profile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximal width of the profile histogram (units = number of characters).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">rc</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Convert the matrix to its reverse complement if value = 1.</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">matrix_distrib</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Returns the theoretical distribution of matrix weight within the defined background model.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">matrix_file</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The matrix file content. Matrix format is specified with the option "matrix_format" (see below) Default format: tab.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_matrix_file</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file containing matrix on the server. You need to supply either this parameter or the previous one (matrix_file).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">matrix_format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Supported fields: tab, cb, consensus, gibbs, meme, assembly.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">matrix_pseudo</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Pseudo counts to apply on the matrix. Default: 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">background</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Background model file is a tab-delimited file containing the specification of oligonucleotide frequencies.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">background_pseudo</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Pseudo frequency for the background models. Value must be a real between 0 and 1. Default: 0.01.</span></pre><i>Type = Float</i></td></tr><tr align="left"><td height="35"><font color="#228888">decimals</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of decimals for the matrix frequencies.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">background_format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Background format.upported formats: all the input formats supported by convert-background-model.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">random_seq</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Generates random sequences.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">sequence_length</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Length of sequence to generate.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">repetition</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of sequences to generate.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">format</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Format of sequence(s) to generate.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">line_width</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A newline character will be inserted in the sequence every # bases, where # is the number provided. 
Default is 70. A value of 0 will prevent newline insertion.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">type</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Type of sequence(s) to generate (protein | DNA | other).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">seed</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Seed for the random generator.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">alphabet</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Alphabet. Must be followed by residue frequencies expressed precisely this way:    a:t # c:g #</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">expfreq</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Expected frequencies of oligomers in sequence(s) to generate. Indicate the file that contains expected oligomer frequencies. 
When this option is used, the sequences are generated according to a Markov chain.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">tmp_expfreq_file</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the file with expected frequencies on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">bg_model</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Background model. Automatically load a pre-calibrated exected frequency file from the RSAT genome distribution. 
When this option is used, the options organism and oligo_length are also required, to indicate the organism and the oligonucleotide length, respectively.
This option is incompatible with the option expfreq.                 
Type of sequences used as background model for estimating expected oligonucleotide frequencies (supported models):
  - equi (equiprobable residue frequencies [default]), 
  - upstream (all upstream sequences, allowing overlap with upstream ORFs. Requires to speciy a model organism), 
  - upstream-noorf (all upstream sequences, preventing overlap with upstream ORFs. Requires to specify a model organism), and 
  - intergenic (intergenic frequencies. Whole set of intergenic regions, including upstream and downstream sequences. Requires to specify a model organism).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">organism</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Name of the organism when using a background model.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">oligo_length</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Length of oligomer when using a background model.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">length_file</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Length file. Allows to generate random sequences with the same lengths as a set of reference sequences.
The length file contains two columns : sequence ID (ignored) and sequence length.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">convert_graph</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Convert graphs between different formats<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">informat</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input format. Supported: tab, gml, adj_matrix.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">ecolors</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Compute an edge color for the GML output. The color intensity is proportional to the weight of the edge.
All weights in the column indicated by the -wcol argument must thus be real values.
Supported: green, blue, red, fire, grey.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">outformat</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Output format. Supported: tab, gml, dot, adj_matrix.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputgraph</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A graph in the format specified by the informat tag.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">wcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains an edge weight or an edge label for the tab-delimited format (no default).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">scol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the source nodes for the tab-delimited format (default = 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the target nodes for the tab-delimited format (default = 2).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">eccol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that containsthe color of the edge (no default).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">sccol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the color of the source node (no default).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tccol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the color of the target node (no default).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">pathcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify the column containing the paths</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">undirected</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The graph is considered as being undirected (useful for the adjacency matrix input and output).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">distinct_path</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The nodes belonging to different paths are duplicated with this option</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">layout</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify whether the disposition of each node has to be calculated using the $RSAT/bin/fr_layout program.
This option is only useful for GML output.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">ewidth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Calculate the edge width for the GML output. The width is proportional to the weight of the edge.
This value can only be computed for the GML output. All weights in the column indicated by the -wcol argument must thus be real values.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">target_xpos_col</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Column containing the X position of the target node</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">target_ypos_col</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Column containing the Y position of the target node</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">source_xpos_col</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Column containing the X position of the source node</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">source_ypos_col</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Column containing the Y position of the source node</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">alter_graph</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Alter a graph either by adding or removing edges or nodes<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">informat</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input format. Supported: tab, gml, adj_matrix.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">outformat</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Output format. Supported: tab, gml, dot, adj_matrix.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputgraph</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A graph in the format specified by the informat tag.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">wcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains an edge weight or an edge label for the tab-delimited format (no default).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">scol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the source nodes for the tab-delimited format (default = 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the target nodes for the tab-delimited format (default = 2).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">directed</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specifies whether the edges must be considered as directed, i.e., an     edge from node A to node B is different from an edge from B to A (by default, edges are not directed).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">duplicate</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specifies whether more than one edge may link two nodes. (by default, duplicated edges are not allowed).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">self</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Allows self loops (by default, self loops are not allowed)</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">target</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Nodes that have to be removed in the graph (if existing). The node names must be separated by comas.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">add_nodes</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of edges to add. This value can either be a percentage value or a discrete number.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">rm_nodes</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of edges to remove. This value can either be a percentage value or a discrete number.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">add_edges</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of edges to add. This value can either be a percentage value or a discrete number.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">rm_edges</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of edges to remove. This value can either be a percentage value or a discrete number.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">graph_cliques</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Find all cliques in a graph<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">informat</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input format. Supported: tab, gml, adj_matrix.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputgraph</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A graph in the format specified by the informat tag.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">scol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the source nodes for the tab-delimited format (default = 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the target nodes for the tab-delimited format (default = 2).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">min_size</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Minimum size of the clique to return.</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">display_graph</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Produces the figure of a graph<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">informat</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input format. Supported: tab, gml, adj_matrix.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">outformat</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Output format. Supported: ps, png, jpeg.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">ewidth</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Calculate the edge width for the GML output. The width is proportional to the weight of the edge. This value can only be
computed for the GML output. All weights in the column indicated by the -wcol argument must thus be real values.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputgraph</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A graph in the format specified by the informat tag.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">wcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains an edge weight or an edge label for the tab-delimited format (no default).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">scol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the source nodes for the tab-delimited format (default = 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the target nodes for the tab-delimited format (default = 2).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">eccol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that containsthe color of the edge (no default).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">sccol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the color of the source node (no default).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tccol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the color of the target node (no default).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">layout</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Calculates the layout according to the Fruchterman and Reingold algorithm.
This option must be provided if the input graph is not GML.
	      </span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">draw_heatmap</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Produces the figure of a heatmap<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">outformat</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Output format. Supported: png, jpeg.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">html</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Returns an HTML file that load the heatmap. The name of this file is the name of the output file on the server with the html extension </span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputfile</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A graph in the format specified by the informat tag.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">row_names</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Use this option if the first column contain the row names.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">no_text</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Using this option, the values are not written in the cells of the heatmap</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">col_width</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Width of the columns (in pixel). If the row height is to small, the label of the heatmap will not be indicated. (Default : 50 px)</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">row_height</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Height of the rows (in pixel). If the row height is to small, the label of the heatmap will not be indicated. (Default : 30 px)</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">min</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Minimal value of the heatmap. By default, this value is the minimal value of the input file.
If the specified value is larger than the minimal value of the heatmap, then the minimal value of the heatmap will be used as minimal value.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">max</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximal value of the heatmap. By default, this value is the maximal value of the input file.
If the specified value is smaller than the maximal value of the heatmap, then the maximal value of the heatmap will be used as maximal value.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">gradient</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Color of the intensity gradient of the heatmap. Default is grey.
Supported : green, blue, red, fire, grey.</span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">compare_graphs</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Computes the union / difference or intersection of two graphs<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">Qinformat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input format of query graph. Supported: tab, gml.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">Rinformat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input format of reference graph. Supported: tab, gml.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">outformat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Output format. Supported: tab, gml, dot.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">outweight</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Q (weights of the query graph), R (weights of the reference graph), sum (sum of the weights of the two graphs), 
mean (mean of the weights of the two graphs), mean.g (geometrical mean of the weights of the two graphs),
min (minimum weight), max (maximum weight), Q::R (weight of the two graphs) (default).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">Rinputgraph</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The reference graph in the format specified by the informat tag.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">Qinputgraph</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The query graph in the format specified by the informat tag</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">Qwcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the query input graph that contains an edge weight or an edge label (default none) for the tab-delimited format.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">Qscol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the query input graph that contains the source nodes for the tab-delimited format (default = 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">Qtcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the query input graph that contains the target nodes for the tab-delimited format (default = 2).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">Rwcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the reference input graph that contains an edge weight or an edge label (default none) for the tab-delimited format.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">Rscol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the reference input graph that contains the source nodes for the tab-delimited format (default = 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">Rtcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the reference input graph that contains the target nodes for the tab-delimited format (default = 2).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">return</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">intersection, union, difference, R.and.Q, Q.and.R, Q.or.R, Q.not.R, R.not.Q, Q.and.R+Q, 
Q.and.R+R, R.and.Q+Q, R.and.Q+R, intersection+Q, intersection+R.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">directed</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Indicates whether the graphs must be considered as directed, i.e., an arc from node A to node B is different from an arc from B to A.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">self</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Indicates whether the graphs can admit self-loops, i.e., an arc from a node to itself. 
Note that the graphs do not specially need to contain actual self-loops, the question is whether 
it would or not be acceptable for the considered input graphs to contain self-loops.</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">graph_node_degree</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Calculates the in / out / global degree for a selection of seed nodes<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">informat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input format. Supported: tab, gml.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">all</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The degree of all nodes will be computed.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputgraph</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A graph in the format specified by the informat tag.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">nodefile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A file containing the nodes for which you want to know the degree.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">wcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains an edge weight or an edge label (default none) for the tab-delimited format.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">scol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the source nodes for the tab-delimited format (default = 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the target nodes for the tab-delimited format (default = 2).</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">graph_topology</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Calculate the node degree, the closeness and the betweenness of each node and specifies if this node is a seed or a target node.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">informat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input format. Supported: tab, gml.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">all</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The statistics will be computed for all nodes.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">return</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">return type : degree, closeness, betweenness. More than one statistics can be returned by separating the fields with ','.
To return all implemented statistics, you can use all.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputgraph</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A graph in the format specified by the informat tag.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">nodefile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A file containing the nodes for which you want to know the degree and other statistics.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">directed</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specifies whether the graph is directed or not (i.e. edge A-B corresponds to edge B-A).
In this case, the betweenness and the closeness calculation will be rather different.
By default the graph is not directed.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">wcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains an edge weight or an edge label (default none) for the tab-delimited format.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">scol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the source nodes for the tab-delimited format (default = 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the target nodes for the tab-delimited format (default = 2).</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">graph_cluster_membership</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Map
 a clustering result onto a graph, and compute the membership degree 
between each node and each cluster, on the basis of egdes linking this 
node to the cluster.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">informat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input format. Supported: tab, gml.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputgraph</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A graph in the format specified by the informat tag.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">clusters</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">clustering as a tab-delimited file.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">stat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Determines how the node membership will be calculated (edge, weight, relw).</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">decimals</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of decimals to print for the membership. Note that by selecting this option, the entries of the membership-vectors (rows) won't sum up to 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">wcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains an edge weight or an edge label (default none) for the tab-delimited format.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">scol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the source nodes for the tab-delimited format (default = 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the target nodes for the tab-delimited format (default = 2).</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">random_graph</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Generate random graphs either from scratch of from an existing graph using different randomization models<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">informat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input format. Supported: tab, gml.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">outformat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Output format. Supported: tab, gml, dot.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputgraph</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A graph in the format specified by the informat tag.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">random_type</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Randomization type:
- scratch (de novo graph) : don't forget to specify the number of nodes and of edges,
- ER (Erdos-Renyii randomization) : corresponds to the randomization of a input graph, keeping the nodes and the number of edges but changing its characteristics,
- node_degree : each node will keep the same degree that in the input graph (edge randomization),
- node_degree_distrib : the global distribution of node degree will remain the same as in the input graph.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">wcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains an edge weight or an edge label (default none) for the tab-delimited format.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">scol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the source nodes for the tab-delimited format (default = 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the target nodes for the tab-delimited format (default = 2).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">edges</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of requested edges (for scratch randomization type).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">degree</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximal degree of the nodes in the ER random graph.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">nodes</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Number of requested nodes (for scratch randomization type).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">self</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Allows self loops (by default, self loops are not allowed)</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">mean</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Mean value on the weight of the arcs.
This argument can only be used with the scratch and ER randomization type and must be combined with the -sd option.</span></pre><i>Type = Float</i></td></tr><tr align="left"><td height="35"><font color="#228888">sd</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Standard deviation value of the weight of the arcs.
This argument can only be used with the scratch and ER randomization type and must be combined with the -mean option.</span></pre><i>Type = Float</i></td></tr><tr align="left"><td height="35"><font color="#228888">directed</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Indicates whether the graphs must be considered as directed, i.e., an arc from node A to node B is different from an arc from B to A.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">no_single</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Prevent the ER / scratch graph from containing nodes with no neighbour.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">duplicate</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specifies whether more than one edge may link two nodes (by default, duplicated edges are not allowed).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">col_conservation</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Only compatible with ER randomization of a graph. Source and target nodes stay source and target nodes in the randomized graph.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">normal</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">This option can only be used with ER randomization type and if the input graph is weighted. 
Using this option will generate randomly the weight of the output random graph according to a normal distribution of weights. 
The mean and standard deviation can then be chosen (-mean and -sd option) or will be calculated according to the weights the input graph.</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">graph_get_clusters</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Compares a graph with a classification/clustering file.<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">informat</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input format. Supported: tab, gml.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">return</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return format. Supported: table, cluster, graph.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">outformat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Output format. Supported: tab, gml, dot.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputgraph</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A graph in the format specified by the informat tag.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">clusters</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specification of the clusters to which belong the nodes.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">wcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains an edge weight or an edge label (default none) for the tab-delimited format.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">scol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the source nodes for the tab-delimited format (default = 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the target nodes for the tab-delimited format (default = 2).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">distinct</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">As some nodes may belong to more than one group, using this option will duplicate the nodes belonging to more than one group.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">induced</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Using this option, only the first column of the cluster file will be taken into account. 
The output graph will thus consist in the graph induced by all nodes of the first column.</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                             </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">graph_neighbours</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Find the neihbours up to a certain distance of a collection of seed nodes<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">informat</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Input graph. Supported: tab, gml.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">direction</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Direction of the neighbours (default all). Supported: in, out, all. This option cannot be used with the stat output and when specifying the number of steps</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">all</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The neighbours of all nodes will be searched.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">stats</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Only valid when the number of step is equal to 1. The output file is presented differently, with one line for each seed node.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">self</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Include each node in its neighborhood, with a distance of 0, even if there is no self-loop at this node. 
This allows to extract the node together with its neighborhood, rather than the neighborhood only (default). This option cannot be used with the stats option.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputgraph</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A graph in the format specified by the informat tag.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">seedfile</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A list of nodes for which you want to know the neighbours.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">wcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains an edge weight or an edge label (default none) for the tab-delimited format.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">scol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the source nodes for the tab-delimited format (default = 1).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tcol</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Specify a column of the input graph that contains the target nodes for the tab-delimited format (default = 2).</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">steps</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Maximal number of steps between a seed node and its neighbours. Default: 1.</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">mcl</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Clustering via Stijn van Dongen MCL algorithm<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputgraph</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A graph in the tab delimited format first column : source node, second column : target node, third column : edge weight</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inflation</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Sets the main inflation value. This value is the main handle for affecting cluster granularity. It is usually chosen somewhere in the range [1.2-5.0].
-I 5.0 will tend to result in fine-grained clusterings, and -I 1.2 will tend to result in very coarse grained clusterings.</span></pre><i>Type = Float</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">rnsc</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Clustering via Andrew King RNSC algorithm<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">output</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return type.
Accepted values: 'server' (result is stored on a file on the server), 'client' (result is directly transferred to the client),
'both' (result is stored on the server and transferred to the client), and ticket (an identifier, allowing to monitor the job status and retrieve the result when it is done, is returned to the client).
Default is 'both'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">inputgraph</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">A graph in the format required by RNSC : an adjacency list in which each edge appears only once.
The vertices are labelled with the integers 0, 1, ..., n-1. The list of neighbours for vertex v appears as v n_1 n_2 ... n_x -1.
The input network in the correct format can be produced by the convert-graph program.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">max_clust</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Allow no more than "num" clusters.  "num" must be between 2 and n, where n is the number of vertices in the graph.
If this option is not specified or an invalid value is given, n clusters are used.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tabulength</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Set the tabu length to "num". Default value is 1. Note that when the tabulist option is used, vertices can appear 
on the tabu list more than once and moving them is only forbidden when they are on the tabu list more than TabuTol times, where TabuTol is the tabu list tolerance.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">tabulist</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Set the tabu list tolerance to "num". Default value is 1.
The tabu list tolerance is the number of times a vertex must appear on the tabu list before moving it is forbidden.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">naive_stop</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Set the naive stopping tolerance to "num". Default value is 5. This is the number of steps that the naive scheme will continue without improving the best cost.
If you run the scaled scheme, using a higher naive stopping tolerance isn't likely to improve your results.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">scale_stop</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Set the scaled stopping tolerance to "num". Default value is 5. This is the number of steps that the scaled scheme will continue without improving the best cost.
Setting the tolerance to 0 will cause the algorithm to skip the scaled scheme.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">exp_nb</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Run "num" experiments. The best final clustering over all experiments will be written to file. Default is 1.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">div_freq</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Set the diversification frequency to "num". Without this option, no diversification will be performed.
If the shf_div_len flag is also used, then "num" is the shuffling diversification frequency.
If the -d flag is not used, then "num" is the destructive diversification frequency.
It is recommended that the shf_div_len flag is used, because destructive diversification isn't much help.</span></pre><i>Type = Integer</i></td></tr><tr align="left"><td height="35"><font color="#228888">shf_div_len</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Set the shuffling diversification length to "num". That means that the last "num" moves in the diversification period will be diversification moves.
Don't set this to be higher than the diversification frequency.</span></pre><i>Type = Integer</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                    </span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">command</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The stand-alone command executed on the server.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">The results.</span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">monitor</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Monitoring the status of a job<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">ticket</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Ticket of a job submitted with its output_choice set to 'ticket'.                                                                                                                                                                                                          </span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">status</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return status ('Running' or 'Done') of a job submitted with its output_choice set to 'ticket'.                                                                                                                                                                                             </span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr><tr><th align="left" valign="top"><table frame="void" rules="none" border="0" cellpadding="5" width="100%"><tbody><tr align="left" valign="top"><th align="left" valign="top" width="120"><font color="#888888" size="4">Method</font></th><td align="left" valign="top"><font color="#222288" size="4">get_result</font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Description</font></th><td align="left" valign="top"><font color="#444444" size="3">Get result of a job<br></font></td></tr><tr align="left" valign="top"><th align="left" valign="top"><font color="#888888" size="3">Parameters</font></th><td align="left" valign="top"><font color="#444444" size="3"><font color="#0000DD">Input Parameters</font><tr align="left"><td height="35"><font color="#228888">ticket</font><br>(required)</td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Ticket of a job submitted with its output_choice set to 'ticket'.                                                                                                                                                                                                           </span></pre><i>Type = String</i></td></tr><br><font color="#0000DD">Output Parameters</font><tr align="left"><td height="35"><font color="#228888">client</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Return result of job submitted with its output_choice set to 'ticket'.</span></pre><i>Type = String</i></td></tr><tr align="left"><td height="35"><font color="#228888">server</font><br></td><td><pre><span style="font-family:Times; font-size:12pt;color:black">Location of the result file on the server. This can be used as input for a further request.                                                                                                                                                                                             </span></pre><i>Type = String</i></td></tr><br></font></td></tr></tbody></table></th></tr></tbody></table></th></tr></tbody></table></body></html>