<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Transcription factor binding motif (TFBM) clustering</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>
<h1> Transcription factor binding motif (TFBM) clustering
<img class="toplogo" src="../../../doc/html/logo.jpg" alt="[R logo]">
</h1>
<hr>
<div align="center">
<a href="../../../doc/html/packages.html"><img src="../../../doc/html/left.jpg" alt="[Up]" width="30" height="30" border="0"></a>
<a href="../../../doc/html/index.html"><img src="../../../doc/html/up.jpg" alt="[Top]" width="30" height="30" border="0"></a>
</div><h2>Documentation for package &lsquo;TFBMclust&rsquo; version 1.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td width="25%"><a href="TFBMclust-package.html">TFBMclust-package</a></td>
<td>Transcription factor binding motif (TFBM) clustering</td></tr>
<tr><td width="25%"><a href="align.leaf.and.cluster.html">align.leaf.and.cluster</a></td>
<td>Align one single node with a cluster of (already aligned) motifs.</td></tr>
<tr><td width="25%"><a href="align.motifs.html">align.motifs</a></td>
<td>Align all the motifs, using as base the merge order of one hierarchical tree.</td></tr>
<tr><td width="25%"><a href="align.two.clusters.html">align.two.clusters</a></td>
<td>Align two already aligned clusters of motifs.</td></tr>
<tr><td width="25%"><a href="align.two.leaves.html">align.two.leaves</a></td>
<td>Align two single nodes.</td></tr>
<tr><td width="25%"><a href="alignment.test.html">alignment.test</a></td>
<td>Using a given agglomeration method, checks if the clusters will be aligned.</td></tr>
<tr><td width="25%"><a href="build.distance.matrix.html">build.distance.matrix</a></td>
<td>Build distance matrix from a motif comparison table</td></tr>
<tr><td width="25%"><a href="closest.or.farthest.motifs.ids.html">closest.or.farthest.motifs.ids</a></td>
<td>Identify the "central" or farthest leaves (motifs) between two clusters.</td></tr>
<tr><td width="25%"><a href="color.code.clusters.html">color.code.clusters</a></td>
<td>Assign one color to each cluster.</td></tr>
<tr><td width="25%"><a href="convert.hclust.to.JSON.html">convert.hclust.to.JSON</a></td>
<td>Convert 'hclust' tree to Newick format</td></tr>
<tr><td width="25%"><a href="convert.hclust.to.newick.html">convert.hclust.to.newick</a></td>
<td>Convert 'hclust' tree to Newick format.</td></tr>
<tr><td width="25%"><a href="define.clusters.bottom.up.html">define.clusters.bottom.up</a></td>
<td>Traverse the tree in a bottom-up approach to separate the motifs clusters.</td></tr>
<tr><td width="25%"><a href="draw.heatmap.motifs.html">draw.heatmap.motifs</a></td>
<td>Draw a heatmap of the input motifs, using the same distance tab and the same aggolmeration method.</td></tr>
<tr><td width="25%"><a href="fill.downstream.html">fill.downstream</a></td>
<td>Add gaps to the downstream end of the consensus</td></tr>
<tr><td width="25%"><a href="get.comparison.number.html">get.comparison.number</a></td>
<td>Given two sets of IDs, return the number of comparison from the given comparison table</td></tr>
<tr><td width="25%"><a href="get.comparison.numbers.html">get.comparison.numbers</a></td>
<td>Given two sets of IDs, return the number of comparison from the given comparison table</td></tr>
<tr><td width="25%"><a href="get.consensus.html">get.consensus</a></td>
<td>Extract the motif IUPAC consensus (or its reverse complement) from a reference table.</td></tr>
<tr><td width="25%"><a href="get.id.html">get.id</a></td>
<td>Extract the id of a motif from a reference table.</td></tr>
<tr><td width="25%"><a href="get.name.html">get.name</a></td>
<td>Extract the name of a motif from a reference table.</td></tr>
<tr><td width="25%"><a href="get.spacer.nb.html">get.spacer.nb</a></td>
<td>Counts the number of gaps ("-") in both ends of a consensus string.</td></tr>
<tr><td width="25%"><a href="hclust.motifs.html">hclust.motifs</a></td>
<td>Hierarchical clustering of the motifs.</td></tr>
<tr><td width="25%"><a href="identify.JSON.tree.branches.html">identify.JSON.tree.branches</a></td>
<td>Identify the order of the branches in the JSON tree, relative to the order of one hclust tree.</td></tr>
<tr><td width="25%"><a href="inverted.alignment.html">inverted.alignment</a></td>
<td>Invert the aligned consensus and orientation information for a group of aligned motifs.</td></tr>
<tr><td width="25%"><a href="leaves.per.node.html">leaves.per.node</a></td>
<td>Collect the nodes of each node of the tree.</td></tr>
<tr><td width="25%"><a href="oct4.tfbm.compa.html">oct4.tfbm.compa</a></td>
<td>Comparison matrix between 21 TFBM discovered by RSAT peak-motifs in Oct4 ChIP-seq peaks</td></tr>
<tr><td width="25%"><a href="TFBMclust-package.html">TFBMclust</a></td>
<td>Transcription factor binding motif (TFBM) clustering</td></tr>
</table>
</body></html>
